{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Новий розділ"
      ],
      "metadata": {
        "id": "Ypo2u1wk1L0z"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Спочатку ця клітинка, запускати і далі **УВАЖНО!!** читати інструкції"
      ],
      "metadata": {
        "id": "C4tWdx58vrDr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Install and load the necessary package\n",
        "install.packages(\"usethis\")\n",
        "library(usethis)\n",
        "\n",
        "# Specify the GitHub repository URL\n",
        "repo_url <- \"https://github.com/PacktPublishing/R-Bioinformatics-Cookbook.git\"  # Replace with the actual repository URL\n",
        "\n",
        "# Clone the repository using system commands\n",
        "system(sprintf(\"git clone %s\", repo_url), intern = TRUE)\n",
        "setwd('/content/R-Bioinformatics-Cookbook')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 104
        },
        "id": "fbmPguHmoXqa",
        "outputId": "a67b8d6d-364c-47ee-de22-7b80723821e6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "Warning message in system(sprintf(\"git clone %s\", repo_url), intern = TRUE):\n",
            "“running command 'git clone https://github.com/PacktPublishing/R-Bioinformatics-Cookbook.git' had status 128”\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [],
            "text/markdown": "",
            "text/latex": "",
            "text/plain": [
              "character(0)\n",
              "attr(,\"status\")\n",
              "[1] 128\n",
              "attr(,\"errmsg\")\n",
              "[1] \"Resource temporarily unavailable\""
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Далі тут будуть коменти і назви проектів з кукбук, ваші скрипти повинні працювати прям з гітхабу, нічо не змінюючи, окрім додавання рядків з інсталлюванням пакаджів\n",
        "\n",
        "\n",
        "скрипти тут: https://github.com/PacktPublishing/R-Bioinformatics-Cookbook\n",
        "в папках chapter шукайте свій рецепт, нижче приклад, треба **вставити комірку з кодом** і туди просто свій скрипт вставити\n",
        "\n",
        "\n",
        "**Підписуйте** свої проекти над комірками з кодом, **вставляючи** комірку з текстом\n",
        "\n",
        "Також беріть до уваги що пакаджи будуть грузитись доволі таки довго і це **нормально**"
      ],
      "metadata": {
        "id": "UnYrQ24oTNln"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#приклад рядків встановлення\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "\n",
        "BiocManager::install(\"Змінити_на_назву_свого_пакедж\")\n",
        "#або, якщо це не біокондактор, а таке наче рідко бувало\n",
        "install.packages(\"Змінити_на_назву_свого_пакедж\")"
      ],
      "metadata": {
        "id": "0XXe6cNi2WTz",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ba040cfb-0b5e-4988-f5c3-1a7e47912746"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'Змінити_на_назву_свого_пакедж'\n",
            "\n",
            "Warning message:\n",
            "\"package 'Змінити_на_назву_свого_пакедж' is not available for Bioconductor version '3.18'\n",
            "\n",
            "A version of this package for your version of R might be available elsewhere,\n",
            "see the ideas at\n",
            "https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages\"\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "Installing package into '/usr/local/lib/R/site-library'\n",
            "(as 'lib' is unspecified)\n",
            "\n",
            "Warning message:\n",
            "\"package 'Змінити_на_назву_свого_пакедж' is not available for this version of R\n",
            "\n",
            "A version of this package for your version of R might be available elsewhere,\n",
            "see the ideas at\n",
            "https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages\"\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Вставляючи ваші скрипти, не забувайте **додати рядки, в яких відбувається встановлення бібліотек**, так як в гіт їх не написали"
      ],
      "metadata": {
        "id": "JQeJF8Cy0aDK"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Plotting features on genetic maps with karyoploteR"
      ],
      "metadata": {
        "id": "yWVJXxl4T1xX"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "IgnQvcQlS9e7",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "3caf6f5b-a51f-4b08-cf62-794d4a9441d9"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'karyoploteR'\n",
            "\n",
            "also installing the dependencies ‘formatR’, ‘lambda.r’, ‘futile.options’, ‘checkmate’, ‘htmlwidgets’, ‘abind’, ‘SparseArray’, ‘filelock’, ‘matrixStats’, ‘futile.logger’, ‘snow’, ‘BH’, ‘rjson’, ‘gridExtra’, ‘htmlTable’, ‘viridis’, ‘Formula’, ‘S4Arrays’, ‘DelayedArray’, ‘ProtGenerics’, ‘lazyeval’, ‘plogr’, ‘BiocFileCache’, ‘png’, ‘BSgenome’, ‘Biostrings’, ‘BiocGenerics’, ‘XVector’, ‘zlibbioc’, ‘bitops’, ‘BiocParallel’, ‘Rhtslib’, ‘XML’, ‘RCurl’, ‘GenomicAlignments’, ‘BiocIO’, ‘restfulr’, ‘GenomeInfoDbData’, ‘Hmisc’, ‘dichromat’, ‘SummarizedExperiment’, ‘ensembldb’, ‘AnnotationFilter’, ‘RSQLite’, ‘biomaRt’, ‘Biobase’, ‘KEGGREST’, ‘MatrixGenerics’, ‘regioneR’, ‘GenomicRanges’, ‘IRanges’, ‘Rsamtools’, ‘rtracklayer’, ‘GenomeInfoDb’, ‘S4Vectors’, ‘biovizBase’, ‘bezier’, ‘GenomicFeatures’, ‘bamsignals’, ‘AnnotationDbi’, ‘VariantAnnotation’\n",
            "\n",
            "\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'GenomicRanges'”\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "Loading required package: regioneR\n",
            "\n",
            "Loading required package: GenomicRanges\n",
            "\n",
            "Loading required package: stats4\n",
            "\n",
            "Loading required package: BiocGenerics\n",
            "\n",
            "\n",
            "Attaching package: ‘BiocGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    combine, intersect, setdiff, union\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    IQR, mad, sd, var, xtabs\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
            "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
            "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
            "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
            "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
            "    table, tapply, union, unique, unsplit, which.max, which.min\n",
            "\n",
            "\n",
            "Loading required package: S4Vectors\n",
            "\n",
            "\n",
            "Attaching package: ‘S4Vectors’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    first, rename\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:utils’:\n",
            "\n",
            "    findMatches\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    expand.grid, I, unname\n",
            "\n",
            "\n",
            "Loading required package: IRanges\n",
            "\n",
            "\n",
            "Attaching package: ‘IRanges’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:dplyr’:\n",
            "\n",
            "    collapse, desc, slice\n",
            "\n",
            "\n",
            "Loading required package: GenomeInfoDb\n",
            "\n",
            "\n",
            "Attaching package: ‘GenomicRanges’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:magrittr’:\n",
            "\n",
            "    subtract\n",
            "\n",
            "\n",
            "No predefined canonical chromosomes found for the requested genome. Applying a heuristic chromosome filtering.\n",
            "\n",
            "To get the unfiltered genome, please set chromosomes=\"all\" in the plotKaryotype call\n",
            "\n",
            "No predefined canonical chromosomes found for the requested genome. Applying a heuristic chromosome filtering.\n",
            "\n",
            "To get the unfiltered genome, please set chromosomes=\"all\" in the plotKaryotype call\n",
            "\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "plot without title"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAADAFBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJyco\nKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6\nOjo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tM\nTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1e\nXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29w\ncHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGC\ngoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OU\nlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWm\npqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4\nuLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnK\nysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc\n3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u\n7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////i\nsF19AAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3dC3gU5b348XdJNjEECJeCXIxc\nvIsICHK0UGtrFIWgaFWuChKoJQpShUasSr3UeMBab3io1ONB6+VvQSv0tHhaEBUlFbFFaLCN\niIpQBQxYUCCEvP/Z3Swk7Mxkd+f3zibL9/M8bia/zMy7Ps9+k80ySZQG4JlK9R0A0gEhAQII\nCRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQII\nCRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQII\nCRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQII\nCRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQII\nCRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEHH0hVZev3pvq+4C0cxSF9NZV\nvYev0RVnKNVybqrvC9LN0RNSWVAFVauNA3PHXNFCLU71vUGaOXpCKgy+VP1Zr7EZb2r9j9yC\nVN8bpJmjJ6R2Y62bZeq80Pb4Nim+M0g3R09IwVnWzR71o9D2bZmpvS9IO0dPSN2vDd3m3Rq6\nHXFsau8L0s7RE1JR9pvRzVXBH6TyniANHT0hVbQJzIxsjQ1mvpPa+4K0c/SEpMsLbo9s9Mp/\nJbX3BOnnKArpkC2pvgNIP6ZDqlr/7r7azU1fGl4rfpWbUn0PkGaMhbTs/G6XlOmlnZVqVXtB\njjrm3v2mVovH2iFdB82tDm+WHI1fiGGSqUfU25mqVbPct1vlX3t1G/XHyFLdMk57zdBycViZ\nrZoH1XcrQ9uEBGGmHlHDOq7V2753fO9vrOdR3S6OLFWyuq+64M0GDjRmaPDlmn0PBs/eowkJ\n4kw9otrdY92sVv8T2r63bWSpEl39y/bqvAW7DK3pLj90iZBeljWkmpAgztQjKvNp62ar+t/Q\n9pORC3KskLTeU9pOZZxddO8jhtZ1Frwz/OZpNZWQIM7UI+rYWdbNCvVwaPu2yAU54ZC0/vqZ\nYblK+f9IPu7SyNuZajYhQZqpR9TItsv3v9/rtOM/07q8zZWRpUqiH9z/1xceN7Sus6mBR6tC\nb2vGqWlTCAmyTD2iNrS0vuq0Le/a/HvnZmb8JbJUSQPHmLXjeBX5KaSaqan4goj0ZuwRtW7U\nf4z/QK8bEFA9fheZZN9uaq34bC+eVru16ARCgizjj6jd28Jvqp7ffuRHbEZ+SvHySC9+fWre\nqWL+Aclm5KcUL4/0QkiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBJgNaet7y9dG\nrrU7sKTSceSnFC+PdGUypCe6qZBTn3cd+SnFyyN9GQzpcZVdMK54zICAWuAy8lOKl0caMxjS\nyYMjv+XkoxNPdxn5KcXLI40ZDClrZe3G3GyXkZ9SvDzSmMGQ2q+q3ZjXxWXkpxQvjzRmMKQJ\nt9VuFE5zGfkpxcsjjRkMaeuA0Ys3fFK+cEhBxWaLw8hPKV4eacxgSKo+h5GfUrw80pjBR9Pw\nEfU4jPyU4uWRxvi0DAggJECAwZBqXizs0zPCZeSnFC+PNGYwpDlKNc+LcBn5KcXLI40ZDOm4\nwRvjGPkpxcsjjRkMKVgWz8hPKV4eaczkV6RV8Yz8lOLlkcYMhjSjOJ6Rn1K8PNKYwZB2Dx69\ntLwizGXkpxQvjzTmzyVCLiM/cXkQTDH4iBo1rijKZeSnFC+PNManZkCA0ZCqrf/2lb1X4z7y\nU4qXR9oyGFJ18ZVab+qh1KDdLiM/pXh5pDGDIZWqm7UeEphc3KzUZeSnFC+PNGYwpDOu0Pqz\ngPVt/YQ+LiM/pXh5pDGDIbWYp/WT6s9az23tMvJTipdHGjMYUkvrYTsqd7/Wj+W6jPyU4uWR\nxkw+tRutP29xubUx6RSXkZ9SvDzSmMGQ7lPndlYrtF6QNcNl5KcUL480ZjCkveNz8h6x3nbq\nVeky8lOKl0caM35lw8KqVQfqT9aW6yNHfjqwMKXLIz2ZDmmnev/I0ZjrDa/pbr3akdL1kZZM\nh1Sp/nbkaPQPDa/pbp3antL1kZZMh7Rd/eLgEaOBFxle090q9a+Uro+0ZDqkg+fmnPBI/Svb\nvj/M8Jru3lVfpHR9pCXzP0ax7f4uraZ+WmfAUzukHz9+Hmnv/NMDTx9+t9c5Pqzp7A21NaXr\nIy3584N9NS0WHn7nsrG+rOnkfb4iQZ5PPyGb97vD2zy1Q/ohJEAAIQECCAkQQEiAAEICBBAS\nIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQ4FNI7f9weHvc\nDf6s6eAfgV0pXR9pKa6QilSFx2U+qj68vSPFfwrC6/8LECvpkKpubdZP/N4ATVSyIZWf1dIl\npI0vAk3ev82H9FVO/4ps55B+mNMFaOIynhcK6V9FnZuf+dCBUEgb7++elX93jdYj1BcFx7yi\nv7ylSruENPHiMqCJa/8bmZC2dcmb8kChKgqFdF3f0tn56jmtr1GjL7lvXfjjhIS0JhXSZPWq\ndTtUrbdCGlSl9Rp1qdYT1EXRv3dESEhrQiHVtMu3nsrpjcu3WyG9HBpk9A99cXo2ugMhIa0J\nhbRFXRjdLLK+KlnyeoY23yUkHBWEQvpQFR4OqeJwSIdewCMkpDWhkPaoQYSEo5jUiw3t21VZ\ntx88uj6JkM79H6CJayMU0kQ137odqdYkHtJMBTR1gaUyIW3umHnjnEJ1rd1TuxUlJSUZHa2b\nHQmsBaQttysbPh7bIdjjF9V2IZVGq+VSakD79vNIQHojJEAAIQECCAkQQEiAAEICBBASIICQ\nAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQ\nAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQ\nAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECfAlp/zvLP6r7/oHNGz7a\nJXj+mo1/eumlZZ8KnhFIjNmQ7lkeup3XRinV76+1s/2PnZNlva86jyuTWaTylg4q7Pi7v5E5\nI5AosyGpEuvm9yr78usHqrwPw6OvBqicM88Jnjimb0DNkFhja3d10vhZs2ffPqqz6l0pcUYg\nYT6EdFJeuXW7KHBdePRjNc16Vreh68P64+HqKYE1ioIv1m5Vzw3cJHBCIHHmQ9qmbgtvD+8S\nfnPc0PCb37Tao6v7nSWwRscJh7dH5AucEEic+ZA+Vc+Et28Pht8E7w6/2aBWan1XjsAawZ8f\n3v5ZlsAJgcSZD6k6rzS8PaFt+E2nq8NvFqrXtZ7aTmCNrlcf3r6sm8AJgcQZDmnU6ortM0/8\n2trckDssPBoXmF+j9ftdc/fod1pfLLDGTYE5+yJbe+4Mf08G+M9wSBELtX42t9k74dGmtqrz\neacH1GO6Opj7jsAaO89SLS8Yf+MN485vrr6zW+CEQOLMhvTUL2fdNG74+cu0nttlSe1s44iW\nKvM7oX9g+sk6kUX2P9gnI5Rr8JwnqkVOCCTMr0uEdh+s886eg477JWfvP9esqdgvfFIgfv5d\na7ejIp6R5BrV77/FZUPwh38hlcQuZTOSWeOtG6ybZ461nu/1fl14CcBOWob0WlaLGv1b1eKq\n4gubZb8rvAZgIy1DOr+D9RSve9et1mZZzjDhNQAbZkPqV0dH5TQSX6PVdK13qYfD25NaC6wB\nNMBsSM2aZR+SoZxG4mvk3qH1vsCi8PZdxwisATTAbEglLQ+/jFb7tMtmJL7GwJO+1vrb00Ob\n+3r3FlgDaIDZkKr69q+Kbtc+yG1G4mssUWe9emBNpwVfV5V9X/1KYA2gAYZfbCjPmR7djFZj\nMxJfY36uyjm9q8rIUIGbayTWANyZftXuqy+jWytKnUfia3w+Z3DXltnt+k1dI7IC0AB/f4tQ\n1fPb4xg1/jWAI/gb0k71Zhyjxr8GcARCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQ\nEiDAj5C2vrd87bbw1oEllY6jxr8G4Mh8SE90C/+21VOfdx01/jUAF8ZDelxlF4wrHjMgoBa4\njBr/GoAb4yGdPDjy12I/OvF0l1HjXwNwYzykrJW1G3OzXUaNfw3AjfGQ2q+q3ZjXxWXU+NcA\n3BgPacJttRuF01xGjX8NwI3xkLYOGL14wyflC4cUVGy2OIwa/xqAG+MhqfocRo1/DcCN8YfY\n8BH1OIwa/xqAGz5XAwIICRBgPKSaFwv79IxwGTX+NQA3xkOao1TzvAiXUeNfA3BjPKTjBm+M\nY9T41wDcGA8pWBbPqPGvAbgx/xVpVTyjxr8G4MZ4SDOK4xk1/jUAN8ZD2j149NLyijCXUeNf\nA3Dj5yVCLqPGvwbgxvjDbNS4oiiXUeNfA3DD52tAgA8hVVv/7St7r8Z91PjXAJwZD6m6+Eqt\nN/VQatBul1HjXwNwYzykUnWz1kMCk4ublbqMGv8agBvjIZ1xhdafBazv+Cf0cRk1/jUAN8ZD\najFP6yfVn7We29pl1PjXANwYD6ml9Ygelbtf68dyXUaNfw3AjfmndqP15y0utzYmneIyavxr\nAG6Mh3SfOrezWqH1gqwZLqPGvwbgxnhIe8fn5D1ive3Uq9Jl1PjXANz4dGXDb79cdaD+ZNcL\n+siRRwt3xJzw2T3CawD2fAqp/cIjJ79vKb3GsS/GjJq9Jr0IYMunkL712yMnS1pIr9Hh/8WM\nAsulFwFs+RRSzoVz39hVb/LroPga3ytdtG5fvZGK/SIFmOBTSNcNOzlDdS2c+UJ59HuW58T/\n3ErRFWfnqWbdL7rx0Vc3HYyMAr+XXgSw5d+PUexfv6CksKMKnn7VrMVbTTy1C6l898VZV/Vr\nrrJ6FEz91Z828tQOPvH755G2LP3Psb2z1Dn64Qxzi1R/+MeHigu6BhR/9xI+SckP9lVdcaV+\npbnxZb4Zr5YZXwQISc1PyE4ca+ipXX0/UTy1gz8ICRBASIAAQgIEEBIggJAAAYQECCAkQAAh\nAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEpCakG4r0n75lfpk7Mt42\nvwig4wypSFXIrrr7K12zRfaUdr4pN78GEJJsSJW3HJ/V7bJV8ncIaIqSDOnLbmroHWMyj3nf\n/oDyX6EpW+LxUXUUSjKkG9Sj1u0iNcT+gEk5XdB0tePv7ibMLaR/FXVufuZDB0Ihbby/e1b+\n3TVaj1BfFBzzip52QZW1Q01OV/sjJ15chqbrQfO/BTftuIS0rUvelAcKVVEopOv6ls7OV89p\nfY0afcl962r32BccaH8oITVphJQ4l5Amq1et26FqvRXSIOsL0Bp1qdYT1EUHD+3xcPgJng1C\natIIKXHOIdW0y7eeyumNy7dbIb0cGmT0D31xevbQHiuyBjn8hVZCatIIKXHOIW1RF0Y3i6yv\nSpa8nqHNd6PT57LP+tLhWEJq0ggpcc4hfagKo5u1r9pFQqp9Aa/mTnXxv52OJaQmjZAS5xzS\nHjUoumkTUs0ENaXa8VhCatIIKXEuLza0bxd6ifuDR9fbhXSTus/lrBPPfhRN148IKWEuIU1U\n863bkWqNTUiL1E1uZ/2ZQlN2suAj7CjhEtLmjpk3zilU19o9tTtBTSkJq/TnbgKNm9uVDR+P\n7RDs8Ytqu5AOfe7a5MedBBq71Pw8EpBmCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQ\nQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQ\nQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQ\nQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAP9D2lFR97397yz/qMFDqt9/69O4zl1d\nvnpvw2sC4vwPqSSy5D3LQ7fz2iil+v3Vcee3brBunjnW2qn3624nfeuq3sPX6IozlGo513lN\nwJSUhaRKrJvfq+zLrx+o8j502Pe1rBY1+reqxVXFFzbLftf5nGVBFVStNg7MHXNFC7XYcU3A\nlNSGdFJeuXW7KHCdw77nd7Cek3XvutXaLMsZ5nzOwuBL1Z/1Gpvxptb/yC1wXBMwxZ9HWL86\nOh4OaZu6Lbw9vIvDca2ma71LPRzentTa+fztxlo3y9R5oe3xbRzXBEzx5xHWrFn2IRmHQ/pU\nPRPevj3ocFzuHVrvCywKb991jPP5g7Osmz3qR6Ht2zId1wRM8ecRVtLy8MtmdZ7aVeeVhrcn\ntHU4buBJX2v97emhzX29ezufv/u1odu8W0O3I451XBMwxZ9HWFXf/lXR7WhIo1ZXbJ95ohWK\n3pDr9O3PEnXWqwfWdFrwdVXZ99WvnM9flP1mdHNV8AeOawKm+PQIK8+ZHt2MhhSxUOtnc5u9\n43Tc/FyVc3pXlZGhAjfXOJ++ok1gZmRrbDAzerLYNQFT/HqEffVldGtF5OncU7+cddO44ecv\n03pulyXOx30+Z3DXltnt+k1d43r68oLbIxu98l9xXhMwpRF8qt59UPBkWwTPBcQt1ZcIhVVu\nqvuew1VDRx6XyOVA9Rdw3xdIQsr+QVavHdJ10NzquiP3q4YOfZ8T7+VAsQs47wt4k7KQVmar\n5kH13co6I/erhqIP+ngvB7JZwHFfwKOUhTQ0+HLNvgeDZ+/RcV41FH3Qx3s5kM0CjvsCHqXs\nEqH80FU9elnWkOp6IdW/asjuMp94LweyWcBxX8CjlF0iFLwz/OZpNbVeSPWvGrK7zCfey4Fs\nFnDcF/AoZZcIHXdp5N2ZarbzVUN2l/nEezmQzQKO+wIepewSoamBR8OjmnFq2hSnq4bsLvOJ\n93IgmwUc9wU8StklQjuOV5FXCmqmKuV41ZDNZT7xXg5ks4DjvoBHKbtESG8vnlY7WXRC5F7Y\nXTVkc5lPvJcDxS7gvC/gTWo+JVc9v911FLlqKHav+pMtTueKe+Y2BxKQmpB2qjeTGtns5Gnm\nNgcSQEiEBAGEREgQQEiEBAGEREgQQEiEBAGEREgQQEiEBAGEREgQQEiEBAGEREgQ4GdIW99b\nvnZbeOvAksoERjY7eZq5zYGk+BfSE93CPyRx6vMJjmx28jRzmwNJ8i2kx1V2wbjiMQMCakFC\nI5udPM3c5kCyfAvp5MG7wm8/OvH0hEY2O3mauc2BZPkWUtbK2o252QmNbHbyNHObA8nyLaT2\nq2o35nVJaGSzk6eZ2xxIlm8hTbitdqNwWkIjm508zdzmQLJ8C2nrgNGLN3xSvnBIQcVmS9wj\nm508zdzmQLJ8C0nVF/fIZidPM7c5kCzfHkXDR9QT98hmJ08ztzmQLD4dAwIICRDgW0g1Lxb2\n6RmR0MhmJ08ztzmQLN9CmqNU87yIhEY2O3mauc2BZPkW0nGDNyY1stnJ08xtDiTLt5CCZcmN\nbHbyNHObA8ny7yvSquRGNjt5mrnNgWT5FtKM4uRGNjt5mrnNgWT5FtLuwaOXlleEJTSy2cnT\nzG0OJCsVlwglNGrgEp+EZ25zIFm+PZJGjSuKSmhks5OnmdscSBafkgEBPoZUbf23r+y9mkRH\nNjt5mrnNgeT4FlJ18ZVab+qh1KDdCY1sdvI0c5sDyfItpFJ1s9ZDApOLm5UmNLLZydPMbQ4k\ny7eQzrhC688C1jf3E/okNLLZydPMbQ4ky7eQWszT+kn1Z63ntk5oZLOTp5nbHEiWbyG1tB68\no3L3a/1YbkIjm508zdzmQLL8e2o3Wn/e4nJrY9IpCY1sdvI0c5sDyfItpPvUuZ3VCq0XZM1I\naGSzk6eZ2xxIlm8h7R2fk/eI9bZTr8qERjY7eZq5zYFk+Xtlw9L39aoD9Ufr/hAzWv1azOh3\n/zxyYjv797yYI/Wex2NnITWP7bWdA0nwN6SLbosZ3XFBzOiGq2JG/R+IPZnN7A11MGZWpvba\n3ped6m+2cyAJTSWkObEns5m9bhPSKseQ/mo7B5Lgb0gnXxkzGtEx5ue+L+wbs1fnq+u/v/+9\nBdPzflR3srf8D3NnnB8T0j+fvsIhpB2hlxsAGf6G1P/6mNHScwIDF9V/+I/8fsxe0zPvim5+\n8vvSkT0zVefBV70Verfqwz/P/+nob3dSKvuUiydOqHPM16+XXtpedSi8xv6+7FR/SeZ/AbCT\n8qd2WldMbd79/l11BjZP7fTi5tcd2L3qVzec10Y171/00LIdBze/seBn47/bNUNldv/ehHue\nWbml7qXcm56d0i8zo0/xMx863hee2kGQvyF1GWw73nZXXss6755zks0+b2cNCQR6DL/jt/+o\nDr/fUwW6DLrmjv9evsnmpbcJqt3Qe5e7X9q9Q/1fnPcaaJC/IZ3v9A+g/9W1zjuTLrPb5bzi\nei+zFV67z3mde/o0/JNGfEWCoEbw1C7k1yfUecfuqZ3WBVNV3V9VcsVUl3VKz2n4vhASBBES\nIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBAS\nIICQAAGEBAjwN6Th9zp84Pledd6ZMc5ul0t/mvlZnXfH3uqyziOxvxkvxtfBfzS8ExCfuEIq\nqve1wIPdVQ4fOFj399rt+9r+2Hp/OuJrl18ipKu/iuPO8JcoICfZkDZO6pH1rcv4XaVAWJIh\nfdAua+ysMcHg2/YH/O1+4Jf2v3Q9PSUZ0oWB163bl9TVdrtrPanVqTjanaJWeXtsNiluIf2r\nqHPzMx86EApp4/3ds/LvrtF6hPqi4JhX9O0zQztUB3vbHznx4jIc7d5UDs9X0pJLSNu65E15\noFAVhUK6rm/p7Hz1nNbXqNGX3Leudo/P1HD7QwkJhBQ1Wb1q3Q5V662QBlVpvUZdGvr19BdF\n/wbL16+d2XK1/aGEBEKqVdMuP/Sb6Dcu326F9HJokNE/9MXp2dqP5yk1dqPDsYQEQqq1RV0Y\n3SyyvipZ8nqGNt+tHd76w283G+RQEiGBkGp9qAqjm7Wv2kVCqvMC3mu5Z8b+zdYQQgIh1dqj\nBkU3nULSo1W57bGEBEKKat8udEHPB4+ujw3pszMjf5j1CmX/asPEPj/H0e4uQoqYqOZbtyPV\nGpuvSMdlhf4W+T9atLD/x+vSNsCxTq9FpSOXkDZ3zLxxTqG61u6p3csZwZE/HZ+rHvPnXgKN\nnNuVDR+P7RDs8Ytq2++Ryoa3z2hdsNiPuwg0fv7+PBKQpggJEEBIgABCAgQQEiCAkAABhAQI\nICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQI\nICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQI\nICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgIBUhPRVyYaGd6rclOzp\nq9a/uy/ZY4HkpCKkzWqJw0fWDuk6aG51eLMk8Xu27Pxul5TppZ2VajXXw90DEudjSEVRo9RF\nRUV2e6zMVs2D6ruVoe3EQ3o7U7Vqlvt2q/xrr26j/uj13gKJ8DEkVY/dHkODL9fsezB49h6d\nTEjDOq7V2753fO9vrOeF3S72fHeBBPgY0o8z+izdGfJ39cLOnXZ75I8N3S7LGlKdTEjt7rFu\nVqv/CW3f29bTXQUS5Of3SKv7BCbv0i7fIwXvDL95Wk1NJqTMp62brep/Q9tPZiZ9L4Ek+Ppi\nw4H7czovdAnpuEsjb2eq2UmEdOws62aFeji0fduxyd5HIBk+v2r34QVq2KeOIU0NPFoVelsz\nTk2bkvA9G9l2+f73e512/Gdal7e50tv9BBLj+8vfT7VtMcsppB3Hq4LwRs1Uh5cj3GxoaR3U\ntrxr8++dm5nxF093EkiQ//+O9MVI5fjvSNuLp9VuLToh8Xu2btR/jP9ArxsQUD1+l/TdA5KR\nin+Q/cMt5UbPv3ub0dMDsVJ0rV3DFwDtqPByfm9HA4nyM6SELgBK4hIhsaOBRPn4gEvsAiBC\nQlPi4wMusQuACAlNiY8PuIYvAOpXR8eE75m3owEvfHzANXwBULNm2YdkJHzPvB0NeOHjA67h\nC4BKWh5+sS3xJ2fejga88PEB1/AFQFV9+1dFtxNPwdvRgBc+PuDiuACoPGd6dDOJFLwdDXjg\n5wMujguAvvoyurWiNPEFvB0NJC+Fn7mrnt/ueQ9zRwOJSGFIO9WbnvcwdzSQCEICBBASIICQ\nAAGEBAggJEAAIQECCAkQQEiAAEICBBASIMD3kLa+t3xt5NdlHVhSmeQe3s4PyPM5pCe6hf+m\ny6nPe9jD2/kBE/wN6XGVXTCueMyAgFqQ9B7ezg8Y4W9IJw/eFX770YmnJ72Ht/MDRvgbUtbK\n2o252Unv4e38gBH+htR+Ve3GvC5J7+Ht/IAR/oY04bbajcJpSe/h7fyAEf6GtHXA6MUbPilf\nOKSgYrMlqT28nR8wwt+QVIN/2bzhPbydHzDC3wfb8BH1JLWHt/MDRvBZGxBASIAAf0OqebGw\nT8+IpPfwdn7ACH9DmqNU87yIpPfwdn7ACH9DOm7wRs97mDsaSJq/IQXLvO9h7mggaT5/RVrl\nfQ9zRwNJ8zekGcXe9zB3NJA0f0PaPXj00vKKsKT38HZ+wIiUXSKU9B7ezg8Y4e8DbtS4oqik\n9/B2fsAIPnMDAvwOqdr6b1/ZezVe9vB2fsAAf0OqLr5S6009lBq0O+k9vJ0fMMLfkErVzVoP\nCUwubub0x5Ib3sPb+QEj/MZnEckAAA0PSURBVA3pjCu0/ixQpPWEPknv4e38gBH+htRintZP\nqj9rPbd10nt4Oz9ghL8htbQe6KNy92v9WG7Se3g7P2CEz0/tRuvPW1xubUw6Jek9vJ0fMMLf\nkO5T53ZWK7RekDUj6T28nR8wwt+Q9o7PyXvEetupl9Mfimh4D2/nB4xIwZUNi9/Rqw647zLn\n3w3t4eL/3mjw/ICwFIQ05CcN7bFXefm5olHXezgYSAohAQJSENIpwxra4xu1xMP5z/quh4OB\npKQgpIE/bGiPveoND+e/5GoPBwNJSUFIlzT40rS3p3YjGwwVkJaCkPK/39Ae36iXPJz/1H4e\nDgaSkoKQLrq5oT28fUW6ip+Ohe941Q4QQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiA\nAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASICAFIY28p6E9DuSu93D+66d7OBhI\nSlwhFakKwSWrDja4y14v5z9Q7eVoIBleQvqx4hdfAWEeQlqd4RjSX0rQFNzPF28pyYd0oE9v\nx5Amtjsbjd+ZarOHhw7qcgvpX0Wdm5/50IFQSBvv756Vf3eN1iPUFwXHvBL66P2BPzqHdHEZ\nGr9F6lP5R9RRyiWkbV3ypjxQGIqlSF3Xt3R2vnpO62vU6EvuW2d99MOcyTsJqWkjJDkuIU1W\nr1q3Q9V6K6RBVVqvUZdqPUFdFHnR7YJOuwipiSMkOc4h1bTLt57K6Y3Lt1shvRwaZPQPfXF6\nNvzRp9RCTUhNHCHJcQ5pi7owulmkwv9AmtcztPluaPOLtoWakJo6QpLjHNKHqjC6WfuqXSSk\n8ObIFp8QUpNHSHKcQ9qjBkU3Y0L6g7pj8+bNf1ejNn9leywhNQmEJMflxYb27aqs2w8eXR8b\n0i0qqsT20Ik9b0Xj9yNCEuMS0kQ137odqdbEhlS+JOQFddGSDbaHPtgDTcEZ/zbwkDo6uYS0\nuWPmjXMK1bV23yOFOX+PBBxl3K5s+Hhsh2CPX1QTEtCQFPw8EpB+CAkQQEiAAEICBBASIICQ\nAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQ\nAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQ\nAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAADMhbfrSyGlD\nKjdFt2o2/umll5Z9arfTVyUbjN0BwIaZkNQx9+6XPN/aIV0Hza0Ob5bU3uPKWzqosOPv/iZm\n/81qieTyQEMMhdQt47TX5E63Mls1D6rvVoa2a0Pa2l2dNH7W7Nm3j+qselfW7lcUNUpdVFQk\ntz7QEEMhlazuqy54U+p0Q4Mv1+x7MHj2Hn0opKLgi7UfrJ4buCm6bD1SqwMNMxWSrv5le3Xe\ngl0ip8sfG7pdljWk+lBIHScc/vCI/NqNH2f0Wboz5O/qhZ07RZYG4mIsJK33lLZTGWcX3fuI\n59MF7wy/eVpNPRRS8OeHP/yzrOjW6j6ByaF2+R4JPjMYktZfPzMsV+I51nGXRt7OVLOjIXW9\n+vCHL+t2aPPA/TmdFxISfGc0JMv+v77wuOfTTQ08WhV6WzNOTZsSucc3Bebsi3xwz52HV7N8\neIEa9ikhwWemQxKx43hVEN6omRr9ArfzLNXygvE33jDu/ObqO7vr7f1U2xazCAn+MhNS9u2y\n59tePK12a9EJtfd4/4N9MkKvzQXPeaL6iL2/GKkICf5qyi8S7/3nmjUVtv/w+4dbyv2+Mzi6\nmQ9pR0Xs7PB1PrWOuKbnwOYNHzm8ch5zqO3IflnAFPMhRa/psbvO55C6rw7sf+ycrNCTts7j\nyrTzoW5ncxwChvgXks11PrbX9Hw1QOWceU7wxDF9A2qG46F2Z7NdFvCDfyHZXOdje03Pj9U0\n61ndhq4P64+Hq6ecDrU7m+2ygB/MPNz61dGxdgmb63xsr+k5bmj4zW9a7dHV/c5yOtTubLbL\nAn4w83Br1iz7kIzaJWyu87G9pid4d/jNBrVS67tynA61PZvdsoAfzDzcSloefs0s+jC3uc5H\n213T0yly8c9C9brWU9s5HWp7NrtlAT+YebhV9e1fFd2OPqJtrvMJO/KannGB+TVav981d49+\np/XFTofans1uWcAPhh5u5TnTo5vRR7TNdT616l/Ts6mt6nze6QH1mK4O5r7jdKj92WyWBfxg\n6uH21aHf2rCitHbD5jqfWvWv6dk4oqXK/M5ya+sn65wPtT+bzbKAD/z5vF31/Hb3We01PdHZ\nnoOux8Y3chgCBvgT0k4V+2PnHmbxjRyGgAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGE\nBAggJECAyZC2vrd87bbw1oElld5n8Y0choBR5kJ6olv4FzGc+rzMLL6RwxAwzFhIj6vsgnHF\nYwYE1AKJWXwjhyFgmrGQTh4c+Q2PH514usQsvpHDEDDNWEhZK2s35mZLzOIbOQwB04yF1H5V\n7ca8LhKz+EYOQ8A0YyFNuK12o3CaxCy+kcMQMM1YSFsHjF684ZPyhUMKKjZbvM7iGzkMAdOM\nhVT/txErr7P4Rg5DwDRjD7XhI+rxOotv5DAETONzNiCAkAABxkKqebGwT88IiVl8I4chYJqx\nkOYo1TwvQmIW38hhCJhmLKTjBm+UnMU3chgCphkLKVgmOotv5DAETDP3FWmV6Cy+kcMQMM1Y\nSDOKRWfxjRyGgGnGQto9ePTS8oowiVl8I4chYJoflwhJzOIbOQwB04w93EaNK4qSmMU3chgC\npvF5GxBgMKRq6799Ze/VCM3iGzkMAbOMhVRdfKXWm3ooNWi3xCy+kcMQMM1YSKXqZq2HBCYX\nNyuVmMU3chgCphkL6YwrtP4sYH3HP6GPxCy+kcMQMM1YSC3maf2k+rPWc1tLzOIbOQwB04yF\n1NJ6RI/K3a/1Y7kSs/hGDkPANHNP7Ubrz1tcbm1MOkViFt/IYQiYZiyk+9S5ndUKrRdkzZCY\nxTdyGAKmGQtp7/icvEest516VUrM4hs5DAHTDF/Z8PPPVx04YlQ5S8fM9BPrYmYrFsXst+nB\nmNGTf7M529pf2wwBgwyHlL00ZrRSVcXu1/uhmNG04TGj5zrFjAb8p82qD/SL684BYggJEGA4\npGa/jhktsgup3Yk3PlFW76Ke6n6n1X13/99f/s+iU3Oi737z8arF8++ZcvV5ORNizrXliZ4d\nkr/HQDIMhxRcHDNaYRfS/Fsu6xFodsIVsxZWHAwP9v6g9W/CGzUf/9/cqRf3yFAt+111/dSX\n5/6s+PKBJ7dQSrU4aeDlxT+7dV2989SsntUv0GnME/L/J4CbRvLULuTfb82bPChP5Q6Y9Miy\n9QOP//sXbz4188ozc1Rm517nfufb/brlWPW0Oe28q6fcPX/xqo+/iT1B5W8ndQr0m7WaC7/h\nO8MhBf4rZvSi2u92xKbFP7/6tGbBdp/p41X+BT96cJgKdDjj+2OmlT71+3c3ux6pe2de+sQW\nb/cXSE4j+opUx087XKN1wfTQ5mMnV8e72Ojxidw1QFDjDGlWR6uJwTNDm3Pj/1uw1/DT5UgV\nQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAA\nQgIEEBIggJAAAYZD6vyXmNH6NgcbPOyhk36s9ajZoc3nB8W92NTpCdwzQFJcIRWpiiRPb9dM\nwx1pXV1zaL94do+o4RfaIVWSDOkpFXGPgbsEND1JhvRLNaokZLn9AW+UAClx68cei0hSkiHN\nUqvdDpjY8QIgFXKeSjYFb9xC+ldR5+ZnPnQgFNLG+7tn5d9tfQ8yQn1RcMwr+ib375omXlwG\npELn/xYuJE4uIW3rkjflgUJVFArpur6ls/PVc1pfo0Zfct86PU5tr9683fFQQkKKNMKQJqtX\nrduhar0V0qAqrdeoS7WeoC4KvYw2XP20jVInP+twKCEhRRpfSDXt8kMvJ29cvt0K6eXQIKN/\n6ItTOJ7zVY/Sp2e2UvPsjyUkpEjjC2mLujC6WWR9VbLk9QxtvhvaXLZwj3X79+y29n9ohZCQ\nIo0vpA9VYXSz9lW7SEh1X2W4XL1jeywhIUUaX0h71KGLcxxDul7Z/0MSISFFGl9Iun270NWl\nHzy6Pjak3Y8/F95lkNpoe+jEnrcCqdC68YU0Uc23bkeqNbEhHezSYoP15neqr/2hj/cAUuLk\nt8QbiYtLSJs7Zt44p1Bda/fU7pVAbtEdlwdarfHnXgKNnNuVDR+P7RDs8Ytq2++R3r6kdWbn\na5O9KBxIM4Z/Hgk4OhASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAA\nIQECCAkQQEiAAEICBBASIOD/A/MeG3/cXO4PAAAAAElFTkSuQmCC"
          },
          "metadata": {
            "image/png": {
              "width": 420,
              "height": 420
            }
          }
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "plot without title"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAADAFBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJyco\nKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6\nOjo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tM\nTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1e\nXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29w\ncHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGC\ngoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OU\nlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWm\npqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4\nuLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnK\nysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc\n3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u\n7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////i\nsF19AAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3dCXwTZfoH8Ddt09IbqOUu5VIR\n5EYWpLIo1SIUBQS5KVJQqXIoIIIHikr5g4oKZRHWAzxwkUNAVzxAXFAqlyJY3C2IyA1ySYGe\nef8zk6RNm5nJ9byTEn7fz2eTt0/emXdY55eZTN4kjAOAz5i/NwAgECBIAAQQJAACCBIAAQQJ\ngACCBEAAQQIggCABEECQAAggSAAEECQAAggSAAEECYAAggRAAEECIIAgARBAkAAIIEgABBAk\nAAIIEgABBAmAAIIEQABBAiCAIAEQQJAACCBIAAQQJAACCBIAAQQJgACCBEAAQQIggCABEECQ\nAAggSAAEECQAAggSAAEECYAAggRAAEECIIAgARBAkAAIIEgABBAkAAIIEgABBAmAAIIEQABB\nAiCAIAEQQJAACCBIAAQQJAACCBIAAQQJgACCBEAAQQIggCABEECQAAggSAAErr0gFedsv+Lv\nbYCAcw0F6bv+rXrv5Lk3Mxad5e9tgUBz7QQp28zMLOZA58ghfaPYWn9vDQSYaydIqeZVxUda\nDA3ezPl/I5P9vTUQYK6dIMUNlW42sC5ye0Q1P28MBJprJ0jm6dJNHntYbk8L8e+2QMC5doLU\ncLh8G/ukfDugpn+3BQLOtROk9LDN9uZW833+3BIIQNdOkHKrmaZaW0PNIdv8uy0QcK6dIPGc\n5KetjRYJa/y7JRB4rqEglTrq7w2AwCM6SIV7d+TbmgfPCB7LfWcP+nsLIMAIC9KGrg3uzubr\n6zAWY5uQw6q8WCBqNHfs7pGYlFWsNKdciwdiEEnUHvV9CIsJivw+JmH4/dXY59ahGgTf9I2g\n4dywJYxFmNnfz8ptBAmIidqjetXazU/dXr/VZek8qkF361BTtrdh3Ta7WFCYnubVlvxXzbfk\ncQQJyInao+JekG62s3fl9ovVrUNN4cVz41mXJecFjakvQZ4ixDeE9ihGkICcqD0qZKl0c4x9\nJrffsk7IkYLEeV5mHAu+Jf3FNwSNq838rHK3lI1DkICcqD2q5nTpZhN7XW5Ps07IUYLE+aX3\nekUyZvyeXO8e6/1UNhtBAmqi9qiB1TcW/NzipvpHOM+p1s861BT7gwU/frRA0LjaxpnmFcr3\nljQ2YSyCBLRE7VH7oqWjTvWcxIjbO4UE/2AdaoqLZcT6sz6zfgrJMs4fB0QIbML2qD2D/jbi\nV76ng4k1+sRaCXta1FjuOZ0xwdZa2RhBAlrC96iLp5S7wmWnKz6iUjKSn4eHwGLUU/M55vQG\nkkrJSH4eHgILggRAAEECIIAgARBAkAAIIEgABBAkAAIIEgABBAmAAIIEQABBAiAgNkjHdm3c\nbZ1rV7TurGbJSH4eHgKVyCAtasBkTZfplozk5+EhcAkM0gIWlpyWMaSDiS3RKRnJz8NDABMY\npBtSrN9y8luTZjolI/l5eAhgAoMUusXWyArTKRnJz8NDABMYpPittsbCujolI/l5eAhgAoM0\ncpqtkTpBp2QkPw8PAUxgkI51GLx236GcFT2Scw9LNEpG8vPwEMAEBomVp1Eykp+HhwAmcG/q\nPaAcjZKR/Dw8BDA8LQMQQJAACAgMkmV5auvmVjolI/l5eAhgAoM0h7GIWCudkpH8PDwEMIFB\nqpdywI2Skfw8PAQwgUEyZ7tTMpKfh4cAJvKItNWdkpH8PDwEMIFBmpzhTslIfh4eApjAIF1M\nGbw+J1ehUzKSn4eHAGbMFCGdkpEwPQhEEbhHDUpLt9MpGcnPw0MAw1MzAAGhQSqW/pefvcui\nXzKSn4eHgCUwSMUZ/Tg/2IixpIs6JSP5eXgIYAKDlMke57yHaUxGUKZOyUh+Hh4CmMAg3dyX\n8yMm6WX9yNY6JSP5eXgIYAKDFLWQ87fY15xnVdUpGcnPw0MAExikaGm3HRRZwPn8SJ2Skfw8\nPAQwkad2g/mJqD5SY/SNOiUj+Xl4CGACgzSTdarDNnG+JHSyTslIfh4eApjAIF0ZER77hnRf\nu8VZnZKR/Dw8BDDRMxuK0s5uLSpfevdfvGLJSOeGF/pzeAhMooP0J5vwxbHypS49BI+pbxs7\n4dfxISAJPyK1aRXGruuW8r+yUqtOgsfUt5kd9+v4EJAMmLRatHfZk+aFZX937CZ+TB1b2THX\nnQA8Y9Ds79hPytqDHzRmTA172Gm/jg8BCUECIIAgARBAkAAIuBWkdObrl4UgSBDYvA5S4ZNB\n7dwfBkGCwOZtkHLaRiNIAHZeBulCePvcMAQJwEYvSMfT60S0fK1IDtKBWQ1DE2ZYOB/ATiZX\nWcPPTCzkOkE6t6OCJ74say9YVPFRQ303dptfx4erxU5PpmTqBOlU3dixL6eydDlID7TJnJ3A\nPuR8GBt898w9yuM6QcpgAFe9j2mCNIZ9Id32ZHulICUVcr6T3cP5SHZXie1xnSCN6p4NcJWL\nf58kSJa4BPnr3w5sPC0FabVcCG4vH5w+sHdAkCCgEQXpKLvT3kyXjkqS2OZyc4c7QWo0DOAq\nF0ETpP0stSxIuWVBKr2ApxOkfyUDXO26/0ISpDyW5HWQAK4xOhcb4uMKpdtf5+1FkABc0AnS\nKLZYuh3IdiJIAC7oBOlwrZBH56Sy4WqndpumTJkSXEu6+dOYzQSo3PRmNvw+tIa50SvFakHK\ntL9nhd+QBOD4oTEAEggSAAEECYAAggRAAEECIIAgARBAkAAIIEgABBAkAAIIEgABBAmAAIIE\nQABBAiCAIAEQQJAACCBIAAQQJAACCBIAAQQJgACCBEAAQQIggCABEECQAAggSAAEECQAAggS\nAAEECYAAggRAAEECIIAgARBAkAAIIEgABBAkAAIIEgABBAmAAIIEQABBAiCAIAEQQJAACCBI\nAAQQJAACCBIAAQQJgACCBEAAQQIggCABEECQAAggSAAEECQAAggSAAEECYAAggRAAEECIIAg\nARBAkAAIIEgABBAkAAIIEgABBAmAAIIEQABBAiCAIAEQQJAACCBIAAQQJAACCBIAAQQJgIAh\nQSrYtvE3x7+LDu/77Tzh+i0Hvlq1asMfhGsE8IzYIL2wUb5dWI0x1u5HW61gfsdQ6W9WJy2b\nZpCzE2swRf0Zl2nWCOApsUFiU6SbT1lYn4c6s9j9SulCBxbesqO5yZA2JjaZYoxjDdn1I6bP\nnv30oDqs1VmKNQJ4zIAgXR+bI92uND2glB5jE6Szun2Jr/Pfe7N3CMZINy+3tYqzTOMJVgjg\nOfFBOsWmKe3edZW7ej2Vu/dj8nhxu7YEY9QaWdYekECwQgDPiQ/SH+w9pf20Wbkzz1Du9rEt\nnD8fTjCG+aWy9nOhBCsE8Jz4IBXHZirtkdWVu9r3K3cr2Lecj4sjGCPx/rL2vQ0IVgjgOcFB\nGrQ99/TUJpek5r7IXkopzbTYwvnPiZF5fFvV7gRjjDfNybe28p5VXpMBGE9wkKxWcP5BZNA2\npXSwOqvTpZmJzefF5shtBGOca8uiu4149JG0rhHstosEKwTwnNggvTN3+vi03l03cJ5Vd52t\ndmBANAu5TX6D6Yk9JIMUvNo6WI6rueOiYpIVAnjMqClCF0sc/sgr0eznnSv/27kzt4B4pQDu\nM26u3Z+57pQoxyj++TtMGwJjGBekKc5DqZRoxvjuEenmvZrS+V6rb4mHAFATkEH6JjTKwj9m\nUf0z7gwK20E8BoCKgAxS1xrSKV7DxGNSMzu8F/EYACrEBqmdg1pMq0Q+Rswkzs+z15X26KoE\nYwC4IDZIQUFhpYKZVol8jMhnOM83rVTaz1chGAPABbFBmhJddhnNdtqlUiIfo/P1lzi/dZLc\nzG/VimAMABfEBqmwTftCe9u2k6uUyMdYx9p+UbSz9pJLhdl3sDcJxgBwQfDFhpzwSfamPTUq\nJfIxFkey8GaJLDiYmR63UIwBoE/0VbsLZ+ytTZnaJfIxTsxJSYwOi2s3bifJCAAuGPstQoXL\nTrtRqvxjAFRgbJDOsc1ulCr/GAAVIEgABBAkAAIIEgABBAmAAIIEQABBAiCAIAEQQJAACCBI\nAAQQJAACRgTp2K6Nu08praJ1ZzVLlX8MAE3ig7SogfJtq02X6ZYq/xgAOoQHaQELS07LGNLB\nxJbolCr/GAB6hAfphhTrr8X+1qSZTqnyjwGgR3iQQrfYGllhOqXKPwaAHuFBit9qayysq1Oq\n/GMA6BEepJHTbI3UCTqlyj8GgB7hQTrWYfDafYdyVvRIzj0s0ShV/jEA9AgPEitPo1T5xwDQ\nI3wX6z2gHI1S5R8DQA+eqwEIIEgABIQHybI8tXVzK51S5R8DQI/wIM1hLCLWSqdU+ccA0CM8\nSPVSDrhRqvxjAOgRHiRztjulyj8GgB7xR6St7pQq/xgAeoQHaXKGO6XKPwaAHuFBupgyeH1O\nrkKnVPnHANBj5BQhnVLlHwNAj/DdbFBaup1OqfKPAaAHz9cABAwIUrH0v/zsXRb9UuUfA0Cb\n8CAVZ/Tj/GAjxpIu6pQq/xgAeoQHKZM9znkP05iMoEydUuUfA0CP8CDd3JfzIybpFf/I1jql\nyj8GgB7hQYpayPlb7GvOs6rqlCr/GAB6hAcpWtqjB0UWcD4/UqdU+ccA0CP+1G4wPxHVR2qM\nvlGnVPnHANAjPEgzWac6bBPnS0In65Qq/xgAeoQH6cqI8Ng3pPvaLc7qlCr/GAB6DJrZ0GHz\n1qLylW2teMWSjzr+x2mFN+wlHgNAnUFBCku8/b4Hp85565P/7D1eoFSyQsjHqJ/Uc/CYJ2e9\n+dHn3/9yJE8umd6nHgRAlUFBennWMxkD72rXIEaejh2VmDiczycP0qtzXpg4qn/yLTfUrCKP\nEhLXqDFDkMAYRk9aLTyRs3nNO3/rzj+qInCUKyf+u+2r5Yu7ss8EDgJQxj+zv0cN5euixA/z\nBNsofhAAjiABkECQAAi4FaR0RvzFBwgSBBhvg3R2Yv3QBvd6+51XCBIEGC+DdKYB6/nMkJAq\nP3s3KoIEAcbLID3C5km3K1kP70ZFkCDA6AXpeHqdiJavFclBOjCrYWjCDAvnA9jJ5Cpr+IRu\nhVIHS3iid6M+lMY/N+DL7aea/iN+EACuG6RTdWPHvpzK0uUgPdAmc3YC+5DzYWzw3TP32Hrk\nmzurL7prlr5pT856aayLPgSeHjZT/CCBCBNCPKYTpDHsC+m2J9srBSlJOgDtZPdwPpLdVVLa\n43XlBE/FqJimcPWqhw9Dekw7SJa4BPmrrA5sPC0FabVcCG4vH5w+KO2xKTRJY271qO7ZcPV6\nNYJ2J7sWaAfpKLvT3kyXjkqS2OZyc4e9+mFY2zMayyJIVzUEyXPaQdrPUu1N21U7a5BsF/As\nz7Luf2kti1O7qxpO7TynHaQ8lmRvqgTJMpKNLdZc9kd/v1oGn3yg+V8WNOhcbIiPky9x/zpv\nr1qQxrOZRmwewNVBJ0ij2GLpdiDbqRKklWy8IZsHcHXQCdLhWiGPzkllw9VO7RqzsVMU+GoR\nAK4/s+H3oTXMjV4pVgtS6a94HTRiIwEqO/w+EgABBAmAAIIEQABBAiCAIAEQQJAACCBIAAQQ\nJAACCBIAAQQJgACCBEAAQQIggCABEECQAAggSAAEECQAAggSAAEECYAAggRAAEECIIAgARBA\nkAAIIEgABBAkAAIIEgABBAmAAIIEQABBAiCAIAEQQJAACCBIAAQQJAACCBIAAQQJgACCBEAA\nQQIggCABEECQAAggSAAEECQAAggSAAEECYAAggRAAEECIIAgARBAkAAIIEgABBAkAAIIEgAB\nBAmAAIIEQABBAiCAIAEQQJAACCBIAAQQJAACCBIAAQQJgACCBEAAQQIggCABEECQAAggSAAE\nECQAAggSAAEECYAAggRAwPgg/Znr+FfBto2/uVyk+Ofv/nBr3cU526+4HhOAnPFBmmId8oWN\n8u3Caoyxdj9qdv7uEenmvZpSp1bf6q30u/6teu/kuTczFp2lPSaAKH4LEpsi3XzKwvo81JnF\n7tfo+01olIV/zKL6Z9wZFLZDe53ZZmZmMQc6Rw7pG8XWao4JIIp/g3R9bI50u9L0gEbfrjWk\nc7KGicekZnZ4L+11pppXFR9pMTR4M+f/jUzWHBNAFGP2sHYOapUF6RSbprR719VYLmYS5+fZ\n60p7dFXt9ccNlW42sC5ye0Q1zTEBRDFmDwsKCisVXBakP9h7Svtps8Zykc9wnm9aqbSfr6K9\nfvN06SaPPSy3p4VojgkgijF72JTosstmDqd2xbGZSntkdY3lOl9/ifNbJ8nN/FattNffcLh8\nG/ukfDugpuaYAKIYs4cVtmlfaG/bgzRoe+7pqU2koPB9kVovf9axtl8U7ay95FJh9h3sTe31\np4dttje3mu/THBNAFIP2sJzwSfamPUhWKzj/IDJom9ZyiyNZeLNEFhzMTI9btFefW8001doa\nag6xr8x5TABRjNrDLpyxtzZZT+femTt9fFrvrhs4z6q7Tnu5E3NSEqPD4tqN26m7+pzkp62N\nFglrtMcEEKUSPFVfLCFc2VHCdQG4zd9ThBRnDzr+pTFrqOJynkwHKj+Afl8AL/jtDVm+u0di\nUlaxY0l/1lDp6xx3pwM5D6DdF8A3fgvSljAWYWZ/P+tQ0p81ZN/p3Z0OpDKAZl8AH/ktSD3N\nqy35r5pvyeNuzhqy7/TuTgdSGUCzL4CP/DZFKEGe1cM3hPYoLhek8rOG1Kb5uDsdSGUAzb4A\nPvLbFCHzs8rdUjauXJDKzxpSm+bj7nQglQE0+wL4yG9ThOrdY/1zKputPWtIbZqPu9OBVAbQ\n7AvgI79NERpnmqeULGlswlitWUNq03zcnQ6kMoBmXwAf+W2K0J/1mfVKgWUcY5qzhlSm+bg7\nHUhlAM2+AD7y2xQhfjpjgq2ysrF1K9RmDalM83F3OpDzANp9AXzjn6fkwmWndUvWWUPOvcpX\njmqty+2aXh3AA/4J0jm22auSSiefanp1AA8gSAgSEECQECQggCAhSEAAQUKQgACChCABAQQJ\nQQICCBKCBAQQJAQJCCBICBIQQJAQJCBgZJCO7dq4+5TSKlp31oOSSiefanp1AK8YF6RFDZQP\nSTRd5mFJpZNPNb06gJcMC9ICFpacljGkg4kt8aik0smnml4dwFuGBemGlPPK/W9NmnlUUunk\nU02vDuAtw4IUusXWyArzqKTSyaeaXh3AW4YFKX6rrbGwrkcllU4+1fTqAN4yLEgjp9kaqRM8\nKql08qmmVwfwlmFBOtZh8Np9h3JW9EjOPSxxu6TSyaeaXh3AW4YFiZXndkmlk081vTqAtwzb\ni3oPKMftkkonn2p6dQBv4ekYgACCBEDAsCBZlqe2bm7lUUmlk081vTqAtwwL0hzGImKtPCqp\ndPKpplcH8JZhQaqXcsCrkkonn2p6dQBvGRYkc7Z3JZVOPtX06gDeMu6ItNW7kkonn2p6dQBv\nGRakyRnelVQ6+VTTqwN4y7AgXUwZvD4nV+FRSaWTTzW9OoC3/DFFyKOSiyk+Htf06gDeMmxP\nGpSWbudRSaWTTzW9OoC38JQMQMDAIBVL/8vP3mXxtKTSyaeaXh3AO4YFqTijH+cHGzGWdNGj\nkkonn2p6dQBvGRakTPY45z1MYzKCMj0qqXTyqaZXB/CWYUG6uS/nR0zSi/uRrT0qqXTyqaZX\nB/CWYUGKWsj5W+xrzrOqelRS6eRTTa8O4C3DghQt7byDIgs4nx/pUUmlk081vTqAt4w7tRvM\nT0T1kRqjb/SopNLJp5peHcBbhgVpJutUh23ifEnoZI9KKp18qunVAbxlWJCujAiPfUO6r93i\nrEcllU4+1fTqAN4ydmbDyYt8a1H5Ut4Jp9L5P51KR69UrKjWLL85Lcm5Wk2vDuA5Y4N097NO\npefvdCqNH+hU6jjXqVRwy2sVS4Vfmp3nKnwfXKi6LXlBv6hvJIDnjA1SHefU3MMa3PPUv/YV\nO5T+dr1Tr7jYx3bywhM5m9e888pTD9/frXX9KMb67Ni49sM3/++Zx0b3757UulFcGGMmW5Dy\ncr9bveC5R/p0vkHqV6C6LafZlxT/JACZwUFKcSpd2bR4XNfqLMah1NE5SIfm3cp6mRhjUfXb\ndLt/zFOvvJOofAoips4N7brdO+TBKS/MXbx8/Zaf9lu7j5Yeqtq0y/1jZyxeu1XjM0cIEhAy\nNkhdJ2k8MCfR4Y9R96h1SRrD/n287ODSc+CvR85rjTOjxeF8V9tyjv3oqguAu4wN0l3TNB74\nZ2OHPx7pr9YleRxzPLb0HaczTmZH19uCIAEhBAmAAIIEQMCtIKUzoi8JQZAgQHkbpAOjG4Ve\nd+8PHo6GIEGA8jJIv8aFDp0+xGz+3rPRECQIUF4G6U7Tt9LtKna/Z6MhSBCg9IJ0PL1ORMvX\niuQgHZjVMDRhhoXzAexkcpU1/OmpcodicyvPRrt/tsYDK9s5/PHUaLUu/WZEHnf4c6TzbKMy\nC7u73pbLUfgmfSCjE6RTdWPHvpzK0uUgPdAmc3YC+5DzYWzw3TP32HocYb3VF/0oWV03jXqF\nR1S7dXOjj5uPut8HfNDzJPnuWnnpBGkM+0K67cn2SkFKKuR8J7tHOhCwu0psj1/6pmX0dvVF\nRzUaBte6IczDV9BXNe0gWeIS5BmgBzaeloK0Wi4Et5cPTh/YHo9lbKjWydGo7tlwrduMICmO\nstKp2unSUUkS21xu7rAVn3zw1qAkjSSN6vYVXOs+Q5AU+1mqvWm7amcNksOls28iW5Y4LyjJ\nYABsJ9VeehXQDlIeS7I3tYLEB7Mc1WXP7wDYTbSPXhV0LjbEx8kfLf113l7nIB1pOUzp0pdp\nXG0AuLboBGkUWyzdDpQO0M5HpHqh8s+w/jcq6ooRGwlQ2ekE6XCtkEfnpLLhaqd2q4PNA58a\nEcnmG7OVAJWc3syG34fWMDd6pVj1NVJ27/jgqslrjdhEgMoPPzQGQABBAiCAIAEQQJAACCBI\nAAQQJAACCBIAAQQJgACCBEAAQQIggCABEECQAAggSAAEECQAAggSAAEECYAAggRAAEECIIAg\nARBAkAAIIEgABBAkAAIIEgABBAmAAIIEQABBAiCAIAEQQJAACCBIAAQQJAACCBIAAQQJgACC\nBEAAQQIggCABEECQAAggSAAEECQAAggSAAEECYAAggRAAEECIIAgARBAkAAIIEgABBAkAAII\nEgABBAmAAIIEQABBAiCAIAEQQJAACCBIAAQQJAACCBIAAQQJgACCBEAAQQIggCABEECQAAgg\nSAAEECQAAggSAAEECYAAggRAAEECIIAgARDwR5AuTNnnutPZg96uvnDvjnxvlwXwjj+CdJit\n03hkd4/EpKxipTnF8y3b0LXB3dl8fR3GYrJ82DwAzxkYpHS7Qeyu9HS1HlvCWISZ/f2s3PY8\nSN+HsJigyO9jEobfX4197uvWAnjCwCCxctR69DSvtuS/ar4lj3sTpF61dvNTt9dvdVk6L2zQ\n3efNBfCAgUF6LLj1+nOyX9hH586p9UgYKt9uCO1R7E2Q4l6Qbrazd+X2i9V92lQADxn5Gml7\na9OY81znNZL5WeVuKRvnTZBClko3x9hncvutEK+3EsALhl5sKJoVXmeFTpDq3WO9n8pmexGk\nmtOlm03sdbk9raa32wjgDYOv2u3vxnr9oRmkcaZ5hfK9JY1NGOvxlg2svrHg5xY31T/CeU61\nfr5tJ4BnDL/8/U71qOlaQfqzPktWGpZxGpcj9OyLlhaqnpMYcXunkOAffNpIAA8Z/z7SyYFM\n832k0xkTbK2VjT3fsj2D/jbiV76ng4k1+sTrzQPwhj/ekP33xByh6794SujqAZz5aa6d6wlA\nf+b6sn7flgbwlJFB8mgCkBdThMiWBvCUgTucZxOAECS4mhi4w3k2AQhBgquJgTuc6wlA7RzU\n8njLfFsawBcG7nCuJwAFBYWVCvZ4y3xbGsAXBu5wricATYkuu9jm+cmZb0sD+MLAHc71BKDC\nNu0L7W3Po+Db0gC+MHCHc2MCUE74JHvTiyj4tjSAD4zc4dyYAHThjL21KdPzAXxbGsB7fnzm\nLlx22uce4pYG8IQfg3SObfa5h7ilATyBIAEQQJAACCBIAAQQJAACCBIAAQQJgACCBEAAQQIg\ngCABEECQAAgYHqRjuzbutn5dVtG6s1728G39APQMDtKiBspvujRd5kMP39YPIIKxQVrAwpLT\nMoZ0MLElXvfwbf0AQhgbpBtSziv3vzVp5nUP39YPIISxQQrdYmtkhXndw7f1AwhhbJDit9oa\nC+t63cO39QMIYWyQRk6zNVIneN3Dt/UDCGFskI51GLx236GcFT2Scw9LvOrh2/oBhDA2SMzl\nL5u77uHb+gGEMHZn6z2gHK96+LZ+ACHwrA1AAEECIGBskCzLU1s3t/K6h2/rBxDC2CDNYSwi\n1srrHr6tH0AIY4NUL+WAzz3ELQ3gNWODZM72vYe4pQG8ZvARaavvPcQtDeA1Y4M0OcP3HuKW\nBvCasUG6mDJ4fU6uwusevq0fQAi/TRHyuodv6wcQwtgdblBaup3XPXxbP4AQeOYGIGB0kIql\n/+Vn77L40sO39QMIYGyQijP6cX6wEWNJF73u4dv6AYQwNkiZ7HHOe5jGZARp/Viy6x6+rR9A\nCGODdHNfzo+Y0jkf2drrHr6tH0AIY4MUtZDzt9jXnGdV9bqHb+sHEMLYIEVLO/qgyALO50d6\n3cO39QMIYfCp3WB+IqqP1Bh9o9c9fFs/gBDGBmkm61SHbeJ8Sehkr3v4tn4AIYwN0pUR4bFv\nSPe1W2j9UITrHr6tH0AIP8xs+DaHby3S77LksqseOrJ/dLl+AGJ+CFKXMa565LPvfFh/6mAf\nFgbwih+C1OBOVz0us098WH/zv/mwMIBX/BCkRDeCtMqH9Tfr4MPCAF7xQ5B6POGqxxXmy0fG\nBz3kw8IAXkGQAAggSAAEECQAAm4FKZ1RfpcIggSBx5cgPca8+mYEBAkCjw9B2h6MIAFYeR+k\notatECQAK70gHU+vE9HytQOnClAAABvASURBVCI5SAdmNQxNmGHhfAA7mVxljfzoLNPn3gXp\npY9c9SjpccSbFdu8/k8fFgbwik6QTtWNHftyqhyWdPZAm8zZCexDzoexwXfP3CM9uj98zDnN\nID3P4GpQ7ZyIfeqapBOkMewL6bYn2ysFKamQ853sHs5HsrtKlEe71T6vHaRRt8yDyu9Z9gf5\nDnWt0g6SJS5B/na4AxtPS0FaLReC28sHpw+UR99hK7hOkLpng9cmfGXQQCsRJDLaQTrKSieX\npktHJUlsc7m5Q26erJ7K9YJkOyK9mvqaf59zr0az2TCDRsIRiY52kPazVHvTdtXOGiSlOTDq\nkF6QXqhmFc2iq2mrGqvz4LUrhoUZNVQdvEaioh2kPJZkbzoF6d/smcOHD//CBh2+oLv2t9kc\nnUefrPuXZxt7NeryrqdLrGP4Jemrj87Fhvi4Qun213l7nYM0sfS6zxTdtU9ifXQevYtN9HyD\nrzaxLj8PXFFWSNAZEVsCIukEaRRbLN0OZDudg5SzTvYRu2vdPt21p9aoWb7wab28sj/i7zfv\n9XKzrxpnWFeNR/7ao/HAk13D1/k67Hm8l2YwnSAdrhXy6JxUNlztNZJC+zWSXaNJbL/j3zuj\nWNnPJR9mOT26BtLPRnyZ71zbwWpo9B5XR+OBQaP+/qRz9eODnmzKIJbjSXfwmd7Mht+H1jA3\neqXY+yBdDvq2xlKltfcn+fZQ7WEJi0sfXRtRvL/KB95ueOVz3LTAufgxY6dVe58IL/8cU6bz\n808lORX3m0d7sCkfhNZ9wYPu4Duhn0f6iZ3u/bDSah80+Dd+/uauBT3Hlz76fEfOn62tf7WC\nUOFRwQMsZa2ci7NvCvqPau8nmtbQuAyR8O5nVZyObUNiY6+4vSWHqs6cjl8RMJbQIH0Yz2e3\nlBu5bNGtoeO6NTvLn7y99NE+0svwy40ec7GO4vdnO/z1xWaPNuBbhzPHZ2J+Kfvjow0ercct\nQzsFfe9UHHNf44Vqnc9Ev9t3lOpqioI3ngvaUqH4U9D66H8prYvZzotUUHL7bcU/qxzvzt70\nX5fLgpeEBumZLnxL0Hmp8dJNnK++qdZB6aTjutJHGyySbtaF7NJbQ8HixhEmh//8TeI8eQvx\nRYdZ5KeimzT+0/7HnrDa5FfeLTXf7p7mVO0+KXW8c1/OpzcseqWp6noOShFoOatCsWdPPqKn\n0no8wuWWz4k5yPkNs53qSxnmxQsjNEj9HuJXwuQJey2fk26K5bO4n9kx24NnTNuVPu2LtVfw\nWb2qz/zZYVzp39tMbZLKf4nqvP79+4+YU6i69AfBcWU75MSm51vdYetX2LZXI/Ir77tMR1aH\nO31T8o1ZT6h9+9hf1RfyH0yn1NbzbVA+H3NP+drmoJ/4xpATUut4eNASFxvyc5h8zjhV+XY/\nS+qzZQ/0qR/xp8Yy4CuhQWr+GuedpnOew0qvkheEfmFrbTArJ/3Hqr6ivYJu90nhey+m9Dn4\n8c6H46Y5dtgVPPTBB0fHt/5RZeFvw+aO7WH/42j4R/z3Grb3dKbHn1xDfuV91s28qO5r9r++\ntkbBUuWzd+uqda6bz4siV6ut5706nL8fV/5qZtJg6YSt/lypNb55+l0uNmRKZ/l2h0k+di8P\nC51rr1+K+LjBS+78S8ALIoNUJIdmovSEPN3hVXiLl22Nl1ta79+MPGh/LKvCR/5KYpdLt/k1\n59n/rjePrwv+0qFDx77y3cn+5ulOB6V91cfxj2Psh7tHWpVwviVMua62w7xCOlm6veICPrrj\ncc6fvckWgB9j2HH5/ijL+YGdd+p7uaa8e3ebpLaelzpy/jv7xbH0qVl+wTOtjbS+Ksu/CT6u\nvyF9HpFvLQ2kUF9p+MSy4KW2+uqIS6/WKfDknwTuExmkX+U5kSulfbnpzLLikDR7Y7j13tKl\nu/2xu6Mvl1tBDvtdvnumqW33/E+wdHbzeM2yHWlxpO0l0/L4v5WUH/xko3uL+UnTDutfv4cq\nH0Z8O/juT4rzm8tfDp5bxeXnC/lro3VOOyvIC1vP+R/Bm5Q/DtYeqJzS8i2my3+ZnD/u+3r8\nJel2eke1FT10v3STsMihUtJKOZLuY3v42JYlJQmvqS1W5mbr4xNv4zyzxgU+3/yptT68D78Q\ns1RnQfCByCCtjpYCcJz99JPjFaTMtrZGM/s5x76w922t2uxj5d6ywpqKd2spd0fNttPBjGTp\npuCWLva4/Xld6Wugg6Zt5cYuSW4n76vNbOeNIztYs/j9sCr1kusoM3CernuRX/5ulc7vVpyP\nNY8o0X64vE/D5a26Z5CyXU2TC1orswzfqy2F4q2KfS1NZsh3X4ZeUlnR3fJxamCaQ2VhtPV1\nZYfJh6us5PyJW1Q3oCTT+gRTEv5v5f67oOMnYt6UGtMjlJ8kKKwuheixNu7+e8AzIoM0U/kv\n3mjBVMf/8p9WsT7LX7I9eUtmxFtfo59gLfspjS+Z9fQtw/aae6B1HnpRDWXiy6HEFNvbLKNv\nKjtTae/wqloyvdpv8t3D1jX8L8T+yoyfmXvLeusGJHZuHWKq0lb7quEL9b+v+pC7Uy/GKS9d\nPgs7xQtOdGp9gQ9TDrgzbuU8xekM7guzkoy/QjaqrKi5/ANP8xqXFU7H2c6G59V5qLVFfndO\n9Sr2JGYN7CH2P+W+pO6CUS2V/68frCdf5vnKLP2/fDD4Gzf/PeAZkUGy7ktDhzR2nAJ+yHbh\nIdtU+tKhoKr1LZL1YSvClesKvZj1xVK7F60dtgQph7QvQq2B218vVQlQdpDDjvhcW+7gy2Dr\n71ksq64cUtK6qGzeV/0yvz5/vH/I5EvF2S92aycpn8WL183nW6PHqSxZ3gVl425U9vaSxEjG\nWEMpJ3OUd0QfGCIdBXpUXKKP9fmCt5uhsrpoect/LL24KT1dNLe9/jsdalIuT9w8XWWxN80N\nHlcaX5ttB9lHbwj+WmlcajRWun1EPpzzfvIzSz5+PoqcyCC1z5RvF0SYDjlWqy5X7v7h8Jyb\nYn1Xdlbb/KryWd6BoBTlDOSK2X5doa2yjzxg/4DUf2v3KeRH3mrq+ENIO0wOX5hyJN52FDjK\n5LlJJ8N0ZoGuqVe3qqn1pDdnzXrMXG5D59S6wvm3EQ/9oL/b/dQ46j/y5QHrHNSctV9l75Sf\nDb4Ik3f/25/mfFGDCkscCbHu4HxCivPqzjD5EFkc+7G9kB30jb3Zu61yfMy83nmxL82LJlpf\na/7jBlvpG2a/iP558A/cUjdLbm4Jmtz3+uB+uv8i8ILAIFmilKPCT+zWcuXbnlHuHnT4j/mc\n9VX3wJH8gV7S/cQ23wfJb7F8b7J/7uydqtJpvjVlipwaSS1YzbSTjqPVebO0XZTU2X4R73r5\nROmFxnovdS7M+tC6Iksrx2kWl2spx5gNN5uiUmbZNuSS8zSd9yPuz4jazBfVrXAOeEz5VHGD\nf8r7bl75h5670dZXuah4LrVduzaNp9ova/zIlPd67h1m+7u4XdnTxSnrYep30w9yxx1HyxL+\nS+xk/naC0ny8p33JQaUvTQe1Lsq2PdHcd9vDWQuCf1P9fwK8JzBIh6zn8sWxr5crZ/RW7m5x\neEtjfZjyoqfpG3x96DmeV+3tothl0t+vlf40ef6QoMGHP4m4WLrEnt4zd1XYc0f3Km2+GH/Y\n3ky/T3qZXVfnrSpH70U7fGJ03nW24U6vmtAofqG0o+fNui65wqCF40Jetlgejd5y34iKK4v/\nUHo4ZAPnZ9mOcg8U1bNvznHp6JN3683z3nxzbvXutqHXRCl3q8Jtf/8j5hiv6LZx+UvbS6eQ\nQbWeshZONOxbYj9ZTp3g/A87UW3OFMdLhK1VL7yDD8QE6bmBXI6H9Rnzl/IzMBco53RFVT4r\nK51TJqnlBf+HF133Nl943RXee6RUGDy8rMt37SNv7K875poI+8W84nqlb0LypTUs/F+Rbn6i\nurB+Zmm7oP6LZQ8UzI5p/eWc+NovhJefaFqQUuMb6c7ySHT0soor6zZVOkll8lN/rfLXnFeF\nl35ur8kb+Xc2UYKS2+zGX5XSvGbWTalpnUp+uvqrztu5sGrNmPG5F/d9uyBUmch3sX2HS5z/\nZVKm+t2YpfIvWxSZ4Djr6K1qeSqdwAdigvRjyBrO596s+tiWIPmZ/mfm+ETbTN5btpoucP5Q\niqW5tAdm1ZcKjec7dCl5O/Fz3THzSqO5PrRs7s0h9gtPetjd7X61dmnoF8eWex/1+MigmnMv\n85lxjtN6SgbVtp4jWR4JcfqwxATpFOvrEPm55I6p1kr+YeU6S0paaZ8RffvWPWht/tWrqvKK\ncLLtXbUn2il345qpvEI7kzLPNtnjLfNX0pNSjybKViXKH1EpCv3SeQFuSSp3qe/ydaozacF7\ngk7tJiVc5A+qH0HOK29QLq3lWBspvwmpXH7YGPJRyB+c/4/9yv+s8NaQK3fbP9Q9wPHz7Q0W\n7DK5PRvor6q293xKFlR9tsJjufKbPoUthziUHov9yd484LQu+SXL4oZy69F7pZvPb4iRTsY6\nfGLZH1T2Bu0/2XWlH8ArGVNfPm8c8KD1z19N8kWHg2Gq04jKPFF1H0+Pt35GrIf8Em8/O6jW\n77/lvxNgavNA+kRlZSAoSHkNJvDbKu6KNonyu/aP3e1YWiy/Tn5QejnDi2vHKFchEufzf4d5\nNp8ly/pim5+pstahmjZw5B3ur+NJ6xyfnR2i56pPasgOLn1His8O+0ZnVdvZWT5NGTrres4P\nXzdyzXe5vz4c1uJuh/dEf2/i8PLpkGmndNvJfkZ5mzzTZ3hHFzt8Se/GEyNsn6yYLL+V9XkV\nd95D/sN+4RCIiLrY8GnwjuucXjZYpY6TL2CXewtlDzvCeQdlFxrHvpXvRt3Ln/Pw18l/Z9bj\nw/xajmdDb8WHu3hSd3QsbC2/8PnDwf01v3x8bEP7dIR3g5frrepy8Ld8kPIR4m+CrxR36Wzd\nqKOPRWp+nUJb+Zmnrv0F1bvVLvM9QZu0OtvltQm2P28o02PfcO8riPrd61Y3cJewq3b9m7Kf\n1B+ZejvPqd2r3BWIktgVvDhCmRO21/qZmY9ii3q4fi+0vJbWJ/N2kx2L+1mi+xPmpJPMRi2D\nw27TeTX2V8Jo5cXTgX5Baq/qHTSdxzspH/g+yXY/W730LSqVL3aweb4V5wWl0bkU+z7v1V2z\nc6kTpVnbxs5JQe/tehEuT1vEFXBSwoJ0LDb4svojH8btqdmnwlnbnRP5L8zxKHA6aMt173PP\nTFWu8O6u8L0fiXpfredkf/9Zm/U/0/1FfPiQr85ODkty9Qqu/4O8lvWfEPegbaKFvp+klzfW\n63yKh2/fEqT28RBNefIHa1Mmu+4owxVwWuLeR1qsdWa2h1W7v+KHHp7pzD+IL1dpN8w2Z8x9\n3wXJV5AnVJhTfcqTA5I7Clb1Cglu9LHLfi90vGRSpovy25h7B9eGr0mngaX/12w3NR/k2ZY1\nlF59Nl7kup/s/U6erRv0+eHHmAvDBztd0v2sSsHk8p8knRJS3dMLS8V3Rs63FMS/6bqnr06s\n0j5BK7Umao/tGv/Ydm50l4y/nS+pV/ZnK7OHP92bOp4XYFaqf/ghSPyA8zHijOmHO8ufk3zN\nVCaiuWDJirr91XDnz9H5x2/s9XDrc0G+2sclVGwMOTPDYT7VJy9rd1X1RDLfx3z5jTbwmj+C\npObG1+PLf8XdlXCNq+e6fuvKhrjuZQxLTLebPFuiKG7J6IE+jLikNl8biTeI/KKyBCnttvKf\nrub8bY3vT9RnWerpKytxOpudPkDhwvC+Kfpfpq5vBzvzCr7Pzj8qS5AWsnDqiwJ+9zDL8HCJ\nlVEN57vupelS0OaH9SckgiiVJUg/MQ/ffr0KZDFPX+RcimA+fX9+o4V3THPdCwSoLEEqjgm8\nLy/czFZ6ukgq2+3LiL3GJbzjy/LgtcoSJD7hM9d9rjJ/RXj8kxCLmU+/ofdkJ6cvOwZjVJog\nBSLPv0TugotpRy68F8ROuu4FAiBIgWQXi/X3JlyrEKRAciVY/TvvQDgEKaA0Gey6D4iAIAWU\nx/HbsX6CIAEQQJAACCBIAAQQJAACCBIAAQQJgACCBEAAQQIggCABEECQAAggSAAEECQAAggS\nAAEECYAAggRAAEECIIAgARBAkAAIIEgABBAkAAIIEgABBAmAAIIEQABBAiCAIAEQQJAACCBI\nAAQQJAACCBIAAQQJgACCBEAAQQIggCABEBATpINnhKxWdvagvWU58NWqVRv+UOt0Yco+YRsA\noEJMkFiVFwso17e7R2JSVrHSnGLb4rMTazBF/RmXnfofZusohwdwRVCQGgTf9A3d6raEsQgz\n+/tZuW0L0rGG7PoR02fPfnpQHdbqrK1fut0gdld6Ot34AK4ICtKU7W1Yt81Uq+tpXm3Jf9V8\nSx4vDVK6ebntweIs03j7sOVQjQ7gmqgg8eK58azLkvMkq0sYKt9uCO1RXBqkWiPLHh6QYGs8\nFtx6/TnZL+yjc+dIhgZwi7AgcZ6XGceCb0l/8Q2fV2d+VrlbysaVBsn8UtnDz4XaW9tbm8bI\n2cVrJDCYwCBxfum9XpEU51j17rHeT2Wz7UFKvL/s4XsblDaLZoXXWYEggeGEBklS8ONHC3xe\n3TjTvEL53pLGJoy1bvF405x864N5z5aNJtnfjfX6A0ECg4kOEok/67NkpWEZZz/AnWvLoruN\nePSRtK4R7LaL5Xq/Uz1qOoIExhITpLCnadd3OmOCrbWysW2LC15tHSxfmzN3XFRcoffJgQxB\nAmNdzReJr/xv585c1Td+/z0xx+iNgWub+CD9metcK5vnY1NhTk/R4X2/aVw5d1pUtaQ+LIAo\n4oNkn9OjNs+nlOPVgYL5HUPlk7Y6adlce1G9tWkWAQQxLkgq83xU5/Rc6MDCW3Y0NxnSxsQm\nay6qtjbVYQGMYFyQVOb5qM7peYxNkM7q9iW+zn/vzd7RWlRtbarDAhhBzO7WzkEt2xAq83xU\n5/TU66ncvR+Tx4vbtdVaVG1tqsMCGEHM7hYUFFYq2DaEyjwf1Tk95hnK3T62hfPnw7UWVV2b\n2rAARhCzu02JLrtmZt/NVeb5cLU5PbWtk39WsG85Hxentajq2tSGBTCCmN2tsE37Qnvbvker\nzPNRVJzTk2ZabOH858TIPL6tanetRVXXpjYsgBEE7W454ZPsTfserTLPx6b8nJ6D1VmdLs1M\nbD4vNkdu01pUfW0qwwIYQdTudqH0Wxs2ZdoaKvN8bMrP6TkwIJqF3LZRaj2xR3tR9bWpDAtg\nAGOetwuXndav2eb02Gt5JbrLulfSKAIIYEyQzjHnj537UHOvpFEEEABBAiCAIAEQQJAACCBI\nAAQQJAACCBIAAQQJgACCBEAAQQIggCABEBAZpGO7Nu4+pbSK1p31veZeSaMIIJS4IC1qoHwR\nQ9NlNDX3ShpFAMGEBWkBC0tOyxjSwcSWUNTcK2kUAUQTFqQbUqzf8Phbk2YUNfdKGkUA0YQF\nKXSLrZEVRlFzr6RRBBBNWJDit9oaC+tS1NwraRQBRBMWpJHTbI3UCRQ190oaRQDRhAXpWIfB\na/cdylnRIzn3sMTXmnsljSKAaMKCVP7biJmvNfdKGkUA0YTtar0HlONrzb2SRhFANDxnAxBA\nkAAICAuSZXlq6+ZWFDX3ShpFANGEBWkOYxGxVhQ190oaRQDRhAWpXsoBypp7JY0igGjCgmTO\nJq25V9IoAogm7oi0lbTmXkmjCCCasCBNziCtuVfSKAKIJixIF1MGr8/JVVDU3CtpFAFEM2KK\nEEXNvZJGEUA0YbvboLR0O4qaeyWNIoBoeN4GICAwSMXS//Kzd1mIau6VNIoAYgkLUnFGP84P\nNmIs6SJFzb2SRhFANGFBymSPc97DNCYjKJOi5l5JowggmrAg3dyX8yMm6RX/yNYUNfdKGkUA\n0YQFKWoh52+xrznPqkpRc6+kUQQQTViQoqU9elBkAefzIylq7pU0igCiiTu1G8xPRPWRGqNv\npKi5V9IoAogmLEgzWac6bBPnS0InU9TcK2kUAUQTFqQrI8Jj35Dua7c4S1Fzr6RRBBBN8MyG\nh3/fWlShdHykxanGZ3zvVPvkTV6x9sskp1Lmf5xKnG95SaUIIJDgIIU4/7rK56zQuV+9Ls+9\n8f7nP+SWHUaO3NbB3rx8/Ncfvly+aM7TKTE8/+yRA7t3bPhq+bI35896bsrYGuMrrOnc7rXz\nuiZSbT+AewQHKeifTqVVakG6/44uzWuHMcaCrruhU89hQ1PbB9W6dXh6v+T29cKC5ancppCw\nsCohwfaZ3WHV6jZq2e6O5P63vFe2kqz7UlvEMhbR7NZBAv9JACoEByn4PafSWlag1Tvv0I8b\nlv/jpYkjbqzSdLOlaZMhGdP+L6VK5zv6D3v8qVlz31y6/N9f/bDjwMGzGrN/OjeY8NqqHafJ\nNh7AbYKDFLbeqbRF7YhUwfRaIzhPmSo3s9z/naNh+OQE+AuCBEAAQQIggCABEHArSOnM228S\nQZDg2uBlkN6xXYZ+wcWSCBJcG7wM0lw2aIpso4slESS4NngZpOlsu1ur9z1ICxAkuAroBel4\nep2Ilq8VyUE6MKthaMIMC+cD2MnkKmv4eP1XTed22EzasKOibyc4lZx8+PDCHTteXSo3Vz/v\nurvNPxa53RUC1U7XT9NC6ATpVN3YsS+nsnQ5SA+0yZydwD6UnvXZ4Ltn7uFp7HTxYe05BBkM\nwD8+EJASN+gEaQz7QrrtyfZKQUqScr6T3cP5SHZXiVTtzZ6qxtgNWhs9qns2gD/UeVtAStyg\nHSRLXIL83XAHNp6WgrRaLgS3lw9OSni6skaZS6fGsIXqyyJI4CeVL0hH2Z32Zrp0VJLENpeb\nO+TmhhV50u0vYdXVJ6COajQMwB+iKl2Q9rNUe9N21c4aJMerDH3YNtVl/5UM4Bcpu4mS4SHt\nIOWxJHtTM0gPMVdvJAFcE3QuNsTHyVcSf5231zlIFxd8qHRJYvjFVgCuG6RRbLF0O5DtdA5S\nSd2ofdLdJ6yNIRsJUNnpBOlwrZBH56Sy4WqndmtMkenP9DHF7DRmKwEqOb2ZDb8PrWFu9Eqx\n6muk7++uGlJnOH5eEkCBHxoDIIAgARBAkAAIIEgABBAkAAIIEgABBAmAAIIEQABBAiCAIAEQ\nQJAACCBIAAQQJAACCBIAAQQJgACCBEAAQQIggCABEECQAAggSAAEECQAAggSAAEECYAAggRA\nAEECIIAgARBAkAAIIEgABBAkAAIIEgABBAmAAIIEQABBAiCAIAEQQJAACCBIAAQQJAACCBIA\nAQQJgACCBEAAQQIggCABEECQAAggSAAEECQAAggSAAEECYAAggRAAEECIIAgARBAkAAIIEgA\nBBAkAAIIEgABBAmAAIIEQABBAiCAIAEQQJAACCBIAAQQJAACCBIAAQQJgACCBEAAQQIggCAB\nEECQAAggSAAEECQAAggSAAEECYAAggRAAEECIIAgARBAkAAIIEgABBAkAAIIEgABBAmAwP8D\ngibEuO7qUZoAAAAASUVORK5CYII="
          },
          "metadata": {
            "image/png": {
              "width": 420,
              "height": 420
            }
          }
        }
      ],
      "source": [
        "##тут буде код\n",
        "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "  BiocManager::install(\"karyoploteR\")\n",
        "   BiocManager::install(\"GenomicRanges\")\n",
        "\n",
        "\n",
        "\n",
        "library(karyoploteR)\n",
        "library(GenomicRanges)\n",
        "\n",
        "genome_df <- data.frame(\n",
        "  chr = paste0(\"chr\", 1:5),\n",
        "  start = rep(1, 5),\n",
        "  end = c(34964571, 22037565, 25499034, 20862711, 31270811)\n",
        ")\n",
        "genome_gr <- makeGRangesFromDataFrame(genome_df)\n",
        "\n",
        "snp_pos <- sample(1:1e7, 25)\n",
        "snps <- data.frame(\n",
        "  chr = paste0(\"chr\", sample(1:5,25, replace=TRUE)),\n",
        "  start = snp_pos,\n",
        "  end = snp_pos\n",
        ")\n",
        "snps_gr <- makeGRangesFromDataFrame(snps)\n",
        "snp_labels <- paste0(\"snp_\", 1:25)\n",
        "\n",
        "plot.params <- getDefaultPlotParams(plot.type=1)\n",
        "plot.params$data1outmargin <- 600\n",
        "\n",
        "kp <- plotKaryotype(genome=genome_gr, plot.type = 1, plot.params = plot.params)\n",
        "kpPlotMarkers(kp, snps_gr, labels = snp_labels)\n",
        "\n",
        "\n",
        "### додатково....\n",
        "\n",
        "numeric_data <- data.frame(\n",
        "  y = rnorm(100,mean = 1,sd = 0.5  ),\n",
        "  chr = rep(\"chr4\", 100),\n",
        "  start = seq(1,20862711, 20862711/100),\n",
        "  end = seq(1,20862711, 20862711/100)\n",
        ")\n",
        "\n",
        "\n",
        "numeric_data_gr <- makeGRangesFromDataFrame(numeric_data)\n",
        "\n",
        "plot.params <- getDefaultPlotParams(plot.type=2)\n",
        "plot.params$data1outmargin <- 800\n",
        "plot.params$data2outmargin <- 800\n",
        "plot.params$topmargin <- 800\n",
        "\n",
        "kp <- plotKaryotype(genome=genome_gr, plot.type = 2, plot.params = plot.params)\n",
        "kpPlotMarkers(kp, snps_gr, labels = snp_labels)\n",
        "kpLines(kp, numeric_data_gr, y = numeric_data$y, data.panel=2)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "знов якийсь код\n",
        "\n",
        "\n",
        "# Predicting open reading frames in long\n",
        "reference sequences"
      ],
      "metadata": {
        "id": "RIVj84nus48a"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "BiocManager::install(\"systemPipeR\")\n",
        "BiocManager::install(\"Biostrings\")\n",
        "library(Biostrings)\n",
        "library(systemPipeR)\n",
        "dna_object <- readDNAStringSet(file.path(getwd(), \"datasets\",\"ch2\", \"arabidopsis_chloroplast.fa\"))\n",
        "\n",
        "predicted_orfs <- predORF(dna_object, n = 'all', type = 'gr', mode='ORF', strand = 'both', longest_disjoint = TRUE)\n",
        "predicted_orfs\n",
        "\n",
        "\n",
        "bases <- c(\"A\", \"C\", \"T\", \"G\")\n",
        "raw_seq_string <- strsplit(as.character(dna_object), \"\")\n",
        "\n",
        "seq_length <- width(dna_object[1])\n",
        "counts <- lapply(bases, function(x) {sum(grepl(x, raw_seq_string))}  )\n",
        "probs <- unlist(lapply(counts, function(base_count){signif(base_count / seq_length, 2) }))\n",
        "\n",
        "get_longest_orf_in_random_genome <- function(x,\n",
        "                                             length = 1000,\n",
        "                                             probs = c(0.25, 0.25, 0.25, 0.25),\n",
        "                                             bases = c(\"A\",\"C\",\"T\",\"G\")){\n",
        "\n",
        "  random_genome <- paste0(sample(bases, size = length, replace = TRUE, prob = probs), collapse = \"\")\n",
        "  random_dna_object <- DNAStringSet(random_genome)\n",
        "  names(random_dna_object) <- c(\"random_dna_string\")\n",
        "  orfs <- predORF(random_dna_object, n = 1, type = 'gr', mode='ORF', strand = 'both', longest_disjoint = TRUE)\n",
        "  return(max(width(orfs)))\n",
        "}\n",
        "\n",
        "random_lengths <- unlist(lapply(1:10, get_longest_orf_in_random_genome, length = seq_length, probs = probs, bases = bases))\n",
        "longest_random_orf <- max(random_lengths)\n",
        "keep <- width(predicted_orfs) > longest_random_orf\n",
        "orfs_to_keep <- predicted_orfs[keep]\n",
        "orfs_to_keep\n",
        "\n",
        "##writing to file\n",
        "extracted_orfs <- BSgenome::getSeq(dna_object, orfs_to_keep)\n",
        "names(extracted_orfs) <- paste0(\"orf_\", 1:length(orfs_to_keep))\n",
        "writeXStringSet(extracted_orfs, \"saved_orfs.fa\")"
      ],
      "metadata": {
        "id": "OfIo1cphgIzW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Splitting sequence files into operational taxonomic units"
      ],
      "metadata": {
        "id": "7byzYTC5C6ef"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "\n",
        "BiocManager::install(\"kmer\")\n",
        "BiocManager::install(\"magrittr\")\n",
        "library(kmer)\n",
        "library(magrittr)\n",
        "library(dplyr)  # Load the dplyr package explicitly\n",
        "\n",
        "seqs <- ape::read.fastq(file.path(getwd(),\"datasets\",\"ch5\",\"fq\",\"SRR9040914ab.fq.gz\"))\n",
        "otu_vec <- otu(seqs, k = 6, threshold = 0.99 )\n",
        "\n",
        "# Create a data frame with the results\n",
        "result_df <- data.frame(seqid = names(otu_vec), cluster = otu_vec, row.names = NULL)\n",
        "\n",
        "# Use dplyr to group by cluster and summarize\n",
        "result_summary <- result_df %>%\n",
        "  group_by(cluster) %>%\n",
        "  summarize(count = dplyr::n())\n",
        "\n",
        "# Print the result summary\n",
        "print(result_summary)\n"
      ],
      "metadata": {
        "id": "CgKuy4pOAUFl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Converting MS formats to move data between tools"
      ],
      "metadata": {
        "id": "Kdtg8OlaE9qk"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "\n",
        "BiocManager::install(\"mzR\")\n",
        "library(mzR)\n",
        "mzxml_file <- file.path(getwd(), \"datasets\", \"ch6\", \"threonine_i2_e35_pH_tree.mzXML\")\n",
        "mzdata <- openMSfile(mzxml_file)\n",
        "header_info <- header(mzdata)\n",
        "peak_data_list <- spectra(mzdata) #must be a list - if only one spectrum, remember to wrap it in list()\n",
        "writeMSData(peak_data_list, file.path(getwd(), \"datasets\", \"ch6\", \"out.mz\"), header = header_info, outformat = \"mzml\", rtime_seconds = TRUE ) #can import mzml mzxml and mzdata and export mzml and mzxml"
      ],
      "metadata": {
        "id": "tfqLe2jdFDBn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "install.packages(\"tidyverse\")\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "\n",
        "BiocManager::install(\"SummarizedExperiment\")\n",
        "BiocManager::install(\"DESeq2\")\n",
        "\n",
        "# Load required libraries\n",
        "library(tidyverse)\n",
        "library(DESeq2)\n",
        "library(SummarizedExperiment)\n",
        "\n",
        "count_dataframe <- readr::read_tsv(file.path(getwd(), \"datasets\", \"ch1\", \"modencodefly_count_table.txt\" ))\n",
        "genes <- count_dataframe[['gene']]\n",
        "count_dataframe[['gene']] <- NULL\n",
        "count_matrix <- as.matrix(count_dataframe)\n",
        "rownames(count_matrix) <- genes\n",
        "\n",
        "pheno_data <- readr::read_table2(file.path(getwd(), \"datasets\", \"ch1\", \"modencodefly_phenodata.txt\"))\n",
        "\n",
        "experiments_of_interest <- c(\"L1Larvae\", \"L2Larvae\")\n",
        "columns_of_interest <- which( pheno_data[['stage']] %in%\n",
        "experiments_of_interest )\n",
        "\n",
        "library(magrittr)\n",
        "grouping <- pheno_data[['stage']][columns_of_interest] %>%\n",
        " forcats::as_factor()\n",
        "\n",
        "counts_of_interest <- count_matrix[, columns_of_interest]\n",
        "\n",
        "# DESeq2 analysis\n",
        "library(\"DESeq2\")\n",
        "dds <- DESeqDataSetFromMatrix(countData = counts_of_interest,\n",
        " colData = grouping,\n",
        " design = ~ stage)\n",
        "\n",
        "\n",
        "dds <- DESeq(dds)\n",
        "res <- results(dds, contrast = c(\"stage\", \"L2Larvae\", \"L1Larvae\"))\n",
        "\n",
        "# SummarizedExperiment analysis\n",
        "load(file.path(getwd(), \"datasets/ch1/modencodefly_eset.RData\"))\n",
        "summ_exp <- makeSummarizedExperimentFromExpressionSet(modencodefly.eset)\n",
        "ddsSE <- DESeqDataSet(summ_exp, design = ~ stage)\n",
        "\n",
        "ddsSE <- DESeq(ddsSE)\n",
        "resSE <- results(ddsSE, contrast = c(\"stage\", \"L2Larvae\", \"L1Larvae\"))\n"
      ],
      "metadata": {
        "id": "GPs-VRRalVYo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "resSE"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 106
        },
        "id": "nOHcjPRgmEM6",
        "outputId": "f02382a5-f285-4939-88df-cc444db5ba4f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ERROR",
          "evalue": "ignored",
          "traceback": [
            "Error in eval(expr, envir, enclos): object 'resSE' not found\nTraceback:\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Estimating differential expression with edgeR"
      ],
      "metadata": {
        "id": "ZEN7NWKuz0_p"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"edgeR\")\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"readr\")\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"magrittr\")\n",
        "\n",
        "\n",
        "count_dataframe <- readr::read_tsv(file.path(getwd(), \"datasets\", \"ch1\", \"modencodefly_count_table.txt\" ))\n",
        "genes <- count_dataframe[['gene']]\n",
        "count_dataframe[['gene']] <- NULL\n",
        "count_matrix <- as.matrix(count_dataframe)\n",
        "rownames(count_matrix) <- genes\n",
        "pheno_data <- readr::read_table2(file.path(getwd(), \"datasets\", \"ch1\", \"modencodefly_phenodata.txt\"))\n",
        "\n",
        "\n",
        "experiments_of_interest <- c(\"L1Larvae\", \"L2Larvae\")\n",
        "columns_of_interest <- which( pheno_data[['stage']] %in% experiments_of_interest ) #gives the numeric indices of the above experiments in the matrix\n",
        "\n",
        "\n",
        "library(magrittr)\n",
        "grouping <- pheno_data[['stage']][columns_of_interest] %>%\n",
        "  forcats::as_factor()\n",
        "\n",
        "counts_of_interest <-  count_matrix[,columns_of_interest]\n",
        "\n",
        "library(edgeR)\n",
        "count_dge <- edgeR::DGEList(counts = counts_of_interest, group = grouping)\n",
        "\n",
        "design <- model.matrix(~ grouping)\n",
        "eset_dge <- edgeR::estimateDisp(count_dge, design)\n",
        "fit <- edgeR::glmQLFit(eset_dge, design)\n",
        "result <- edgeR::glmQLFTest(fit, coef=2)\n",
        "topTags(result)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "p4rLS-xV2tQX",
        "outputId": "3123b7db-e59b-47b8-d587-31e35acdf5b0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'edgeR'”\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'readr'”\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'magrittr'”\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "ERROR",
          "evalue": "ignored",
          "traceback": [
            "Error: '/content/datasets/ch1/modencodefly_count_table.txt' does not exist.\nTraceback:\n",
            "1. readr::read_tsv(file.path(getwd(), \"datasets\", \"ch1\", \"modencodefly_count_table.txt\"))",
            "2. vroom::vroom(file, delim = \"\\t\", col_names = col_names, col_types = col_types, \n .     col_select = {\n .         {\n .             col_select\n .         }\n .     }, id = id, .name_repair = name_repair, skip = skip, n_max = n_max, \n .     na = na, quote = quote, comment = comment, skip_empty_rows = skip_empty_rows, \n .     trim_ws = trim_ws, escape_double = TRUE, escape_backslash = FALSE, \n .     locale = locale, guess_max = guess_max, show_col_types = show_col_types, \n .     progress = progress, altrep = lazy, num_threads = num_threads)",
            "3. vroom_(file, delim = delim %||% col_types$delim, col_names = col_names, \n .     col_types = col_types, id = id, skip = skip, col_select = col_select, \n .     name_repair = .name_repair, na = na, quote = quote, trim_ws = trim_ws, \n .     escape_double = escape_double, escape_backslash = escape_backslash, \n .     comment = comment, skip_empty_rows = skip_empty_rows, locale = locale, \n .     guess_max = guess_max, n_max = n_max, altrep = vroom_altrep(altrep), \n .     num_threads = num_threads, progress = progress)",
            "4. (function (path, write = FALSE) \n . {\n .     if (is.raw(path)) {\n .         return(rawConnection(path, \"rb\"))\n .     }\n .     if (!is.character(path)) {\n .         return(path)\n .     }\n .     if (is_url(path)) {\n .         if (requireNamespace(\"curl\", quietly = TRUE)) {\n .             con <- curl::curl(path)\n .         }\n .         else {\n .             inform(\"`curl` package not installed, falling back to using `url()`\")\n .             con <- url(path)\n .         }\n .         ext <- tolower(tools::file_ext(path))\n .         return(switch(ext, zip = , bz2 = , xz = {\n .             close(con)\n .             stop(\"Reading from remote `\", ext, \"` compressed files is not supported,\\n\", \n .                 \"  download the files locally first.\", call. = FALSE)\n .         }, gz = gzcon(con), con))\n .     }\n .     path <- enc2utf8(path)\n .     p <- split_path_ext(basename_utf8(path))\n .     if (write) {\n .         path <- normalizePath_utf8(path, mustWork = FALSE)\n .     }\n .     else {\n .         path <- check_path(path)\n .     }\n .     if (is_installed(\"archive\")) {\n .         formats <- archive_formats(p$extension)\n .         extension <- p$extension\n .         while (is.null(formats) && nzchar(extension)) {\n .             extension <- split_path_ext(extension)$extension\n .             formats <- archive_formats(extension)\n .         }\n .         if (!is.null(formats)) {\n .             p$extension <- extension\n .             if (write) {\n .                 if (is.null(formats[[1]])) {\n .                   return(archive::file_write(path, filter = formats[[2]]))\n .                 }\n .                 return(archive::archive_write(path, p$path, format = formats[[1]], \n .                   filter = formats[[2]]))\n .             }\n .             if (is.null(formats[[1]])) {\n .                 return(archive::file_read(path, filter = formats[[2]]))\n .             }\n .             return(archive::archive_read(path, format = formats[[1]], \n .                 filter = formats[[2]]))\n .         }\n .     }\n .     if (!write) {\n .         compression <- detect_compression(path)\n .     }\n .     else {\n .         compression <- NA\n .     }\n .     if (is.na(compression)) {\n .         compression <- tools::file_ext(path)\n .     }\n .     if (write && compression == \"zip\") {\n .         stop(\"Can only read from, not write to, .zip\", call. = FALSE)\n .     }\n .     switch(compression, gz = gzfile(path, \"\"), bz2 = bzfile(path, \n .         \"\"), xz = xzfile(path, \"\"), zip = zipfile(path, \"\"), \n .         if (!has_trailing_newline(path)) {\n .             file(path)\n .         } else {\n .             path\n .         })\n . })(\"/content/datasets/ch1/modencodefly_count_table.txt\")",
            "5. check_path(path)",
            "6. stop(\"'\", path, \"' does not exist\", if (!is_absolute_path(path)) {\n .     paste0(\" in current working directory ('\", getwd(), \"')\")\n . }, \".\", call. = FALSE)"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Selecting and classifying variants with VariantAnnotation\n",
        "\n",
        "# Значення методу\n",
        "\n",
        "Призначення: Класичним етапом біоінформатичних досліджень є фільтрація та класифікація та фільтрація окремих варіантів на основі їх певних особливостей. Для цього данізберігають у VCF-файли.Прикладом ознаки, за якою відфільтровують варіанти є глибина покриття(кількість зчитувань) альтернативного алелю у вибірці.\n",
        "\n",
        "Мета:Здійснити формування вихідного VCF-файлу, та провести його фільтрацію з наступним отриманням вибірок в яких альтернативний алель є основним.\n",
        "\n",
        "# Ключові етапи:\n",
        "\n",
        "1\\. Завантаження та встановлення необхідних пакетів\n",
        "\n",
        "2.Формування перфільтру\n",
        "\n",
        "3.Формування фільтру\n",
        "\n",
        "4.Створення VCF-файлу\n",
        "\n",
        "5.Застосуванню префільтру і фільтру до VCF-файлу\n",
        "\n",
        "# 1.Завантаження та встановлення необхідних пакетів\n",
        "\n",
        "Встановлюємо пакет \"Biocunductor\" у його останній версії, за допомогою його менеджера для установики \"BiocManager\", якщо це необхідно.Встановюємо пакет \"VarianAnnotation\",(якщо це необхідно) який буде використаний для подальшого аналізу. Завантажуємо бібліотеку \"VarianAnnotation\".\n",
        "\n",
        "```{r}\n",
        "if (!requireNamespace(\"BiocManager\"))\n",
        "  install.packages(\"BiocManager\")\n",
        "BiocManager::install(version = \"3.18\")\n",
        "\n",
        "BiocManager::install(\"VarianAnnotation\")\n",
        "\n",
        "library(VariantAnnotation)\n",
        "```\n",
        "\n",
        "# 1.1.Створення функції префільтру\n",
        "\n",
        "Створюємо функція liis_not_microsat, яка перевіряє, чи в рядку x не міститься рядок \"microsat\" і повертає здані про зразок лише у разі відсутності даної ознаки.Це досягається завдяки викоритсанню функції grepl, з логічним оператором !(НЕ).\n",
        "\n",
        "```{r}\n",
        "liis_not_microsat <- function(x){ !grepl(\"microsat\", x, fixed =TRUE)}\n",
        "```\n",
        "\n",
        "# 1.2. Завантажння функцію префільтра в об'єкт FilterRules\n",
        "\n",
        "Створюємо об'єкт FilterRules з назвою \"microsat\" та функцією is_not_microsat #для prefilters. FilterRules є функцією, яка створює об'єкт правил фільтрації #для використання в подальших операціях фільтрації даних.\n",
        "\n",
        "```{r}\n",
        "prefilters<-FilterRules(list(microsat = liis_not_microsat))\n",
        "```\n",
        "\n",
        "# 2.1.Створення функції фільтра\n",
        "\n",
        "Створюємо функцію major_alt, яка визначає, чи референсний алель є головним варіантом у менше ніж половині вибірок.\n",
        "\n",
        "```{r}\n",
        "major_alt <- function(x){\n",
        "  af <- info(x)$AF ## also geno() fixed()\n",
        "  result <- unlist(lapply(af, function(x){x[1] < 0.5}))\n",
        "  return(result)\n",
        "}\n",
        "```\n",
        "\n",
        "# 2.2. Завантажння функцію фільтра в об'єкт FilterRules\n",
        "\n",
        "Створюємо об'єкт FilterRules з назвою \"alt_is_major\" та функцією major_alt для фільтрації варіантів.\n",
        "\n",
        "```{r}\n",
        "filters <- FilterRules(list(alt_is_major = major_alt))\n",
        "```\n",
        "\n",
        "# 3.Завантаження VCF-файлу та застосування фільтрів:\n",
        "\n",
        "Створюємо шлях до VCF-файлу в папці \"ch2\".\n",
        "\n",
        "```{r}\n",
        "vcf_file <- file.path(getwd(), \"ch2\", \"sample.vcf.gz\")\n",
        "```\n",
        "\n",
        "# 4.Застосування фільтрів для завантаженого VCF-файлу\n",
        "\n",
        "Функцію filterVcf використовуємо ядля завантаження VCF-файлу, застосування попередніх фільтрів (prefilters) та фільтрів для варіантів (filters), і збереження результату у файл \"filtered\\_.vcf\".\n",
        "\n",
        "```{r}\n",
        "filterVcf(vcf_file, \"hg17\", \"filtered.vcf40\", prefilters = prefilters,\n",
        "          filters = filters)\n",
        "```\n",
        "\n",
        "# Висновок:\n",
        "\n",
        "Створення VCF-файлів та фільтрів до ним використовується для ефективного варіантів за конкретною ознакою, в даному випадку за озкауою переважання кількості альтернативного алеля. Це досягається у 4 ключові етапи: формування перфільтру, формування фільтру, створення VCF-файлу та застосуванню до нього префільтру і фільтру.\n"
      ],
      "metadata": {
        "id": "seweCzBjGAts"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "\n",
        "BiocManager::install(\"VariantAnnotation\")\n",
        "\n",
        "library(VariantAnnotation)\n",
        "\n",
        "\n",
        "is_not_microsat <- function(x){ !grepl(\"microsat\", x, fixed = TRUE)}\n",
        "\n",
        "prefilters <- FilterRules(list(microsat = is_not_microsat) )\n",
        "\n",
        "major_alt <- function(x){\n",
        "  af <- info(x)$AF ## also geno() fixed()\n",
        "  result <- unlist(lapply(af, function(x){x[1] < 0.5}))\n",
        "  return(result)\n",
        "}\n",
        "filters <- FilterRules(list(alt_is_major = major_alt))\n",
        "\n",
        "vcf_file <- file.path(getwd(), \"datasets\", \"ch2\", \"sample.vcf.gz\")\n",
        "filterVcf(vcf_file, \"hg17\", \"filtered.vcf\", prefilters = prefilters, filters = filters)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c67b11c1-5751-45bb-aa44-b3222c660953",
        "id": "uWehY2AJR0Sr"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'VariantAnnotation'\n",
            "\n",
            "also installing the dependencies ‘formatR’, ‘lambda.r’, ‘futile.options’, ‘filelock’, ‘abind’, ‘SparseArray’, ‘futile.logger’, ‘snow’, ‘BH’, ‘plogr’, ‘png’, ‘rjson’, ‘BiocFileCache’, ‘matrixStats’, ‘RCurl’, ‘GenomeInfoDbData’, ‘S4Arrays’, ‘DelayedArray’, ‘bitops’, ‘BiocParallel’, ‘RSQLite’, ‘KEGGREST’, ‘XML’, ‘GenomicAlignments’, ‘BiocIO’, ‘restfulr’, ‘biomaRt’, ‘BiocGenerics’, ‘MatrixGenerics’, ‘GenomeInfoDb’, ‘GenomicRanges’, ‘SummarizedExperiment’, ‘Rsamtools’, ‘zlibbioc’, ‘Biobase’, ‘S4Vectors’, ‘IRanges’, ‘XVector’, ‘Biostrings’, ‘AnnotationDbi’, ‘rtracklayer’, ‘BSgenome’, ‘GenomicFeatures’, ‘Rhtslib’\n",
            "\n",
            "\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "Loading required package: BiocGenerics\n",
            "\n",
            "\n",
            "Attaching package: ‘BiocGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    IQR, mad, sd, var, xtabs\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
            "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
            "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
            "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
            "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
            "    table, tapply, union, unique, unsplit, which.max, which.min\n",
            "\n",
            "\n",
            "Loading required package: MatrixGenerics\n",
            "\n",
            "Loading required package: matrixStats\n",
            "\n",
            "\n",
            "Attaching package: ‘MatrixGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
            "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
            "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
            "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
            "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
            "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
            "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
            "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
            "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
            "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
            "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
            "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
            "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
            "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
            "    rowWeightedSds, rowWeightedVars\n",
            "\n",
            "\n",
            "Loading required package: GenomeInfoDb\n",
            "\n",
            "Loading required package: S4Vectors\n",
            "\n",
            "Loading required package: stats4\n",
            "\n",
            "\n",
            "Attaching package: ‘S4Vectors’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:utils’:\n",
            "\n",
            "    findMatches\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    expand.grid, I, unname\n",
            "\n",
            "\n",
            "Loading required package: IRanges\n",
            "\n",
            "Loading required package: GenomicRanges\n",
            "\n",
            "Loading required package: SummarizedExperiment\n",
            "\n",
            "Loading required package: Biobase\n",
            "\n",
            "Welcome to Bioconductor\n",
            "\n",
            "    Vignettes contain introductory material; view with\n",
            "    'browseVignettes()'. To cite Bioconductor, see\n",
            "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘Biobase’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:MatrixGenerics’:\n",
            "\n",
            "    rowMedians\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    anyMissing, rowMedians\n",
            "\n",
            "\n",
            "Loading required package: Rsamtools\n",
            "\n",
            "Loading required package: Biostrings\n",
            "\n",
            "Loading required package: XVector\n",
            "\n",
            "\n",
            "Attaching package: ‘Biostrings’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:base’:\n",
            "\n",
            "    strsplit\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘VariantAnnotation’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:base’:\n",
            "\n",
            "    tabulate\n",
            "\n",
            "\n",
            "starting prefilter\n",
            "\n",
            "prefiltering 9 records\n",
            "\n",
            "prefiltered to /tmp/RtmprdmxpQ/file28ebab717e\n",
            "\n",
            "prefilter compressing and indexing ‘/tmp/RtmprdmxpQ/file28ebab717e’\n",
            "\n",
            "starting filter\n",
            "\n",
            "filtering 8 records\n",
            "\n",
            "completed filtering\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# *Rarefying counts and correcting for sample differences using metacoder*\n",
        "\n",
        "# Значення методу\n",
        "\n",
        "Призначення: Метагеномний аналіз націлений на аналіз кількості і різноманіття видів у піддослідних зразках.При проведенні ПЛР та секвенування, задля збору метагеномних даних, однією з помилок є різна кількість зчитувань окремих зразків межах одного циклу секвенування. Це призводить до сильного спотворення результатів дослідження, адже теким чином для зразків з більшою кількістью зчитувань збільшується вірогідність детекції рідкісних такоснів, в резульатів ми спостерігаємо штучне збільшення різноманіття в ньому.Для заподігання цьому артефакту використвоують метод рарифікації(розрідження).\n",
        "\n",
        "Мета: Нормалізувати секвеновані зразки на основі зразка з мінімальною кількістью зчитувань для можливості подальшого аналізу і порівняння.\n",
        "\n",
        "# Основні етапи:\n",
        "\n",
        "1\\. Завантаження та встановлення необхідних для роботи пакетів.\n",
        "\n",
        "2.  Завантаження даних для аналізу.\n",
        "\n",
        "3\\. Створення гістограми для перевірки розподілу кількості OTU counts у вибірках\n",
        "\n",
        "4\\. Рарифікація зразків\n",
        "\n",
        "# 1.Завантаження та встановлення необхідних пакетів\n",
        "\n",
        "Встановлюємо пакет \"Biocunductor\" у його останній версії, за допомогою його менеджера для установики \"BiocManager\", якщо це необхідно.Встановюємо пакет \"metacode\", (якщо це необхідно) який буде використаний для подальшого аналізу. Завантажуємо бібліотеку \"metacoder\".\n",
        "\n",
        "```{r}\n",
        "if (!requireNamespace(\"metacoder\"))\n",
        "if (!requireNamespace(\"BiocManager\"))\n",
        "  install.packages(\"BiocManager\")\n",
        "BiocManager::install(version = \"3.18\")\n",
        "\n",
        "if (!requireNamespace(\"metacoder\"))\n",
        "  install.packages(\"metacoder\")\n",
        "BiocManager::install(\"metacoder\")\n",
        "\n",
        "library(metacoder)\n",
        "```\n",
        "\n",
        "# 2.Завантажуємо дані для аналізу\n",
        "\n",
        "Cтворюємо шлях до файлу \"rich_high_count_otu.biom\"в поточні робочій директорії, зберігаючі дані у biom_file . Далі функція parse_qiime_biom використовуюється для обробки biom_file що знаходиться в QIIME форматі. Отримані дані зберігаємо в змінній taxdata, завдяки чому вони стають доступними для роботи в metacoder.\n",
        "\n",
        "```{r}\n",
        "biom_file<- file.path(getwd(), \"rich_high_count_otu.biom\")\n",
        "\n",
        "taxdata <- parse_qiime_biom(biom_file)\n",
        "```\n",
        "\n",
        "# 3.Створюємо гістограму для перевірки розподілу кількості OTU counts у вибірках:\n",
        "\n",
        "Ми створюємо символьний вектор імен зразків за допомогою функції paste(), і вилучаємо іменні індекси зі значеннями що містять стовпці в таблиці otu_table. За допомогою виразу 'otu_table' ми отримуємо самі значення OTU-таблиці з об'єкта taxdata\\$data. Функція colSums() проводить обрахунки кільксті OTU в кожній вибірці, і зберігає данні у вектор hist_data. Далі за допомогою функції hist ми будуємо гістограму, і додаємо криву щільності з використанням функції lines.\n",
        "\n",
        "```{r}\n",
        "sample_ids <- paste0(\"Sample\", 1:6)\n",
        "taxdata$data$otu_table\n",
        "\n",
        "hist_data <- colSums(taxdata$data$otu_table[, sample_ids])\n",
        "hist(hist_data, prob= TRUE, breaks=3)\n",
        "lines(density(hist_data, adjust = 2), col=\"blue\")\n",
        "```\n",
        "\n",
        "# 4.Використовуємо функцію рарифікації та відфільтровуємо малопоширені OTU(Operational Taxonomic Units), які можуть виникати\n",
        "\n",
        "За допомогою функції raryfy_obs ми проводимо рарифікацію даних з OTU-таблиціРезультати рарифікації зберігаємо у змінній taxdata$data$rarefied_otus. Викликається функція rarefy_obs для рарифікації OTU-таблиці. Результат зберігається у змінній taxdata$data$rarefied_otus.Створюється логічний вектор low_otu_index, який вказує,які OTUs мають суму кількостей менше або рівно 20. Викликається функція filter_obs для відфільтрування OTUs на основі low_otu_index, ! є висволом інверсії. Результат зберігається у змінній taxdata$data$rarefied_otus.\n",
        "\n",
        "```{r}\n",
        "taxdata$data$rarefied_otus <- rarefy_obs(taxdata, \"otu_table\",\n",
        "                                         other_cols = TRUE)\n",
        "low_otu_index <- rowSums(taxdata$data$rarefied_otus[, sample_ids])<=20\n",
        "taxdata <- filter_obs(taxdata, \"rarefied_otus\", ! low_otu_index)\n",
        "taxdata$data$rarefied_otus\n",
        "\n",
        "```\n",
        "\n",
        "# Висновок\n",
        "\n",
        "Метод рарифікації вибірки за допомогою metacoder використовується для зменьшення похибки дослідження спречиненою неоднорідною глибиною сиквенування серед вибірок метагеномних данних.Її основними етапами є завантеження пакетів аналізу і самих даних(з конвертацією у відповідний формати), створення гістограми для оцінки кількості OTU у вибірках та сама рарифікація з видаленням вибірок OTUs з сумою кількостей, меньше за певне число.\n"
      ],
      "metadata": {
        "id": "JLVVhcbqFln3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"metacoder\")\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"biomformat\")\n",
        "\n",
        "library(metacoder)\n",
        "\n",
        "biom_file <- file.path(getwd(), \"datasets\", \"ch5\", \"rich_high_count_otu.biom\")\n",
        "taxdata <- parse_qiime_biom(biom_file)\n",
        "\n",
        "\n",
        "sample_ids <- paste0(\"Sample\", 1:6)\n",
        "taxdata$data$otu_table\n",
        "\n",
        "hist_data <- colSums(taxdata$data$otu_table[, sample_ids])\n",
        "hist(hist_data, prob= TRUE, breaks=3)\n",
        "lines(density(hist_data, adjust = 2), col=\"red\")\n",
        "\n",
        "taxdata$data$rarefied_otus <- rarefy_obs(taxdata, \"otu_table\", other_cols = TRUE)\n",
        "\n",
        "low_otu_index <- rowSums(taxdata$data$rarefied_otus[, sample_ids]) <=20\n",
        "low_otu_index\n",
        "\n",
        "taxdata <- filter_obs(taxdata, \"rarefied_otus\", ! low_otu_index)\n",
        "taxdata$data$rarefied_otus"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "BwFQ-7kGu9eF",
        "outputId": "79987d6c-0cb3-4ce9-a7c0-ea58b8c048f3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'metacoder'”\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'biomformat'\n",
            "\n",
            "also installing the dependencies ‘Rhdf5lib’, ‘rhdf5filters’, ‘rhdf5’\n",
            "\n",
            "\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A tibble: 5 × 8</caption>\n",
              "<thead>\n",
              "\t<tr><th scope=col>taxon_id</th><th scope=col>otu_id</th><th scope=col>Sample1</th><th scope=col>Sample2</th><th scope=col>Sample3</th><th scope=col>Sample4</th><th scope=col>Sample5</th><th scope=col>Sample6</th></tr>\n",
              "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><td>ax</td><td>GG_OTU_1</td><td>  24</td><td>15010</td><td>22426</td><td>20482</td><td>4574</td><td>13424</td></tr>\n",
              "\t<tr><td>ay</td><td>GG_OTU_2</td><td> 872</td><td>    0</td><td>16811</td><td> 5174</td><td>4280</td><td> 8042</td></tr>\n",
              "\t<tr><td>az</td><td>GG_OTU_3</td><td>   0</td><td>    7</td><td>  100</td><td>   50</td><td>  25</td><td>   19</td></tr>\n",
              "\t<tr><td>ba</td><td>GG_OTU_4</td><td> 875</td><td>16654</td><td>31109</td><td> 2410</td><td>   0</td><td>13815</td></tr>\n",
              "\t<tr><td>ax</td><td>GG_OTU_5</td><td>1270</td><td>12857</td><td> 7267</td><td> 3675</td><td>4159</td><td>    0</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA tibble: 5 × 8\n\n| taxon_id &lt;chr&gt; | otu_id &lt;chr&gt; | Sample1 &lt;dbl&gt; | Sample2 &lt;dbl&gt; | Sample3 &lt;dbl&gt; | Sample4 &lt;dbl&gt; | Sample5 &lt;dbl&gt; | Sample6 &lt;dbl&gt; |\n|---|---|---|---|---|---|---|---|\n| ax | GG_OTU_1 |   24 | 15010 | 22426 | 20482 | 4574 | 13424 |\n| ay | GG_OTU_2 |  872 |     0 | 16811 |  5174 | 4280 |  8042 |\n| az | GG_OTU_3 |    0 |     7 |   100 |    50 |   25 |    19 |\n| ba | GG_OTU_4 |  875 | 16654 | 31109 |  2410 |    0 | 13815 |\n| ax | GG_OTU_5 | 1270 | 12857 |  7267 |  3675 | 4159 |     0 |\n\n",
            "text/latex": "A tibble: 5 × 8\n\\begin{tabular}{llllllll}\n taxon\\_id & otu\\_id & Sample1 & Sample2 & Sample3 & Sample4 & Sample5 & Sample6\\\\\n <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n\\hline\n\t ax & GG\\_OTU\\_1 &   24 & 15010 & 22426 & 20482 & 4574 & 13424\\\\\n\t ay & GG\\_OTU\\_2 &  872 &     0 & 16811 &  5174 & 4280 &  8042\\\\\n\t az & GG\\_OTU\\_3 &    0 &     7 &   100 &    50 &   25 &    19\\\\\n\t ba & GG\\_OTU\\_4 &  875 & 16654 & 31109 &  2410 &    0 & 13815\\\\\n\t ax & GG\\_OTU\\_5 & 1270 & 12857 &  7267 &  3675 & 4159 &     0\\\\\n\\end{tabular}\n",
            "text/plain": [
              "  taxon_id otu_id   Sample1 Sample2 Sample3 Sample4 Sample5 Sample6\n",
              "1 ax       GG_OTU_1   24    15010   22426   20482   4574    13424  \n",
              "2 ay       GG_OTU_2  872        0   16811    5174   4280     8042  \n",
              "3 az       GG_OTU_3    0        7     100      50     25       19  \n",
              "4 ba       GG_OTU_4  875    16654   31109    2410      0    13815  \n",
              "5 ax       GG_OTU_5 1270    12857    7267    3675   4159        0  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "No `cols` specified, so using all numeric columns:\n",
            "   Sample1, Sample2, Sample3, Sample4, Sample5, Sample6\n",
            "\n",
            "\n",
            "Rarefying to 3041 since that is the lowest sample total.\n",
            "\n",
            "Warning message in vegan::rrarefy(t(count_table), sample = sample_size):\n",
            "“function should be used for observed counts, but smallest count is 7”\n",
            "Warning message:\n",
            "“The following columns will be replaced in the output:\n",
            "   Sample1, Sample2, Sample3, Sample4, Sample5, Sample6\n",
            "”\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<style>\n",
              ".list-inline {list-style: none; margin:0; padding: 0}\n",
              ".list-inline>li {display: inline-block}\n",
              ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
              "</style>\n",
              "<ol class=list-inline><li>FALSE</li><li>FALSE</li><li>TRUE</li><li>FALSE</li><li>FALSE</li></ol>\n"
            ],
            "text/markdown": "1. FALSE\n2. FALSE\n3. TRUE\n4. FALSE\n5. FALSE\n\n\n",
            "text/latex": "\\begin{enumerate*}\n\\item FALSE\n\\item FALSE\n\\item TRUE\n\\item FALSE\n\\item FALSE\n\\end{enumerate*}\n",
            "text/plain": [
              "[1] FALSE FALSE  TRUE FALSE FALSE"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A tibble: 4 × 8</caption>\n",
              "<thead>\n",
              "\t<tr><th scope=col>taxon_id</th><th scope=col>otu_id</th><th scope=col>Sample1</th><th scope=col>Sample2</th><th scope=col>Sample3</th><th scope=col>Sample4</th><th scope=col>Sample5</th><th scope=col>Sample6</th></tr>\n",
              "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><td>ax</td><td>GG_OTU_1</td><td>  24</td><td>1015</td><td> 847</td><td>1936</td><td>1088</td><td>1172</td></tr>\n",
              "\t<tr><td>ay</td><td>GG_OTU_2</td><td> 872</td><td>   0</td><td> 649</td><td> 495</td><td> 946</td><td> 670</td></tr>\n",
              "\t<tr><td>ba</td><td>GG_OTU_4</td><td> 875</td><td>1136</td><td>1236</td><td> 220</td><td>   0</td><td>1198</td></tr>\n",
              "\t<tr><td>ax</td><td>GG_OTU_5</td><td>1270</td><td> 889</td><td> 306</td><td> 385</td><td>1002</td><td>   0</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA tibble: 4 × 8\n\n| taxon_id &lt;chr&gt; | otu_id &lt;chr&gt; | Sample1 &lt;dbl&gt; | Sample2 &lt;dbl&gt; | Sample3 &lt;dbl&gt; | Sample4 &lt;dbl&gt; | Sample5 &lt;dbl&gt; | Sample6 &lt;dbl&gt; |\n|---|---|---|---|---|---|---|---|\n| ax | GG_OTU_1 |   24 | 1015 |  847 | 1936 | 1088 | 1172 |\n| ay | GG_OTU_2 |  872 |    0 |  649 |  495 |  946 |  670 |\n| ba | GG_OTU_4 |  875 | 1136 | 1236 |  220 |    0 | 1198 |\n| ax | GG_OTU_5 | 1270 |  889 |  306 |  385 | 1002 |    0 |\n\n",
            "text/latex": "A tibble: 4 × 8\n\\begin{tabular}{llllllll}\n taxon\\_id & otu\\_id & Sample1 & Sample2 & Sample3 & Sample4 & Sample5 & Sample6\\\\\n <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n\\hline\n\t ax & GG\\_OTU\\_1 &   24 & 1015 &  847 & 1936 & 1088 & 1172\\\\\n\t ay & GG\\_OTU\\_2 &  872 &    0 &  649 &  495 &  946 &  670\\\\\n\t ba & GG\\_OTU\\_4 &  875 & 1136 & 1236 &  220 &    0 & 1198\\\\\n\t ax & GG\\_OTU\\_5 & 1270 &  889 &  306 &  385 & 1002 &    0\\\\\n\\end{tabular}\n",
            "text/plain": [
              "  taxon_id otu_id   Sample1 Sample2 Sample3 Sample4 Sample5 Sample6\n",
              "1 ax       GG_OTU_1   24    1015     847    1936    1088    1172   \n",
              "2 ay       GG_OTU_2  872       0     649     495     946     670   \n",
              "3 ba       GG_OTU_4  875    1136    1236     220       0    1198   \n",
              "4 ax       GG_OTU_5 1270     889     306     385    1002       0   "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Plot with title “Histogram of hist_data”"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAACXBIWXMAABJ0AAASdAHeZh94\nAAAgAElEQVR4nOzde1zO9+P/8dfVWedQKjooOXyWLeQwzWFmGk3O5zm1ImE0zA5yZvbFZycy\nG2PMKYcPNvPhw4y2sYTMcVNCoiSVzsfr98f1/Vy/vqRC9ep697j/sdv7er3f17vn1XvV0/uo\nUqvVAgAAALpPT3YAAAAAVA2KHQAAgEJQ7AAAABSCYgcAAKAQFDsAAACFoNgBAAAoBMUOAABA\nISh2AAAACkGxAwAAUAiKHQAAgEJQ7AAAABSCYgcAAKAQFDsAAACFoNgBAAAoBMUOAABAISh2\nAAAACkGxAwAAUAiKHQAAgEJQ7AAAABSCYgcAAKAQFDsAAACFoNgBAAAoBMUOAABAISh2AAAA\nCkGxAwAAUAiKHQAAgEJQ7AAAABSCYgcAAKAQFDsAAACFoNgBAAAoBMUOAABAISh2AAAACkGx\nAwAAUAiKHQAAgEJQ7AAAABSCYgcAAKAQFDsAAACFoNgBAAAoBMUOAABAISh2gMLt2rVLpVKp\nVCoTExPZWeqQnTt3tmvXrl69eqampu7u7rm5uWUu9lRbpxZuyr1792oiGRgYyM4CQAiKHaBb\n1q1bp/qv9PT0xxfo2bOnZu4bb7xR8/Gg8ccffwwdOvTs2bN5eXm5ubnXr18vLi6WHQpAncC/\nsQCF8/Ly+vTTT4UQz7ZPJTk5uXHjxsXFxVeuXGnZsmVVp1OmDRs2aCasrKzmz5/foEEDY2Pj\n51+t8jZlLYwE6DqKHaBwzZo1mz59+jO/PSIigr1NT+vWrVuaieHDhz/PN/8RytuUtTASoOs4\nFAugPDt27JAdQfcUFBRoJszNzeUmKa0WbspaGAnQdRQ7QOGedMZ9UVHR119/3bNnTzs7O0ND\nQzs7u/bt23/88cf379/XLPDmm2+qVKrffvtN87JVq1Yqlar0HqOHDx8uXbq0U6dO9evXNzIy\natSoka+v74YNG8rcB7NhwwZvb28zM7MGDRq8+eab0dHR8fHx2vMF8/LyNIt9++23mpFu3boV\nFRVNmzbN1ta2UaNGmrlqtXr79u2+vr6azJaWlh07dvzyyy9Lf0XtGrp27ar5+G3atDE1NXV1\ndZ0zZ05hYaEQ4sqVK/7+/jY2Nubm5r169bp06VJlvpOV+bzBwcEqlero0aOalytXrtSEycrK\nKn/lenp6QojIyMjXX39dE6xLly5HjhwpvUz1bcpK2rhxo2Yj1q9f38/PLzo6WqVSlblkhVuq\n/EiV2dAAyqYGoDu++eYb7Q9vWlra4wu89tprmrm+vr6akZ07d2pGjI2NtYsVFBR07969zN8J\nbm5usbGxarXaz8/v8bnTpk3TrCEmJqZx48ZlruHll1++f/9+6VShoaGPLGNsbLx69erHfxFt\n3bpVM/Liiy+uXLlSM62vr6+ZO2rUqDK/4ptvvllSUqJZZvv27ZpBT0/PiIiIR5rH5MmTr1+/\n3rBhw9KDtra26enp5X/nK/l5J06cWOYymZmZZa5Wu3VsbGwOHTpkZGRU+l36+vqHDx9+fOGq\n3ZSVNGPGjEfWYGxsPGfOHG3U0gtXuKXKj1SZDQ2gTBQ7QJdUVbFbs2aNZrBly5bbtm37/fff\n//3vfw8aNEgz2K1bN7VafenSpX379mm/3JYtWyIjI69fv65Wqx88eKBtOU2bNl2zZs3evXtn\nz56tPanfz89P+7WioqK0K/Hy8vrmm282b97cuXNn7WHK0p1Am9bV1dXJycnQ0NDLy6tFixZq\ntfqHH37QzNLT0wsPD79w4cL69eu1XzEiIuKRNTg4ODRp0sTf33/KlClWVlbab0Lv3r0dHR2n\nTZvWsWNHbbBPP/20nG975T9vbGxsZGSkt7e3Znz48OGRkZGRkZHFxcVlrlmbtmHDhq6url5e\nXh988EGvXr20wdq3b//4wlW4KSvpjz/+0L69R48ee/bs2b9//xtvvKH9DpTeiJXZUuVEquSG\nBlAmih2gS0oXu/KVX+zGjRunGVy5cqV2sKCgYMSIEVOmTFm2bJmmiNy9e1e7witXrmiXXLBg\ngWbQ0tIyMTFRO75582bt8tHR0ZrBCRMmaEasra21e7ZycnJcXFwe7wTatEIIDw+PhIQE7axV\nq1b5+fn5+fmV3tXk7++vWXjMmDGPr2HIkCGawd27d2sHTUxM4uLi1Gp1fn5+ixYtNIN9+/Yt\n59v+VJ9XXapez5gxo5zVPpK2a9eueXl5mnHtLis9Pb2CgoJHFq7CTVlJQUFBmvfa2tpmZWVp\nBgsLC7WXspbeiJXcUk+KVMm3AygTV8UCdZGlpaVm4vPPP7e1te3Vq1ejRo0MDQ21R0LLt2fP\nHs3EwIEDHR0dteMjRox455130tLShBA//fRTu3bthBC//vqrZm7fvn0bNGigma5Xr96YMWMW\nLVpUzldZsGBBkyZNtC8nT548efLkR5bRtsOkpKTH1zBr1iztlzY2Ns7PzxdCDBgwwM3NTQhh\nZGTUt2/fv/76SwiRkJBQVZ/3mc2fP197V5Tx48dv2bJFCFFSUpKYmOjq6vqkdz3npqyk0hvR\nzMxMM21gYDBq1KiwsLBHFn6GLVWFbwfqOC6eAHSVk5OTy2Mq+UyC8ePHm5qaCiFu3bo1ZswY\ne3v7Zs2aBQYG7tmzp8Lz09VqtfZqg9atW5eepa+vr92Fc+XKFc3EjRs3NBOtWrUqvfBLL71U\n/hfq0aPHIyNHjhzp37+/u7u7iYmJ5jKCL7/8UjOrzNienp6aCUNDQ1tbW830iy++qF3AwcFB\nM5GZmfmkGE/7eZ+Z9gCuKNVjhBDlX3jxPJuy8rQb0cPDo/T4Cy+8UObyT7ulqvbtQF3GHjtA\nV/3555/W1taPDPbs2VN7SWY5vLy8Dhw4MGXKFG1liYuLi4uLW79+vaur69atW19++eUnvTcn\nJ6eoqEgz/fjtPLS7czRVSa1W5+TkPDLrSe8tTV9fX1vFNNasWRMSEqJdlZubm5GRUWJiovba\nz0cYGxvXq1ev9EvNROlv2iMXK5TpqT7vMzM2NrawsNC+rPxDw55nU1aSWq3WPhJNUyK1Hnmp\n8bRbqmrfDtRx7LED6qju3btfuHDh999/X7hw4RtvvKG9vODGjRv9+vXLzs5+0htNTU21Z7I/\nfPjwkbnafqNZoUql0jYqbcN7ZMkyGRgYaO4AopGVlTVz5kzN9MiRI1NSUi5fvhwTEzN06NCK\nPujzeqrPK8Uzb8pKKn2DlUfW9vg35Dm3lMQNDSgDxQ6ou1Qq1csvvxwWFnbw4MHU1NR9+/bZ\n2NgIIVJSUk6cOFHOu7SHOGNiYkrPKiwsvHz5smZae8TTyclJM6GdpXH+/PnKRz137py2F86a\nNUu7K05zhly1etrPK8WzbcrKc3Z21kxcvXq19Pi5c+ceWfI5t5TEDQ0oA8UOqHNyc3M//vjj\n8ePH9+vXr6SkRDOor6/v7+/fvn17zUvNdQalbwJX+kDYwIEDNRN79+69ffu2dnzjxo2aPVgq\nlap///6awU6dOmkm9u/fr11JVlbWd999V/nMmjwampsMCyEuX7587NixRwarw1N93pr0/Juy\nkrTHc/fv36+5WEQIkZWVVfq64NJfTqP8LVVmJLkbGlACuRflAngqVXUfOy8vL83goEGDfvrp\np+jo6BMnTixYsMDQ0FCzZHJyslqtLioq0owIIbp06RIREXHo0CG1Wp2Wlqa9XrVZs2arV6/e\nvXv3rFmztEddAwMDtV+r9Dl/L7744saNG9evX9++fXvt2Wll3u6kdFq1Wp2YmKg9Mtu3b98L\nFy7s37+/cePG2vuVWFpanjx5Mjk5+UlrcHd314yvWbNGO6g9Jd/d3b2cb/tTfV71M93u5JG0\npa/SvXDhQvVtykr65ZdftHnatWu3ffv27777rn379tqTDvX09DRLVn5LlRmp8m+vfHigTqHY\nAbqkqordhQsXSt9JpDQ9Pb3169drl+zdu3fpudo78ZbzJIaBAwfm5uaWTjV27NhHljE1Nf3k\nk08005Updmq1eurUqY+sxNHRMT4+vvT9R+bNm1cdxe5pP29NFrvn35SVNH78+EfWb2ZmtmrV\nKu1L7U2YK7mlnhSp8m8H8DgOxQJ1kaenZ3R09KJFi7y9vTW3PTM1NW3ZsuWECRPOnTsXEBCg\nXfKbb77p37+/tbW1iYlJ06ZNO3furBl/6aWXLl++vHjx4vbt21tZWRkaGjo4OAwYMGD//v27\nd+9+5KLOb7/9dvny5S1atDA2Nrazsxs8ePCpU6e0tw7RXppQvpUrV37yySf/+Mc/6tWr17hx\n48DAwNOnT7u6un733XctWrQwMDBo0qTJI3dUqUJP9Xlr0vNvykpat27dypUrW7RoYWRkpNmI\nf/zxx6uvvqpdQHtdReW3VJmR5G5oQNep1Gq17AwA6qJNmzZp9uQ5OjomJibKjgMASsB97ABU\nrytXruzbty8hIeHBgwebN2/W7p/TPuarQ4cO8tIBgKJQ7ABUL0NDww8//FBzcCAvL2/69OkG\nBga7du3av3+/ZgHt3WhRk/744w8/P7/KLBkcHLx48eLqzgOgSnAoFkC1W7Bgwfz588ucFRYW\ntnDhwpqNAwCKRbEDUBN+/vnnNWvWnDp1Kjk5WU9Pz97evlOnTpMmTerWrZvsaACgHBQ7AAAA\nheB2JwAAAApBsQMAAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDYAQAAKATFDgAAQCEo\ndgAAAApBsQMAAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDYAQAAKATFDgAAQCEodgAA\nAApBsQMAAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDYAQAAKATFDgAAQCEodgAAAApB\nsQMAAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDYAQAAKATFDgAAQCEodgAAAApBsQMA\nAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDYAQAAKATFDgAAQCEodgAAAApBsQMAAFAI\nih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDYAQAAKATFDgAAQCEodgAAAApBsQMAAFAIih0A\nAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDYAQAAKATFDgAAQCEMZAeoc+7evXvp0iXZKVBdHjx4\nIISoX7++7CCoLi+88IKDg4PsFABQNpVarZadoW4JDAzctGlTvXr1ZAdBtcjKytLX12f7KlVu\nbu6YMWPWrVsnOwgAlI09djWtuLi4T58+ixcvlh0E1cLPz69NmzZsX6WaM2dOcXGx7BQA8ESc\nYwcAAKAQFDsAAACFoNgBAAAoBMUOAABAISh2AAAACkGxAwAAUAiKHQAAgEJQ7AAAABSCYgcA\nAKAQFDsAAACFoNgBAAAoBMUOAABAISh2AAAACkGxAwAAUAiKHQAAgEJQ7AAAABSCYgcAAKAQ\nFDsAAACFoNgBAAAoBMUOAABAISh2AAAACkGxAwAAUAgD2QGeS0FBwfnz57OyslxdXZs2bSo7\nDgAAgEw6s8du8eLFx44dKz2ydu1ae3v7Dh069OjRw83NzdvbOyYmRlY8AAAA6XSm2IWFhR06\ndEj78sCBA8HBwTk5OQMGDJg4caKPj8+ZM2e6d+8eFxcnMSQAAIBEunooNjQ01MrK6uTJk61a\ntdKM7NmzZ/DgwUuWLPn222/lZgMAAJBCZ/bYlZaSknLt2rXJkydrW50QYuDAgf369Tt8+LDE\nYAAAABLpZLHLy8sTQpRudRqenp737t2TkQgAAEA+nSx2jo6OVlZWt2/ffmT8zp07FhYWUiIB\nAABIp0vF7tatW9HR0bGxsWlpaSEhIevXr8/JydHOvXr16o4dO3x8fCQmBAAAkEiXLp7Ytm3b\ntm3bSo8cPHhw0KBBQoitW7dOmDAhNzc3LCxMUjoAAADJdKbYbdiwIb2UjIyM9PR0Gxsbzdz0\n9HRra+vt27e3b99ebk4AAABZdKbYjRs3rpy5Y8aMCQ4O1tPTpSPLAAAAVUtnip2WWq2Oj4+/\nfv16ZmamEMLKysrDw8PJyUl2LgAAAMl0qdilpaUtWbJk8+bNj9/TxNnZOTAwcObMmfXq1ZOS\nDQAAQDqdKXZ379718fGJj4/38PDo06ePi4uLmZmZEOLhw4dxcXHHjx+fO3fu7t27jx07pj3x\nDgAAoE7RmWIXFhZ2+/btiIiIIUOGPD63uLh47dq1U6ZMWbBgwWeffVbz8QAAAKTTmasNDhw4\nMHr06DJbnRBCX18/JCRk6NChe/bsqeFgAAAAtYTOFLvU1FR3d/fyl2nVqlVycnLN5AEAAKht\ndKbYOTo6nj9/vvxlzp075+joWDN5AAAAahudKXb9+/ffuXPnihUr8vPzH5+bnZ09b968ffv2\nDRs2rOazAQAA1AY6c/HE/PnzIyMjZ82atXDhwg4dOjg5OZmbm6vV6qysrJs3b0ZFReXk5HTp\n0mXOnDmykwIAAMihM8XO2tr65MmTq1ev3rRp0y+//FJcXKydZWho2K5du4CAgICAAH19fYkh\nAQAAJNKZYieEMDIyCg0NDQ0NzcvLS0hI0Dx5wtLS0tnZ2cjISHY6AAAAyXSp2GmZmJh4eHjI\nTgEAAFC76MzFE2VasWLFK6+8IjsFAABAraDbxS42Nva3336TnQIAAKBW0O1iBwAAAC2KHQAA\ngEJQ7AAAABRCt4vdsmXLEhISZKcAAACoFXS42GVmZiYmJpqbm8sOAgAAUCvoZLE7fvy4t7e3\npaWlp6fnqVOnNIP+/v5Hjx6VGwwAAEAi3St2UVFRvXr1+vvvv319fbWDKSkpp0+f7tOnz5kz\nZyRmAwAAkEj3it3ChQvt7e0vX768ceNG7aCtre358+ft7e0XLVokLxoAAIBMuvdIsVOnTs2c\nObNJkyZJSUmlx+3s7IKDg5cvX/5Ua4uPj+/YsWNRUVE5y5SUlOTm5ubk5Ojr6z9LYgAAgBqh\ne8UuIyPDycmpzFkODg5ZWVlPtTYXF5eIiIjyi92lS5emT59eXFxMsQMAALWZ7hU7e3v7K1eu\nlDnrxIkTjo6OT7U2PT297t27l7+MqanpU60TAABACt07x65Pnz7h4eFnz54tPZiWlvbRRx9t\n2LDBz89PVjAAAAC5dK/YLViwwNzcvGPHjpoO98EHH7Rp08bBwWHp0qXOzs5z586VHRAAAEAO\n3St29vb20dHRQUFBN2/eFELExMTExMRYWFhMmjTp9OnTjRo1kh0QAABADt07x04IYWdnFx4e\nvnr16nv37mVmZlpYWNDnAAAAdG+PnZZKpbKxsUlLS7t8+XJ8fLzsOAAAAJLpTLFbvHjxsWPH\nSo+sXbvW3t6+Q4cOPXr0cHNz8/b2jomJkRUPAABAOp0pdmFhYYcOHdK+PHDgQHBwcE5OzoAB\nAyZOnOjj43PmzJnu3bvHxcVJDAkAACCRTp5jJ4QIDQ21srI6efJkq1atNCN79uwZPHjwkiVL\nvv32W7nZAAAApNCZPXalpaSkXLt2bfLkydpWJ4QYOHBgv379Dh8+LDEYAACARDpZ7PLy8oQQ\npVudhqen571792QkAgAAkE8ni52jo6OVldXt27cfGb9z546FhYWUSAAAANLpUrG7detWdHR0\nbGxsWlpaSEjI+vXrc3JytHOvXr26Y8cOHx8fiQkBAAAk0qWLJ7Zt27Zt27bSIwcPHhw0aJAQ\nYuvWrRMmTMjNzQ0LC5OUDgAAQDKdKXYbNmxILyUjIyM9Pd3GxkYzNz093draevv27e3bt5eb\nEwAAQBadKXbjxo0rZ+6YMWOCg4P19HTpyDIAAEDV0pliVz5zc3PZEQAAACRjFxcAAIBCKKfY\nxcXF9ezZs2fPnrKDAAAAyKGQQ7FCiMzMzKNHj8pOAQAAII1yil3Lli0vXLggOwUAAIA0yil2\nJiYmnp6eslMAAABIo5xz7IQQqampsbGxslMAAADIoahit3z5cg8PD9kpAAAA5FBUsQMAAKjL\nKHYAAAAKoTMXT3h7e1e4TGJiYg0kAQAAqJ10ptidO3dOCGFoaFjOMkVFRTUVBwAAoNbRmUOx\ns2bNMjMzu3jxYt6TzZw5U3ZMAAAAaXSm2C1atKhZs2YjRowoLCyUnQUAAKA20pliZ2houGXL\nlkuXLn344YeyswAAANRGOnOOnRCiVatWSUlJ5ZxI17t3b2tr65qMBAAAUHvoUrETQlhaWpYz\nt1u3bt26dauxMAAAALWKzhyKBQAAQPkodgAAAApBsQMAAFAIih0AAIBCUOwAAAAUgmIHAACg\nEBQ7AAAAhaDYAQAAKATFDgAAQCEodgAAAApBsQMAAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7\nAAAAhaDYAQAAKATFDgAAQCEodgAAAApBsQMAAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAA\nhaDYAQAAKATFDgAAQCEodgAAAApBsQMAAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDY\nAQAAKATFDgAAQCEodgAAAApBsQMAAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDYAQAA\nKATFDgAAQCEodgAAAApBsQMAAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDYAQAAKATF\nDgAAQCEodgAAAApBsQMAAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDYAQAAKATFDgAA\nQCEodgAAAApBsQMAAFAIih0AAIBCUOwAAAAUgmIHAACgEBQ7AAAAhaDYAQAAKATFDgAAQCEM\nZAcAUBvpZ2WpCgv1srM1L1VFRWqDUr8u9PSKzc3/978AgFqDYgfULaqCAqPERIOUFMOUFP3U\nVMOUFIO0NP0HDwzS0vQfPtTLytLLztbLz6/8CkvMzEpMTYstLIotLYstLIrq1y+ysytq0KDI\n1rbQ1raoUaPChg3VRkbV94kAAFoUO0CxVPn5JtevG8fHGyUkGCUkGCYkGN2+bZiSItRqtZ5e\nccOGhQ0aFNnaFtvY5Ddrlm1jU2JhUWxhUWJmVmJsXGJmVmJioilk2gnNOjW1TzOhl5Wll5Oj\nn5Oj9/Ch/sOH+pmZBg8e1Lt0ySAlxeD+fYO0NCGE0NMrbNSowNk538mpwNm5wMUl39m5wMVF\nbWgo73sDAMpEsQOUwzApyeSvv0z+/lvzX6Nbt1TFxUX16xc4ORU4OWV37Jg2aFBhkyYFTZoU\nNmwo9Kr9FFtVQYHh/fsGSUnGCQlGt24Z3bpldfCgcUKCXlaWWl8/v2nT/ObN81q0yG3RIq9l\ny6IGDao7DwAoHsUO0GF6ubn1Ll40PX++3p9/ml68aJCSUmJiku/hkdeixYNhw/KaNctr0aLY\n2lpWPLWRUYGjY4GjY07btqXHDVJTjWNjTa5dM/n7b8tDh+xWr1YVFBTZ2uZ6eua0bp3z0ku5\nnp4lpqayYgOA7qLYATpG/+FDszNnzKKiTM+eNfnrLyFEXvPmuS+9lPTaa7kvvljg4qKu/l1x\nz6moQYOiBg2yO3bUvFQVFRlfv25y9arpxYuWR47YrV4thMhr3jynbducNm2y27cvql9fal4A\n0BkUO0AHqAoKTGNizH//3fzkSZOrV9XGxjlt2z587bWkGTNyW7cuqVdPdsDnojYwyGvePK95\n83R/fyGEXl5evYsXTc+eNT13znrfPv3s7Dx39+yOHbM6dcru0IE9eQBQDoodUHsZJidbnDhh\nceKE2R9/qAoKcl96KfPVV+/Onp374ov/5+YjylJiYpLt7Z3t7S2EUJWUmFy+bBYVZR4VZbN7\nt6qoKMfLK8vHJ7NLl7wWLWQnBYBaR7F/GwDdZfLXX5Y//2z5888mV68W2tlldu16+5NPstq3\nL6l7N41T6+nlenrmenreDwhQ5eebnTtn/vvvVgcONPr888JGjTK7dn3Yo0d2x45cYAsAGhQ7\noLaod+mS1eHDlocPG92+ndey5cMePW4vXJjXsqVQqWRHqxXUxsZZnTpldeok3n3X8M4di8hI\ny19+cZk2rcTIKKtr14yePbO6dCkxMZEdEwBkotgBkhnHxVn/+KPVwYNGd+7ktm79YPjwhz17\nFjRuLDtXrVbo6Phg2LAHw4bpZWVZnDhhefRok48+EkJkdumS4eub2a2b2thYdkYAkIBiB8hh\nkJpqdeCAzY8/mly5kteq1YMRIzJ69Sp0cJCdS8eUmJtn9OmT0aePXn6+eWSk1X/+0yQsTK2n\n97Bnz/Q338xu374GbtcHALUHxQ6oUaqiIotjx2z27TP/9dciO7t0P7+Ejz/Od3eXnUvnlRgb\nP+zZ82HPnnq5uZbHjln9+KPrxIlFDRum+/ml9+2b36yZ7IAAUBModkANMbp922bXLpu9e/Vy\nch727Hlz7dos9idVg5J69dL79Env08fgwQOrgwetf/jB9ttvc1u3fjBwYEbv3iVmZrIDAkA1\notgB1UtVXGxx/Hj9HTvMT53K8/C4Fxyc/uabdfD61ppXVL9+6qhRqaNGGcfF2ezebf/FFw4r\nVmT07v1g8ODcF16QnQ4AqgXFDqguBmlpNhER9XfuNEhPz3jjjeubN+e8+KLsUHVRvrt70nvv\nJU+fbnn0aP1du9xHjMht2TJt6NB0Pz9dv7czADyCYgdUPePr1xts3mz9ww9FtrapY8ak9etX\nbGUlO1RdpzYyyujdO6N3b+ObN2127Wr0xReNPv00bdCg1OHDCx0dZacDgKpBsQOqUue8vJmn\nTnns35/Tps3tZcsye/So/U9urWvyXVySZsxInjLF+sCBBlu2NPjuu8zXXkt9663stm1lRwOA\n50WxA6qAqqTE8tAh2w0bvr9371TjxnFbtuS2bi07FMqjNjZOGzgwbeBAs6ioBt9/33T8+NyW\nLVPHjs3w9VXr68tOBwDPiGIHPBdVUZH1jz82XLfOMDk5bdCgfhkZDb29F9PqdEd2hw7ZHToY\nJSQ02LLFccECu88/Tx03Lm3AAB5iAUAXcZAIeEaqwsL6O3d6+Pk5fPLJw9df/1krx+cAACAA\nSURBVPvQobvvv59gwD+WdFKBk9Pd99//6/DhtIEDbb/6qkWvXnZr1+pnZMjOBQBPhz9CwFNT\nFRTU37274fr1erm5qW+9lTpqVLGlpexQqALFVlYpEyemjhtnvWdPw40bG27Y8GDIkPtjxxY1\nbCg7GgBUCsUOeAqq4mLrvXvt1q7Vy829P2ZM6ogR3JFOeUqMjR+MGJE2ZIjVwYO269fX37Yt\nbciQlPHji+zsZEcDgApQ7IDKKSmx+ve/G61ebfDgwf0xY+6PHk2lUza1gUF6377pfn5WR47Y\nfvVV/Z07HwwaZJOXJzsXAJSHYgdUzCIystEXXxjduJE6cuT9gABuSleH6Oll9OqV8frrlkeO\n2K1Zs/b69V9athTJyaJRI9nJAKAMXDwBlKfexYtNx493fuednBdf/PvAgeTQUFpdXaRSPXz9\n9dhduz5v08YzMVG4u4sPPxRpabJjAcCjKHZA2Yzu3HGaPdt91Kji+vWv7d17JyyMU6zqOj29\n3xo3ntOvn1izRuzYIdzdxccfi5wc2bEA4P+j2AGP0svKavTZZx7+/oaJidc3bbq1cmWBi4vs\nUKgt1CqVGD1aXLkiFi8WX34pPDzE11+LoiLZuQBACIodUJqqpMRm9+7mb75pdfDg7SVLrm/e\nnPPSS7JDoVYyMhIhIeLaNRESImbNEq1bi717ZWcCAIod8F+m5865Dx/usHx56ltvXdu/P8PX\nV6hUskOhdjMzEx99JOLihK+vGDZMdOsmoqJkZwJQp1HsAGFw716TDz5wGzcur1mzv3/4ISUw\nUG1sLDsUdEfDhuKzz8Tly6JRI/Hyy2LkSHHrluxMAOooih3qNFVxcYNNm5r7+xtfv379u+9u\nL11aZGsrOxR0k7u7iIgQv/4q4uNFy5YiLExkZ8vOBKDOodih7jI9d8596FC7tWuTQkPjtm3L\n8fKSnQi67+WXxe+/i2++ERs3ihYtxJYtQq2WnQlAHUKxQ12kn5HReN48t3Hj8lq1uvbDDw+G\nDRN6/CygiqhUYtQocfWqePttERQkXnlFnDkjOxOAuoI/Zqhj1Grrffs8+vY1PX8+/ttvby9e\nXFS/vuxMUCIzM7FggbhyRTRuLDp0EEFB4v592ZkAKB/FDnWI0e3brkFBjosWpY4eHbtzZ3a7\ndrITQelcXEREhDh6VJw8KVq0EF99JUpKZGcCoGQUO9QJqpKSht9912zAAJUQsf/6V0pQkNrQ\nUHYo1Bndu4tz58SHH4r33hMdO4rTp2UHAqBYFDson8m1a26jRtmuXXv3gw/iv/mmwMlJdiLU\nPYaGYsYMceWKcHcXnTqJkBAeNQugOlDsoGSqwkK78HD3YcMK7eyu7duXNnAg9xyGTI0bi+3b\nxaFD4uhR0bKl+P572YEAKI2B7ADPpaCg4Pz581lZWa6urk2bNpUdB7VLvStXGn/0kUFq6u2P\nP87w9ZUdB/ivnj3Fn3+K5ctFUJDYtEmsWSPc3WVnAqAQOrPHbvHixceOHSs9snbtWnt7+w4d\nOvTo0cPNzc3b2zsmJkZWPNQqquJiu6++chs5Mr9Zs2v79tHqUOsYG4s5c8Sff4riYtG6tVi2\nTBQWys4EQAl0ptiFhYUdOnRI+/LAgQPBwcE5OTkDBgyYOHGij4/PmTNnunfvHhcXJzEkagPj\nuDi3kSMbbN16+5NPEv7nf4qtrWUnAp7Aw0McOSLWrBErVghvb54zC+D56Uyxe0RoaKiVldW5\nc+f27Nnz1Vdf/frrr7t373748OGSJUtkR4M8mktfhw0ratTo2r/+ldGrl+xAQEVUKjF2rLh8\nWbRuLTp3FtOn8yAyAM9DJ4tdSkrKtWvXJk+e3KpVK+3gwIED+/Xrd/jwYYnBIJFRYmLTgADb\nr766ExZ284sviho0kJ0IqDQ7O/H99+LHH8XevaJ1a3HkiOxAAHSVTha7vLw8IUTpVqfh6el5\n7949GYkgmfX+/c0GDxYGBrF79qT16yc7DvBM3nhDXLwo+vQRvr4iKEhkZMgOBED36GSxc3R0\ntLKyun379iPjd+7csbCwkBIJsuhnZjq9917jBQvuTZwY/803hQ4OshMBz8HcXKxaJY4dE8eP\nC09P8dNPsgMB0DG6VOxu3boVHR0dGxublpYWEhKyfv36nJwc7dyrV6/u2LHDx8dHYkLUMNOz\nZ5sNHmzy119x339/f9w47lEHhejaVZw/L4YOFf7+YuxYbmUMoPJ0qdht27atffv2Hh4etra2\nH3/8cWxs7MGDBzWztm7d6u3tnZubGxYWJjckaoaquNhu9eqmAQGZPj5xO3bkPXZcHtBt9eqJ\nlStFZKQ4fZpddwAqT2duULxhw4b0UjIyMtLT021sbDRz09PTra2tt2/f3r59e7k5UQOM7txp\nMnu28Y0bCf/858MePWTHAarNyy+Ls2dFWJjw9xfjx4uVK4WlpexMAGo1nSl248aNK2fumDFj\ngoOD9fR0aQckno3l0aON587Na9kydvfuQjs72XGAamZiIpYvF/37i3HjxIsvig0bxKuvys4E\noPbSmWKnpVar4+Pjr1+/npmZKYSwsrLy8PBw4rHudYCqoMB+xYr6O3akBAenTJyopsej7vDx\nEefPi9mzxeuvi8mTxbJlol492ZkA1Ea6VOzS0tKWLFmyefPmx+9p4uzsHBgYOHPmzHr8slMo\no4QEpxkzDFNTb6xbl80Bd9RBpqbiyy9Fv34iIEAcOiQ2bxb8IAB4jM4Uu7t37/r4+MTHx3t4\nePTp08fFxcXMzEwI8fDhw7i4uOPHj8+dO3f37t3Hjh3TnngHxbA6fNhx3rzcl166uXZtEdsX\ndVnPnuLCBfHOO6JzZxEWJj78UBjozK9xADVAZ34jhIWF3b59OyIiYsiQIY/PLS4uXrt27ZQp\nUxYsWPDZZ5/VfDxUE1VBgcOKFTYREfdCQlICAwWHXwErK/Hdd8LfX0ycKH76SWzaJJo3l50J\nQG2hM38mDxw4MHr06DJbnRBCX18/JCRk6NChe/bsqeFgqD6Gd+64jR1reeTIjXXrUiZMoNUB\n/9+gQeLCBWFjI9q2FV9/LTsNgNpCZ/5Spqamuru7l79Mq1atkpOTayYPqpvZH3+4jxhRYmwc\nu2NHtre37DhA7ePgIH76SfzznyI0VPTuLZKSZAcCIJ/OFDtHR8fz58+Xv8y5c+ccHR1rJg+q\nUUmJXXi464QJaf3731i/vsjWVnYgoLZSqcSECSIqSiQlibZtxeHDsgMBkExnil3//v137ty5\nYsWK/Pz8x+dmZ2fPmzdv3759w4YNq/lsqEL6GRkukyc32Lw54bPPkkND1fr6shMBtd4LL4hT\np8TIkaJPHzFjhijrlySAOkJnLp6YP39+ZGTkrFmzFi5c2KFDBycnJ3Nzc7VanZWVdfPmzaio\nqJycnC5dusyZM0d2Ujw7kytXnENDS0xN47ZvL3BxkR0H0B3GxmLFCuHrK8aOFT//LLZtEy1b\nys4EQAKdKXbW1tYnT55cvXr1pk2bfvnll+LiYu0sQ0PDdu3aBQQEBAQE6LODR2dZ//ij4/z5\nD1977c78+SXcjxB4Bq+/Ls6fF2+/Lby9xeefi7fflh0IQE3TmWInhDAyMgoNDQ0NDc3Ly0tI\nSNA8ecLS0tLZ2dnIyEh2Ojw7VXGx/cqV9bduTXr33dQxY2THAXSZra3Yt0+sWiWmTBGHD4uv\nvxZWVrIzAag5ulTstExMTDw8PGSnQNUwSEtzmjHDODb2xtdfZ3foIDsOoPtUKjF1qujaVYwY\nIby8xLZtolMn2ZkA1BCduXiiTCtWrHjllVdkp8CzM7l61W34cL2srLgdO2h1QFV66SURHS1e\ne0106SKWLRMlJbIDAagJul3sYmNjf/vtN9kp8IysDh1yGzMm18srfvPmQgcH2XEAxTE1FevW\nie+/F8uWCV9fwW0+gTpAt4sddFVJSaMvvmgye/a94OCEZctKjI1lBwKUa9gwcfasyMgQXl7i\n6FHZaQBUL4odappedrbz9On1t2279eWX9wMChEolOxGgdG5u4tdfxfDhwtdXzJsnSt1VAIDC\nUOxQo4wSE93eesvk+vXrW7ZkdukiOw5QZxgZiU8/Fbt3i1WrRM+e4u5d2YEAVAvdLnbLli1L\nSEiQnQKVZXb2rPuIEUUNG8Zt2ZLv5iY7DlD39Osnzp4VeXnCy0v85z+y0wCoejpc7DIzMxMT\nE83NzWUHQaXY7NvnGhiY4et786uvirmxFiCLi4s4cUKMHi169xZz53JYFlAYnSx2x48f9/b2\ntrS09PT0PHXqlGbQ39//KOcF104lJY0+/dRx3rykWbPufPQRj38FJDM0FCtWiH/9S6xaJXx9\nxb17sgMBqDK6V+yioqJ69er1999/+/r6agdTUlJOnz7dp0+fM2fOSMyGx+nl5TnPmFF/586b\n4eGpI0bIjgPgv/r2/d+rZdu0EZGRstMAqBq6V+wWLlxob29/+fLljRs3agdtbW3Pnz9vb2+/\naNEiedHwKIN795qOHWty9er177/P6txZdhwA/5erq/j1VzFggOjRQ/zP/wi1WnYgAM9L9x4p\ndurUqZkzZzZp0iQpKan0uJ2dXXBw8PLly59qbfHx8R07diwqKipnGc1cNb/ynpLJ1asuU6YU\nOjpe37q1yMZGdhwAZTE2FqtWiVdeERMmiN9+Exs3Cn5aAV2me8UuIyPDycmpzFkODg5ZWVlP\ntTYXF5eIiIjyi92lS5emT5+u4nZrT8MiMtJp1qyH3bolLlqkNjKSHQdAuYYPF15eYvBg0a6d\n2LVLtG0rOxCAZ6R7xc7e3v7KlStlzjpx4oSjo+NTrU1PT6979+7lL2NqavpU60T9HTscPv74\nfmBg8uTJ3H8Y0A0tW4o//hCTJgkfH/HllyIwUHYgAM9C986x69OnT3h4+NmzZ0sPpqWlffTR\nRxs2bPDz85MVDEIIUVJiv3Klw7Jld+bPT54yhVYH6BIzM7Fpk/jnP8WUKeLtt0VuruxAAJ6a\n7hW7BQsWmJubd+zYUdPhPvjggzZt2jg4OCxdutTZ2Xnu3LmyA9Zdqvx8p1mzbHbturlmTVr/\n/rLjAHgmkyaJEyfEkSPCx0fEx8tOA+Dp6F6xs7e3j46ODgoKunnzphAiJiYmJibGwsJi0qRJ\np0+fbtSokeyAdZR+RobrhAmmMTHxGzZkdeokOw6A59Chgzh7VtjZCS8v8a9/yU4D4CnoXrET\nQtjZ2YWHh6ekpCQlJV27di0pKSklJSU8PNzOzk52tDrK6PZt91Gj9HNz47Zty2vZUnYcAM+t\nQQNx4ICYOlUMGSLmzRMlJbIDAagU3bt4Qq1Wx8fHX79+PTMzUwhhZWXl4eEhO1SdVu/yZZeQ\nkNxWrRJWrizhQhNAMfT1xeLFomNHMWaMOH1afP+9qF9fdiYAFdClYpeWlrZkyZLNmzffe+wB\nOM7OzoGBgTNnzqxXr56UbHWW+alTzqGhD3v0SFywQG2gS/87AaiUvn3F6dNi4EDh5SV27xbt\n28sOBKA8OvOX+O7duz4+PvHx8R4eHn369HFxcTEzMxNCPHz4MC4u7vjx43Pnzt29e/exY8ds\nuLtmTbHZt89x/vz7b7/NbU0AJWvWTJw8KQIDRbduYu1aMXq07EAAnkhnil1YWNjt27cjIiKG\nDBny+Nzi4uK1a9dOmTJlwYIFn332Wc3Hq4Nsv/7aLjz87ocfPhg6VHYWANXMzExs2yY+/VQE\nBIjTp8XKlcLQUHYmAGXQmYsnDhw4MHr06DJbnRBCX18/JCRk6NChe/bsqeFgdZCqpMRx0SLb\nr79O+Oc/aXVAHRIaKo4cETt2iB49xP99qCOAWkJnil1qaqq7u3v5y7Rq1So5Oblm8tRZqvx8\np3fftTp8+Ma6dQ979JAdB0DN6tZNnDkjCgtFu3bi1CnZaQA8SmeKnaOj4/nz58tf5ty5c0/7\nSDE8Ff2sLNfg4HqXL1//7rscLy/ZcQDI0KSJOH5c9OkjuncX33wjOw2A/0Nnil3//v137ty5\nYsWK/Pz8x+dmZ2fPmzdv3759w4YNq/lsdYTB/ftNx483SEu7vnlzvpub7DgA5DE2Ft98Iz7/\nXEyZIiZOFAUFsgMB+F86c/HE/PnzIyMjZ82atXDhwg4dOjg5OZmbm6vV6qysrJs3b0ZFReXk\n5HTp0mXOnDmykyqTUUKC64QJRQ0a3Fy3rtjKSnYcALXAxImidWsxeLC4cEHs2iU4YALUAjpT\n7KytrU+ePLl69epNmzb98ssvxcXF2lmGhobt2rULCAgICAjQ19eXGFKpTK5edZ00Kbdly4R/\n/rOEOwUC0OrcWURHi8GDhbe32LVLdO4sOxBQ1+lMsRNCGBkZhYaGhoaG5uXlJSQkaJ48YWlp\n6ezsbGRkJDudYpmdOeM8ZUpmt26JixdzC2IAj3J0FL/8IqZOFa++Kr78UkyYIDsQUKfp5N9p\nExMTHiNWMyyOHXOeNevB4MF333tP6OnMGZkAapSRkVi7VrRrJ6ZOFWfPii++EPxjG5CEP9V4\nIusffnB+992UoKC7779PqwNQgQkTxM8/i337uMsdIJFy/lrHxcX17NmzZ8+esoMoRIOtWxvP\nmXP3vffuTZwoOwsAHeHjI6KjRWGhaN9enD4tOw1QFymn2GVmZh49evTo0aOygyiB3Vdf2S9f\nnrh06YMRI2RnAaBTGjcWx4+Lnj1F165i82bZaYA6RyfPsStTy5YtL1y4IDuF7lOr7Zcvrx8R\nceuzzzK7dZOdBoAOMjERGzaINm1EQIA4d078z/8IrrsCaopyfthMTEw8PT1lp9BtqpISx/nz\nrQ4fvrlmTXb79rLjANBl77wjPD3FsGHi4kWxfbuoX192IKBOUM6hWCFEampqbGys7BS6SlVY\n6DRrluWxY/Hr1tHqAFSBHj1EVJRIThYdOohLl2SnAeqEiovdyy+/vHbt2oyMjBpI85yWL1/O\nbVCejV5+vvM775jGxFzfsCGXHZ8AqkrTpuK334SXl3j5ZbF3r+w0gPJVXOyio6ODg4MdHBxG\njhz5n//8p6SkpAZioSbpZWe7BAcbX79+fePG/GbNZMcBoCzm5mLnTjFrlhg8WCxaJNRq2YEA\nJau42CUlJa1du7Zz584RERG9evVydXWdM2cORzwVQz8jo2lQkEFqavymTQVOTrLjAFAilUqE\nhYndu8Xy5WLIEJGdLTsQoFgVXzzRoEGDCRMmTJgwISUlZffu3RERER9//PGSJUteeeWVcePG\nDR061MLCogaCent7V7hMYmJiDSRREoPUVNegIKGnF79xYxGnNgOoVv36iZMnRb9+wsdH7N0r\nXF1lBwIU6CmuirW1tQ0ODg4ODr579+6OHTs2btwYGBj4zjvvjBkzJjQ0tHnz5tWXUghx7tw5\nIYShoWE5yxQVFVVrBoUxSElpOmFCiZHRjbVri62tZccBUAe88IKIihLDhglvbxERIXr0kB0I\nUJqnvio2Nzf3t99++/XXX//++28hRMOGDdevX+/p6blgwQJ1dZ45MWvWLDMzs4sXL+Y92cyZ\nM6svgMIYJSa6jR5dZGUVv349rQ5AzalfXxw8KEaOFG+8Ib76SnYaQGmeotj99ttvQUFB9vb2\nQ4YM+emnnwYOHHjs2LGbN2/GxcX5+/vPnz9/wYIF1Rd00aJFzZo1GzFiRGFhYfV9lTrC+ObN\npmPHFri43PzqqxJzc9lxANQxBgbiiy/EmjVi2jQREiL4rQ5UnYqLXUJCwpIlS5o3b/7KK6+s\nW7fO3d191apVd+7c+f7777t37y6EcHJy2rlzZ8+ePdesWVN9QQ0NDbds2XLp0qUPP/yw+r5K\nXWAcH980ICCvRYubX3xRYmIiOw6Auurtt8WxY2LPHtGjh7h3T3YaQCEqPsfO1dW1pKTEysoq\nODg4MDCwXbt2jy+jUqn69+9f3c9pbdWqVVJSUjkn0vXu3duao4rlMrl61XXChOz27W9/8oma\nh/wAkKtzZxEdLfr3F97eYu9e0bat7ECAzqv4T7uPj8/bb789dOjQevXqlbOYr6/v7t27qy5Y\n2SwtLcuZ261bt2483vTJ6l286DpxYmbXrolLlqj1FPXQEQC6qkkTcfy4GDtWdO8uNm8W/frJ\nDgTotor/ui9durRv375ltrqoqChtmWvWrNmAAQOqOB2qjmlMTNOgoMwuXWh1AGoXMzOxc6f4\n6CMxaJB4/33uYAw8j4r/wHfp0uXEiRNlzoqMjAwKCqrqSKh6ZmfOuE6cmN679+2lS2l1AGod\nlUrMni127BCrVolRo0RuruxAgK564qHY2NhY7eMlzp07Z/LYWfa5ubkRERH5+fnVmA5VweyP\nP1ymTEkbNOju7NlCpZIdBwCeYNAg0ayZ6NdPdO0q9u4VjRvLDgTonicWu127dn3wwQea6YUL\nFz5pscGDB1d9KFQd899/d37nnQcjRiTNmCE7CwBU5KWXRFSUGDRIdOgg/vUv0aGD7ECAjnli\nsXv//ffHjh17+vTpfv36jR49+h//+McjC+jr67u5ufn7+1dzQjw7i8hI59DQ+6NHJ0+bJjsL\nAFSOnZ04elSEhIhu3cS6dWLUKNmBAF1S3lWxDg4O/v7+fn5+ISEhnTp1qrFMqBIWx445z5yZ\nEhBwb/Jk2VkA4GkYGYl164Snpxg3Tly8KJYsEZwcDFRO2cUuKSnJ2NjYxsZGCLFu3TrNyJNW\nYW9vX03h8Mwsjx51mjXr3sSJKRMnys4CAM9k+nTxj3+IYcPEpUtiyxZhYSE7EKADyi52Dg4O\nvr6+//73vzXT5a+iWh8Ri2dg+Z//OL333r2QkBSuWQag03r1En/8Ifz9RefOYt8+4eYmOxBQ\n25Vd7IYNG+bl5aWdrsE8eF5Whw83mT373tSpKQEBsrMAwHNr3lycOiWGDRMdO4pduwR3oQfK\nVXax2759e5nTqOWsDh1q8v77ydOm3R83TnYWAKgi1tbiwAExY4Z4/XWxapWYMEF2IKD2quzT\nQouLi/X19TXT+fn5MTExRkZGXl5eKu6LVmv87766KVNodQCUxsBAfP65aN1aTJ4szpwRq1YJ\nQ0PZmYDaqOLrjIqLiydPnjx8+HDNyxs3bvzjH//o1KlT27Ztu3btmpWVVc0JUSlWhw41ee+9\ne1OnpgQGys4CANUjMFD8/LPYu1f4+Ym0NNlpgNqo4mK3fPny8PBwZ2dnzcvJkyfHx8dPmjQp\nJCTk999/X7VqVTUnRMWsfvqpyezZSTNnprz9tuwsAFCdfHzEqVPi7l3RqZP46y/ZaYBap+Ji\nt2XLloEDB65cuVIIkZiYePDgwYCAgPDw8NWrV48bN27Hjh3VHxLlsfr3v5t8+GHSzJmpb70l\nOwsAVL+mTcXvv4sWLUSnTuLwYdlpgNql4mJ348aNXr16aaYPHTqkVqtHjBihedmuXbsbN25U\nXzhUyPLw4Sbvv3/vnXdodQDqEAsLsW+fmDZN+PmJTz6RnQaoRSq+eKL05RFHjhwxMzPr0qWL\n5qVarS4sLKyuaKiI1aFDTWbPTnr33dQxY2RnAYCapVKJ+fNFs2YiKEjExorVq4WRkexMgHwV\n77FzcXE5ceKEECI5OfmHH37o1auX0X9/eM6fP9+kSZPqDYgn0FwDm0yrA1CXvfWWOHZMHDgg\nXn9d3L8vOw0gX8XFbuTIkVu3bu3cuXPbtm2zsrKm/fdx8ps2bfruu+/8/f2rOSHKYKlpddOm\n3afVAajjOnUSUVEiK0t06CAuXJCdBpCs4mIXGho6bty4mJiY7OzsL774ott/7/r9/vvvt2jR\n4oMPPqjmhHiU5ZEjTrNnJ0+den/8eNlZAKAWaNJEREYKb2/h4yP275edBpCp4mJnYmKyYcOG\nnJyc9PT0qVOnasf37Nlz9uxZGxub6oyHR1keO+Y0a9a9KVPu88QwANAyNRU7doiZM8XAgWLZ\nMsFDzFFXVfbJE4/r1KlTFeZAZVgcP+40Y8a9kBDuVwcAj1KpxNy54oUXxNix4uJFsW6dMDGR\nnQmoaRUXO7VavWvXrk2bNt2+fbvMa2AvXrxYDcHwKItff3V+992UCRNSgoJkZwGA2mrQIOHm\nJvr3F927i3/9Szg4yA4E1KiKi93KlStnzZolhDA1NTXk2XySmP/+u/P06Snjx98LDpadBai7\nkpOTr1y5MnHiRNlBUAHLbt2CDx9u4OER3qvXTVvbSr7r1q1bQgjtk5agPP7+/n5+frJTVK+K\ni93nn3/u6+sbHh7u5uZWA4HwOLM//nB+5537o0ffmzJFdhagTrtz545KpXrw4IHsIKjAAyHC\nfHwmnDkzc//+1d7evzs5VeZdv/76a4MGDczNzas7HqQ4ffp0QUEBxU4kJyfv2rWLVieLWXS0\ny5QpD0aMSP7vjWYASOTl5TVv3jzZKVBZDzZunP7ZZ295eSVPnixK3W+/TFFRUWxfBZszZ47s\nCDWh4mLXqFEjNZcXSWJ69qzL5MlpQ4YkzZghOwsA6J7748blu7s3ee8949jY20uXlpiayk4E\nVK+Kb3cyYsSIzZs310AUPML0zz9dQ0LS+/W7O2uW7CwAoKsyu3S5vmWLybVrbqNHG925IzsO\nUL0qLnZz586Ni4sbNWrUoUOHrly5EvuYGkhZB5lcveoSEpLx+ut33n+/wsMHAIBy5Lu5xW3d\nWlS/vvuwYWanT8uOA1Sjig/FWlhYaCa2bt1a5gIcqK1yJn/91TQwMLNbt8QFC4RexeUbAFC+\nYiurm1995fDJJ64TJ9756KO0QYNkJwKqRcXFbsSIEUZGRgYGz34rYzwV47g416CgLB+fxEWL\naHUAUFXU+vp3Pvwwz8PDcfFik7//TnrvPbW+vuxQQBWruK49aUcdqoPRzZtNJ0zIadfu9tKl\nalodAFS1B0OG5Lm7O4eGmvz1V8KnnxbxYEwoy1NUh8zMzEuXLqWnp1dfmjrO6Pbtpm+/nfvC\nCwnLl/PvSACoJjlt217fskU/I8Nt9Gjj+HjZcYCqVKlid/z4cW9vb0tL1UR7wwAAIABJREFU\nS09Pz1OnTmkG/f39jx49Wp3Z6hbDu3ebvv12vofHrRUr1Bz4BoDqVNCkyfXvv89zc3MbOdIi\nMlJ2HKDKVFzsoqKievXq9ffff/v6+moHU1JSTp8+3adPnzNnzlRnvLrC4N69poGB+U5Otz77\nTG1kJDsOAChfiZnZrc8+ezBypPPUqQ03bJAdB6gaFe8ZWrhwob29/W+//WZgYODw36cp29ra\nnj9/vn379osWLdq7d281h1Q4g9TUpoGBRQ0b3vryyxJjY9lxAKDO0NNLnjo1z8OjcViYyd9/\nG3OTB+i+ivfYnTp1atKkSU2aNHlk3M7OLjg4+MSJE9UTrK7QT093DQoqsbC4sXp1Sb16suMA\nQJ2T8cYb8Zs2mUVHRyQn18/Lkx0HeC4VF7uMjAynJzw+2cHBISsrq6oj1SH6mZmuEyYIA4Mb\na9aU8NhpAJAkt1WruO3bi1SqFceP17twQXYc4NlVXOzs7e2vXLlS5qwTJ044OjpWdaS6Qi87\n2yU4WFVYGL92bbGlpew4AFCnFTVoMMLO7qydndv48dY//CA7DvCMKi52ffr0CQ8PP3v2bOnB\ntLS0jz76aMOGDX5+ftWWTcn08vNdpk41yMi48fXXxdxFCQBqgQKValWbNnfff7/x3LmOCxao\niotlJwKeWsXFbsGCBebm5h07dtR0uA8++KBNmzYODg5Lly51dnaeO3du9YdUGqOSEuepUw3v\n3o1fv77I1lZ2HADA//dg8OCba9ZY/ec/LlOm6Gdmyo4DPJ1KHYqNjo4OCgq6efOmECImJiYm\nJsbCwmLSpEmnT59u1KhR9YdUFP2SklmnTxvHx99Yv76Q7x4A1D5ZnTrFbdtmmJTkNmIEdzCG\nbqnUDYrt7OzCw8NTUlKSkpKuXbuWlJSUkpISHh5uZ2dX3fmUprg4KDKyeVrajbVrCzg9EQBq\nqwInp7gtW/KbNXMfMcLy2DHZcYDKqtQTDq5du3bq1Kl79+4ZGBg0bty4a9eu1R1LmUpKREDA\nC3fvfujjM8HNTXYaAEB5SkxNb/3zn43Cw52mT783bVrK+PFCpZIdCqhABcUuKipq2rRp2seI\naahUKn9//xUrVjRr1qw6symLWi0mTRL796/o1esWvxoAQCfo6SVPmZLXvHnjOXNMrl5NXLiw\nxMREdiagPOUVu0OHDvXv3z8vL69t27a+vr6NGzcuLCyMjY09cODAvn37fvnll4MHD7788ss1\nllW3zZ4tvv9e/PvfN7/9VqSlyU4DAKisjF698l1dnd95x+2tt2598QUn0qA2e+I5dunp6WPG\njNHT09u5c+eZM2eWLl06efLk6dOnr1q1KjY2dtWqVTk5OQMGDMjIyKjJuLpqzhyxapX48UfR\npYvsKACAp5bXvHnctm3FVlZuw4ebRUfLjgM80ROL3caNG+/du7dq1arBgwc/MktfX3/y5Mmf\nfvppcnJyeHh4NSfUfR9/LJYvF7t3i1dflR0FAPCMim1sbnz9dUbv3q5BQfW3b5cdByjbE4vd\ngQMHmjRpMnbs2CctMGnSJGdn53379lVPMKX44gsxd67Ytk307i07CgDguaj19e9+8MGdefMc\nli9vPH++qqBAdiLgUU8sdhcuXOjSpYue3hMX0NPTe/XVV69evVo9wRRh/Xrx7rti40YxcKDs\nKACAqpHWv3/8hg0WkZFNAwIMUlJkxwH+jyf2tgcPHjg4OJT/Zjs7O86xe6Jt28TEiSI8XIwa\nJTsKAKAq5bz4Yuz27UKIZsOHm/75p+w4wP/3xGJX+P/au/OAKKvF/+Nn2JFdRVkEVKC0vIq5\noKk3U1NDr2sKmZIimpKVC7aogOCSXblm1/BmfnNBSwOhrMwWvYgbhpqapanAiCIghaDs6/z+\nmO+XH6nhks7heXi//po5z5mZDx6Xj888S1WVqanpHV785/vzmrovvhAvviiio8X06bKjAAAe\nvGpHR+2GDUV9+7abMsXhs89kxwH+111doBj35vvvxfjxIiJCzJ4tOwoA4GHRmZldiYws69jR\nJSrK8uzZnNdf15nwryoka+i34MGDBxcvXtzwhAccRwVSUsSYMSIkRCxcKDsKAOChuxYQUP7o\no+5z51r8+uulVauqW7aUnQhNWkPF7tChQ4cOHTJYFDU4dkwMHSomTxarVsmOAgAwkNKuXdO3\nbXOfPdvz+ecvrV5d9vjjshOh6frTYrdlyxZD5lCD06fF0KFi7Fjx73/LjgIAMKgqJ6eMzZtd\nlyxp/+KLV8LCCkeOlJ0ITdSfFruJEycaMofinT8vBg8WgwaJ9eu5SzQANEE6c/OspUvLHnvM\ndfFiyzNncufP55A7GB6ntT4ImZli0CDRo4fYskUYG8tOAwCQJn/ChIvr19t9803b4GCT/HzZ\ncdDkUOz+spwcMWiQePRRERcn7nSBGACA6pV0757+6adGFRWe/v6WXOUOhkWx+2t+/10884xo\n3Vp8/rmwsJCdBgDQKFQ5OWVs2lT85JPtp0xxSEyUHQdNCF///wXXr4uhQ4W5udi1S1hZyU4D\nAGhEdObmV6Kiyh5/3GXpUsuff8556y0d3+rg4WOP3f0qKRHDh4uyMvHtt8LOTnYaAEBjdM3f\nX7thg21ycrspU0zz8mTHgfpR7O5LRYUYPVrk5IjvvxdcixIA8OdKfXzStm8Xxsae48dbHT8u\nOw5UjmJ372pqxKRJ4swZ8f33wsVFdhoAQGOnv7FswciRbadOddywQXYcqBnH2N2j2loRGCiS\nk8X+/aJdO9lpAADKoDM2vjpnToWnp0tUlPm5c9mRkbWccoeHgD1290KnEzNnit27xXffiUcf\nlZ0GAKAwhSNGZGzZ0uzkyfYvvGB26ZLsOFAhit29ePNNsXWr+OIL0aWL7CgAAEUq79gxPS6u\nulUrL39/2717ZceB2lDs7pbxkiVizRrx1Veib1/ZWQAAClZjZ3cxJib/hRfc5s5tFRMjamtl\nJ4J6UOzuyjQhjN95R8THi6eflp0FAKB8RkZXZ8269N57LT7+2OPll42vX5cdCCpBsbsrXYWo\n3rxZDBsmOwgAQD2K+vdP377dNC/P09/f8uxZ2XGgBhS7u7JMiNqxY2WnAACoTaW7e8bHH5d2\n7dp+0iSHzz6THQeKx+VO7soV2QEAAGpVa2GR9fbbpV26uCxZ0uzUqey33tKZm8sOBaVijx0A\nAPJdCwjQbtpkffBg+8BAsyvsT8B9otgBANAolHbunB4fX2tj4+nvb7N/v+w4UCSKHQAAjUW1\ng8PFDz+8Nn68+yuvtF6zRsOVUHCPKHYAADQiOiOjq6++eunf/26+fbvH9Okm167JTgQlodgB\nANDoFD31VHpcnHFRkee4cc1OnJAdB4pBsQMAoDGqdHXNiI0t/vvf2wUFtdiyReh0shNBAbjc\nCQAAjZTO3PxKRERJ164uS5ZYnTiRFRVVa20tOxQaNfbYAQDQqBWOGJGxbZt5WppXQIDFuXOy\n46BRo9gBANDYlXt5pW/bVvbYY+1feMEhIUF2HDReFDsAABSg1srq8j//mTtvnsvy5W0WLjQq\nL5edCI0RxQ4AAMW49vzzGbGxzY4fb//88+Zarew4aHQodgAAKEnZ44+nx8VVubl5BgTYf/WV\n7DhoXCh2AAAoTI2tbeZ771195RXX8PA2CxbwtSzqUOwAAFAgjSZ/4kTtxo1WR4+2f+EF84sX\nZQdCo0CxAwBAqUq7dEmLj69ycvIMCLD/+mvZcSAfxQ4AAAWrsbfPfP/9vJdecl2wwCUy0qii\nQnYiyESxAwBA4TSa36dM0W7aZHPwYPsJEzhbtimj2AEAoAalPj5p8fFVzs6eAQH2X34pOw7k\noNgBAKASNfb2mWvW5L38smt4uOuiRUZlZbITwdAodgAAqIhG83tgYMbmzVbHjnkGBFhcuCA7\nEAyKYgcAgNqUde6cHh9f4enZ/vnnm+/YITsODIdiBwCACtXY2FxatSo3NNR5xQq30FDj4mLZ\niWAIFDsAAFTrWkBA2qefmqene40Z0+zkSdlx8NBR7AAAULMKT8+MTz4p7tmz3ZQpLTduFDqd\n7ER4iCh2AACoXK2l5ZWlS68sWdJq3bq2M2ea5OfLToSHhWIHAECTUDh8eNqnnxoXFHg995x1\nSorsOHgoKHYAADQVlR4eGVu3Fvr5ecyY4fTuu5rqatmJ8IBR7AAAaEJ0pqa58+dnxsTY79zZ\nftIks8uXZSfCg0SxAwCgySnu2zdtx44aW1uvcePsv/hCdhw8MBQ7AACaouqWLS9+8MHVWbNc\nFy92nzfP+MYN2YnwAFDsAABoqjSa/IkTM7ZuNT93ztPfv9lPP8kOhL+KYgcAQJNW9thj6XFx\nJT16tAsMdPzwQ01trexEuH8UOwAAmrraZs2uREVlvfNOy02b2k6dapqbKzsR7pOyi11NTc3p\n06cPHz58mZN6AAD4a64PGZKWkCBqa73GjrX79lvZcXA/lFTsDh8+PGvWrLqnW7dudXV17dy5\nc58+fdzd3X18fPbv3y8xHgAASlfl7Hxx48bfX3yxzZtvui5aZFRSIjsR7o1iit2+ffuefvrp\nzZs363Q6IcSOHTsmTZpUUlIybty4kJCQZ5555vTp04MHDz5+/LjspAAAKJjOyOi36dMzNm+2\nOnHC67nnmp06JTsR7oFiil1kZKS9vf2JEyc0Go0Q4vXXX/fw8Dh//nxcXFxMTMx33313+PBh\nIyOjyMhI2UkBAFC8ss6d0+Liinv1avfii625R4VyKKbY/fjjj4GBgV5eXkKI69eva7XauXPn\nOjs7103w9fWdOHHigQMH5GUEAEA9aq2ssiMiLv/zn80TEtpNnmyWlSU7Ee5MMcWupqbG0tJS\n/9jCwkKj0bRp0+amOW3atCkvLzd4NAAAVOvG4MFpCQm1FhZezz3n8NlnsuPgDhRT7Hx8fLZv\n315aWiqEMDc37927d0pKSv0JFRUViYmJjz76qKSAAACoU1Xr1hfXr78aEuKybJn77NnGBQWy\nE+FPKabYvfnmmxcuXOjXr993331XXV29Zs2ajz/+ODY2trS0tKqq6ocffvDz8zt16lRISIjs\npAAAqI5Gkx8YmL59u9nly95jx1ofPCg7EG5PMcVu+PDh69evP3fu3JAhQ2xtbSdNmmRmZvbi\niy/a2tpaWlr26tUrKSlp7ty506ZNk50UAAB1KvfySt+2rXDYMI+XX3ZZvtyookJ2ItxMMcVO\nCBEcHJyenr5y5cq///3vJSUl165dMzc3t7e39/HxefXVV48dO/avf/1Lf84sAAB4GHRmZrnz\n5l1cv95m3z7PceMsf/5ZdiL8gYnsAPemdevWoaGhoaGhsoMAANB0lfTsmZaQ4LxsWfuJE397\n6aXfpk/XGRvLDgUhFFfshBA6nU6r1WZkZBQVFQkh7OzsvL293dzcZOcCAKAJqbGxyVqxomjA\nAJclS2wOHMhavryibVvZoaCoYldQULBs2bItW7bk5eXdtMnd3T04ODg0NLTukigAAOBhuz54\ncImPT5uICM9x43Lnzr0WECA4JkoqxRS7nJycPn36aLVab29vPz8/Dw8PKysrIcSNGzfS09OT\nk5PDw8MTEhKSkpIcHBxkhwUAoKmobtXq4tq1zePjnaKjbfftu7JkSVWrVrJDNV2KKXZhYWFZ\nWVlxcXHjxo27dWtNTc26detmzZoVGRm5evVqw8cDAKDp0miujR9f7OvbZuFCr9GjcxYsKBw2\nTHamJkoxZ8Xu2rVr0qRJt211QghjY+OQkJDx48cnJiYaOBgAABBCVHp4aGNjf58yxTUszH3e\nPK5jLIViil1+fr6np2fDczp27Hj16lXD5AEAADfRGRn9Fhycvn272cWL3qNH2yYlyU7U5Cim\n2Lm4uJw6darhOSdOnHBxcTFMHgAAcFvljzyS/umn+YGBbnPnus2bZ3z9uuxETYhiit2oUaPi\n4+Ojo6Mrbned65KSkoiIiJ07d/r7+xs+GwAAqE9nYvJbUFDG5s0W5897jR1rfeiQ7ERNhWJO\nnli8ePGBAwfmz58fFRXVs2dPNzc3a2trnU5XXFycmZmZmppaWlrar1+/RYsWyU4KAACEEKKs\nc+f0HTtar17dNiTk2pgxuaGhtVZWskOpnGKKnb29fUpKSkxMTGxs7L59+2pqauo2mZqaduvW\nLSgoKCgoyJgrXwMA0GjUmpvnvPHGjYEDXcPCvMaMubJ0aUmPHrJDqZliip0QwszMbM6cOXPm\nzCkvL798+bL+zhO2trbu7u5mZmay0wEAgNsr6d49LSHBadWqdsHB+f7+V+fMqeWGAg+Hkopd\nHQsLC29vb9kpAADA3apt1ix70aIbzzzjGh5uc/DglaVLS554QnYoFVLMyRO3FR0d3bdvX9kp\nAADAXSn29b2QkFDi69t2yhTnd94xKi+XnUhtlF3s0tLSDnGiDQAAylFrbX0lIiLzP/+x3bvX\na+zYZj/+KDuRqii72AEAACUqfvLJC4mJJT16tGPX3QNFsQMAABLUWltfWbw484MPbP/7X6+x\nY62OH5edSA0odgAAQJri3r0vJCYW+/q2DQpyWb7cqLRUdiJlU3axW7FixeXLl2WnAAAA96/W\nyio7PDzzww+tk5O9Ro+2PnJEdiIFU3CxKyoqunLlirW1tewgAADgryr29U377LPip55q+9JL\nrpGRxsXFshMpkiKLXXJycvfu3W1tbTt16nTk/3r9iBEj9u7dKzcYAAC4b7XNmmUvWKDdsMHq\n6FGvkSNtkpNlJ1Ie5RW71NTUwYMHnz9/fsiQIXWDv/3229GjR/38/I5z6CUAAEpW0q1bWkLC\ndT8/91dfbfPmm8aFhbITKYnyil1UVJSTk9OZM2c2bdpUN+jo6Hjq1CknJ6clS5bIiwYAAB6A\nWnPz3HnzMrZutTh3znvkSLtvv5WdSDGUd0uxI0eOhIaGtmnTJjc3t/54q1atZsyYsXLlynt6\nN61W6+vrW11d3cAc/VadTncfaQEAwP0p+9vf0uPiHNevb/PWW3Zff529cGF1q1ayQzV2yit2\n169fd3Nzu+0mZ2fn4ns81tLDwyMuLq7hYvfLL7/Mnj1bo9Hc0zsDAIC/SGdqmhcScuOZZ1wj\nIrxHjcqdN69gzBjBv8h/TnnFzsnJ6ezZs7fdtH//fhcXl3t6NyMjo/79+zc8p1mzZvf0ngAA\n4AEq9/bO2Lq1xdatzitW2O/efSUiovJPdvFAecfY+fn5rV279sc/3lquoKBg4cKFGzduHDZs\nmKxgAADgIdEZGf0eGJj22Wc6IbzGjGm5aZOmpkZ2qMZIecUuMjLS2tra19dX3+Heeuutrl27\nOjs7L1++3N3dPTw8XHZAAADwUFS2aXNx/fqcBQsc169vP2GCxblzshM1Osordk5OTseOHZs2\nbVpmZqYQ4uTJkydPnrSxsZk5c+bRo0dbt24tOyAAAHhoNJqC0aMv7NxZ5erq6e/f+t13jSoq\nZGdqRJR3jJ0QolWrVmvXro2JicnLyysqKrKxsaHPAQDQdFS3bHlp1Srb//7Xedkyu++/z46I\nKPb1lR2qUVBesdPpdFqtNiMjo6ioSAhhZ2fn7e0tOxQAADC0GwMGFPfs6fTee22nTy/8xz9y\nQkNr7O1lh5JMScWuoKBg2bJlW7ZsycvLu2mTu7t7cHBwaGiopaWllGwAAMDwaq2tsxcuLBw2\nzDUy0nvkyNw33ij085MdSibFFLucnJw+ffpotVpvb28/Pz8PDw8rKyshxI0bN9LT05OTk8PD\nwxMSEpKSkhwcHGSHBQAAhlPq45MWF+f40UeuYWH2X36ZvWhRpaur7FByKKbYhYWFZWVlxcXF\njRs37tatNTU169atmzVrVmRk5OrVqw0fDwAASKQzNc2bMeP60KEukZFeo0blhYTkBwbqjI1l\n5zI0xZwVu2vXrkmTJt221QkhjI2NQ0JCxo8fn5iYaOBgAACgkaho21a7YUPOggWOH33k6e9v\n+dNPshMZmmKKXX5+vqenZ8NzOnbsePXqVcPkAQAAjZH+eihffFHu6dl+0iTn5cuN7vF2o4qm\nmGLn4uJy6tSphuecOHHiXm8pBgAA1Ke6efOsd97J/M9/bA4c8B450vb772UnMhDFFLtRo0bF\nx8dHR0dX3O46hCUlJRERETt37vT39zd8NgAA0AgVP/lk2uefF/7jH26vv77ohx9aNIFdd4o5\neWLx4sUHDhyYP39+VFRUz5493dzcrK2tdTpdcXFxZmZmampqaWlpv379Fi1aJDspAABoLGrN\nza/Onn19+HCrqVOXff65iI4Ws2cLE8X0n3ulmB/M3t4+JSUlJiYmNjZ23759NfVu/Wtqatqt\nW7egoKCgoCDjpnf+CwAAaFi5l1dYv34jf/tt8ttviy1bxAcfiN69ZYd6KBRT7IQQZmZmc+bM\nmTNnTnl5+eXLl/V3nrC1tXV3dzczM5OdDgAANF46IZIfeWTyZ5+J0FDRt6+YNk28/bZQ3bVv\nlVTs6lhYWHAbMQAAcM9atRKxsWLKFDFzpujQQURHi4kThUYjO9YDo5iTJwAAAB6Mp58Wp06J\nV14RL70kBg4Uv/4qO9ADo55il56ePmjQoEGDBskOAgAAGj1zc7FokTh9Wpibiy5dxMKForRU\ndqYHQD3FrqioaO/evXv37pUdBAAAKISnp9i9W3z8sdi8WTz+uPjyS9mB/ir1FLsOHTqcPn36\n9OnTsoMAAABFee45cfasGD1ajBkjRo4UmZmyA90/9RQ7CwuLTp06derUSXYQAACgNDY2YtUq\ncfy4yM8Xjz0m3n5bVFbKznQ/lF3sKisrjx49mpSUpNVqZWcBAAAK17mzOHBAvP++ePdd0bmz\n2LNHdqB7pphit3Tp0qSkpPoj69atc3Jy6tmz54ABA9q3b9+9e/eTJ0/KigcAANRAoxFTpohf\nfxVPPy2GDBH+/uLKFdmZ7oFiil1YWNi3335b93TXrl0zZswoLS0dPXr0Sy+91KdPn+PHj/fv\n3z89PV1iSAAAoAbNm4v//Ef88IPQav/3cndVVbIz3RXFFLubzJkzx87O7sSJE4mJiR988MHB\ngwcTEhJu3LixbNky2dEAAIAqdO8ujhwR0dHi7bdF165i3z7Zge5MkcXut99+u3Dhwssvv9yx\nY8e6wTFjxowcOfK7776TGAwAAKiKkZF46SVx7pzo3VsMHCgmTBDZ2bIzNUSRxa68vFwIUb/V\n6XXq1CkvL09GIgAAoF4tW4r160VKirhwoZF/M6vIYufi4mJnZ5eVlXXTeHZ2to2NjZRIAABA\n5Xr2FD/8IKKjxYoVYv582WluT0nF7tKlS8eOHUtLSysoKAgJCfnoo49K693949dff/3000/7\n9OkjMSEAAFAzIyMxfbo4f14MHSo7yu2ZyA5wD7Zt27Zt27b6I7t37x47dqwQ4pNPPpk+fXpZ\nWVlYWJikdAAAoGlo3pxi91dt3LixsJ7r168XFhY6ODjotxYWFtrb22/fvr1Hjx5ycwIAAMii\nmGI3efLkBrYGBgbOmDHDyEhJ3ywDAAA8WIopdg2ztraWHQEAAEAydnEBAACoBMUOAABAJSh2\nAAAAKkGxAwAAUAmKHQAAgEpQ7AAAAFSCYgcAAKASFDsAAACVoNgBAACoBMUOAABAJSh2AAAA\nKkGxAwAAUAmKHQAAgEpQ7AAAAFSCYgcAAKASFDsAAACVoNgBAACoBMUOAABAJSh2AAAAKkGx\nAwAAUAmKHQAAgEpQ7AAAAFSCYgcAAKASFDsAAACVoNgBAACoBMUOAABAJSh2AAAAKkGxAwAA\nUAmKHQAAgEpQ7AAAAFSCYgcAAKASFDsAAACVoNgBAACoBMUOAABAJSh2AAAAKkGxAwAAUAmK\nHQAAgEpQ7AAAAFSCYgcAAKASFDsAAACVoNgBAACoBMUOAABAJSh2AAAAKkGxAwAAUAmKHQAA\ngEpQ7AAAAFSCYgcAAKASFDsAAACVoNgBAACoBMUOAABAJSh2AAAAKkGxAwAAUAmKHQAAgEpQ\n7AAAAFSCYgcAAKASFDsAAACVoNgBAACoBMUOAABAJSh2AAAAKkGxAwAAUAmKHQAAgEpQ7AAA\nAFSCYgcAAKASFDsAAACVoNgBAACoBMUOAABAJSh2AAAAKkGxAwAAUAmKHQAAgEpQ7AAAAFSC\nYgcAAKASFDsAAACVoNgBAACoBMUOAABAJSh2AAAAKkGxAwAAUAmKHQAAgEpQ7AAAAFSCYgcA\nAKASFDsAAACVoNgBAACoBMUOAABAJSh2AAAAKkGxAwAAUAmKHQAAgEpQ7AAAAFSCYgcAAKAS\nFDsAAACVoNgBAACoBMUOAABAJSh2AAAAKkGxAwAAUAmKHQAAgEpQ7AAAAFSCYgcAAKASFDsA\nAACVoNgBAACoBMUOAABAJSh2AAAAKkGxAwAAUAkT2QHumU6n02q1GRkZRUVFQgg7Oztvb283\nNzfZuQAAACRTUrErKChYtmzZli1b8vLybtrk7u4eHBwcGhpqaWkpJRsAAIB0iil2OTk5ffr0\n0Wq13t7efn5+Hh4eVlZWQogbN26kp6cnJyeHh4cnJCQkJSU5ODjIDgsAACCBYopdWFhYVlZW\nXFzcuHHjbt1aU1Ozbt26WbNmRUZGrl692vDxAAAApFPMyRO7du2aNGnSbVudEMLY2DgkJGT8\n+PGJiYkGDgYAANBIKKbY5efne3p6NjynY8eOV69eNUweAACAxkYxxc7FxeXUqVMNzzlx4oSL\ni4th8gAAADQ2iil2o0aNio+Pj46OrqiouHVrSUlJRETEzp07/f39DZ8NAACgMVDMyROLFy8+\ncODA/Pnzo6Kievbs6ebmZm1trdPpiouLMzMzU1NTS0tL+/Xrt2jRItlJAQAA5FBMsbO3t09J\nSYmJiYmNjd23b19NTU3dJlNT027dugUFBQUFBRkbG0sMCQAAIJFiip0QwszMbM6cOXPmzCkv\nL798+bL+zhO2trbu7u5mZmay0wEAAEimpGKnp9PpsrOzMzMz626uACaUAAAVyElEQVQpZm5u\nzi3FAAAAlFTsuKUYAABAAxRT7LilGAAAQMMUU+y4pRgAAEDDFHMdO24pBgAA0DDFFDtuKQYA\nANAwxRQ7bikGAADQMMUUO24pBgAA0DDFnDzBLcUAAAAapphixy3FAAAAGqaYYie4pRgAAECD\nlFTs6lhYWHh7e8tOAQAA0Lgo5uSJm+Tm5o4aNSolJUV2EAAAgMZCqcWuuLh4586dV65ckR0E\nAACgsVBqsQMAAMBNKHYAAAAqobCTJ0JDQ/UPCgsLhRCxsbFHjhyp2xodHS0nFgAAQCOgsGK3\nY8cO/YPq6mohxJEjR3766ae6rRQ7AADQlCms2F28eFH/IC0tzdvbe+3atc8995zURAAAAI0F\nx9gBAACoBMUOAABAJSh2AAAAKqGwY+zqeHl5VVVVGRsbyw4CAADQWCh4j11ZWdmZM2f01z0B\nAACAIotdcnJy9+7dbW1tO3XqVHcduxEjRuzdu1duMAAAAImU91Vsamrq4MGDzc3NhwwZ8u23\n3+oHf/vtt6NHj/r5+R0+fLhbt253/25ardbX11d/Vbw/o9+q0Wj+Suw6xsbGX3/9dVJS0gN5\nNzQ2xcXFOTk5rK9asb7qxvqqW1lZWWBgoOwUD51Gp9PJznBvhg8ffvr06UOHDpmYmDg7O+/e\nvXvo0KFCiLy8vB49enTt2vXzzz+/+3erra3dv39/w8VOp9Pl5eW98MILfzW6EEKInJycX375\n5YG8FRqha9euCSGaN28uOwgeCtZX3Vhf1Xv88cednZ1lp3i4lLfH7siRI6GhoW3atMnNza0/\n3qpVqxkzZqxcufKe3s3IyKh///4PMt+dODs7q/53FQAAkEJ5x9hdv37dzc3ttpucnZ2Li4sN\nnAcAAKCRUF6xc3JyOnv27G037d+/38XFxcB5AAAAGgnlFTs/P7+1a9f++OOP9QcLCgoWLly4\ncePGYcOGyQoGAAAgl/JOnsjNze3Zs2dOTk7nzp1//PFHHx8fIcTZs2crKirc3d1TU1Nbt24t\nOyMAAIAEyttj5+TkdOzYsWnTpmVmZgohTp48efLkSRsbm5kzZx49epRWBwAAmizl7bGro78K\nSVFRkY2NDX0OAABAeZc70el0Wq02IyOjqKhICGFnZ+ft7S07FAAAgHxKKnYFBQXLli3bsmVL\nXl7eTZvc3d2Dg4NDQ0MtLS2lZAMAAJBOMV/F5uTk9OnTR6vVent79+nTx8PDw8rKSghx48aN\n9PT05OTk7OzsLl26JCUlOTg4yA4LAAAggWL22IWFhWVlZcXFxY0bN+7WrTU1NevWrZs1a1Zk\nZOTq1asNHw8AAEA6xeyxc3Z29vPz++ijjxqYExAQcPjw4UuXLhksFQAAQOOhmMud5Ofne3p6\nNjynY8eOV69eNUweAACAxkYxxc7FxeXUqVMNzzlx4gS3FAMAAE2WYordqFGj4uPjo6OjKyoq\nbt1aUlISERGxc+dOf39/w2cDAABoDBRzjF1hYeHAgQN//PFHGxubnj17urm5WVtb63S64uLi\nzMzM1NTU0tLSfv36ff3119bW1rLDAgAASKCYYieEqKysjImJiY2NPX36dE1NTd24qalpt27d\ngoKCgoKCjI2NJSYEAACQSEnFrk55efnly5f1d56wtbV1d3c3MzOTHQoAAEAyRRY7AAAA3Eox\nJ08AAACgYeopdunp6YMGDRo0aJDsIAAAAHIo5pZid1RUVLR3717ZKe6sd+/eR44ckZ0CAIAm\np1evXikpKbJTPFzqKXYdOnQ4ffq07BR31r59e0dHx4iICNlB8FBERkYKIVhftWJ91Y31VbfI\nyEgbGxvZKR469RQ7CwuLTp06yU5xZ2ZmZi1atOjWrZvsIHgoWrRoIYRgfdWK9VU31lfd9Our\nesordjqdTqvVZmRk6C93Ymdn5+3t7ebmJjsXAACAZEoqdgUFBcuWLduyZUteXt5Nm9zd3YOD\ng0NDQy0tLaVkAwAAkE4xxS4nJ6dPnz5ardbb29vPz8/Dw8PKykoIcePGjfT09OTk5PDw8ISE\nhKSkJAcHB9lhAQAAJFBMsQsLC8vKyoqLixs3btytW2tqatatWzdr1qzIyMjVq1cbPh4AAIB0\nirmO3a5duyZNmnTbVieEMDY2DgkJGT9+fGJiooGDAQAANBKKKXb5+fmenp4Nz+nYsePVq1cN\nkwcAAKCxUUyxc3FxOXXqVMNzTpw44eLiYpg8AAAAjY1iit2oUaPi4+Ojo6MrKipu3VpSUhIR\nEbFz505/f3/DZwMAAGgMFHPyxOLFiw8cODB//vyoqKiePXu6ublZW1vrdLri4uLMzMzU1NTS\n0tJ+/fotWrRIdlIAAAA5FFPs7O3tU1JSYmJiYmNj9+3bV1NTU7fJ1NS0W7duQUFBQUFBxsbG\nEkPeDTMzM9kR8BCxvurG+qob66tuTWR9NTqdTnaGe1ZeXn758mX9nSdsbW3d3d0VtFoFBQVC\nCC62p1asr7qxvurG+qpbE1lfRRY7AAAA3EoxJ08AAACgYRQ7AAAAlaDYAQAAqATFDgAAQCUo\ndgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAA\nACpBsQMAAFAJih0AAIBKUOwMp7CwcPbs2W3btjUzM3NxcQkODs7JyZEdCv+roKAgNDTUw8PD\n3Ny8Xbt2o0aNOnLkSP0Jd1w+A0zAAzF37lyNRhMcHFx/kPVVut27dz/11FM2Njb29vYDBgzY\nt29f/a2sr6L9+uuvkyZNcnZ2NjU1dXR0HD16dGpqav0JrO8f6GAQFRUVTzzxhBBi7Nixy5Yt\nCwoKMjU1bdeu3bVr12RHgy4/P79t27ZCiGHDhoWFhb3wwgsmJiYWFhY//fSTfsIdl88AE/BA\nHD161NjYWAgxderUukHWV+k2bNgghPD09Fy0aFFoaKijo6OZmdmhQ4f0W1lfRfv5559tbGya\nN28eHh4eGxu7ZMkSJycnExOTvXv36iewvjeh2BnIqlWrhBDvvPNO3cinn34qhJg3b57EVNB7\n+eWXhRBr1qypG0lISBBC+Pn56Z/ecfkMMAF/XVVVlY+PT5cuXW4qdqyvol29etXa2rpr167F\nxcX6kQsXLlhbW4eEhOifsr6KNmHCBCHEf//737qRU6dOCSH69++vf8r63oRiZyA+Pj42Njbl\n5eX1B728vFq1alVbWysrFfRmz549cODAysrKupHa2lpLS0sPDw/90zsunwEm4K9bsWKFRqPZ\nvXv3TcWO9VW0lStXCiG++eab+oP1f1VZX0Xz9fUVQtT/+1mn09na2rZt21b/mPW9CcfYGUJ5\nefnp06d79uxpbm5ef7xv3755eXlarVZWMOi9++67e/bsMTU1rRuprKysrq5u06aNuIvlM8CE\nB/8zNz3p6emRkZEzZszo1atX/XHWV+n27NljaWk5YMAAIURFRcWNGzeEEBqNRr+V9VW6Dh06\nCCHOnTtXN/L7778XFxd37NhRsL63Q7EzhMuXL9fU1Li5ud007uHhIYTIyMiQEQoNWbduXVVV\nVUBAgLiL5TPAhAf2gzVhL730kr29/dtvv33TOOurdL/++mu7du1+/vnnvn37Wlpa2tnZeXl5\nbdq0Sb+V9VW6N954w8HBYeLEiQcPHszNzT1x4kRAQICFhUVERIRgfW+HYmcIRUVFQggrK6ub\nxq2treu2ovFITk6eP39+3759Z8yYIe5i+Qww4YH8XE3Zpk2b9u7du2bNGjs7u5s2sb5Kd+3a\ntZKSkmHDhvXq1Ss+Pv69996rqqqaMmXKJ598Ilhf5evYsWNKSkpVVVW/fv2cnZ2feOKJCxcu\n7NmzR/8VLet7KxPZAZqQuq8G6uh0utuOQ6Jt27ZNmTKlU6dOO3fuNDH5/39A7rh8BpiA+5OX\nlzdv3rzhw4ePHTv2z+awvspVWVmZmZm5efPmwMBA/ci4ceMeeeSRefPm+fv760dYX+U6e/bs\nsGHDqqur//Wvfz3yyCN5eXmrVq169tlnd+zYMWjQIP0c1rc+ip0h2Nraitv1ev2xIDY2NhIy\n4RY6nW7x4sVRUVFDhw6Ni4urW5c7Lp8BJvz1n64pe+211yorK2NiYm67lfVVOmtr6+rq6uee\ne65uxNnZ+dlnn42Pjz9z5gzrq3RBQUFXr149f/68q6urfiQgIOCRRx6ZPHmyVqtlfW/FV7GG\n4O7ubmJikpmZedN4enq6EMLb21tGKPyBTqcLDg6Oiop65ZVXvvrqq/p/Vu+4fAaY8AB+wqZq\n9+7d27dvnzNnjpGRUVZWVlZWVnZ2thCitLQ0Kyvrxo0brK/S6S9CWf/kJyGEo6OjEKKoqIj1\nVbTi4uIffvjB19e3rtUJIZo1azZw4MArV66cP3+e9b0NA5+F22T5+vo2a9aspKSkbqSmpsbF\nxcXNzU1iKtR57bXXhBDLly+/7dY7Lp8BJuD+zJs3r4G/AN944w0d66tws2bNEkIcOXKk/uDg\nwYOFEJcuXdKxvkqWl5cnhOjdu/dN4+PHjxdCHDt2TMf63oI9dgYyderU0tJS/fWW9D788MPs\n7Oyb7msEKRITE997773XXnvtrbfeuu2EOy6fASbg/kydOvXLP9q+fbsQYvDgwV9++eXkyZMF\n66twkydP1mg0CxYsqKio0I8cO3Zsz549nTt31p/JyPoql6OjY7t27Y4dO3b+/Pm6wcLCwj17\n9tja2nbq1EmwvreS3Sybiurq6n79+gkhRo4cGRkZGRAQoNFo/va3v9X/HwBk8fT0FEK88sor\nb9xCf8eYOy6fASbgQSkoKBB/vEAx66t0s2fPFkL4+PhERkZOmzbN0tLSzMwsKSlJv5X1VbTE\nxEQjI6MWLVosXLhww4YNy5Yta9eunRAiJiZGP4H1vQnFznCKior0t5k3NTV1dXV9+eWX8/Pz\nZYeCTqfTNfA/H61Wq59zx+UzwAQ8ELcWOx3rq3C1tbUffPBBly5dLCws7Ozs/Pz8UlNT609g\nfRXt8OHDo0aNcnR0NDExcXBwGDRo0K5du+pPYH3r0+ga/FcNAAAASsExdgAAACpBsQMAAFAJ\nih0AAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0A\nAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBK\nUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwA\nAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgCUKjg4WKPRpKWl/dmEgIAAjUaTm5tr\nsEgmJia9evUy2McBwE0odgBUy8fHZ8iQIebm5nczecWKFQ10xIdEyocCUDGKHQDVevPNN7/5\n5hsHB4c7zszJyXnrrbcM3LGkfCgAdaPYAYA4evRoE/lQAOpGsQOgbEZGRu+880779u3Nzc3d\n3d2XLFmi0+n0m+ofY1dRUbFy5couXbrY2dnZ2Nh07tx55cqVtbW1Qojhw4ePHDlSCPHss89q\nNJqDBw/e5Ud//fXX3bp1s7S0bNWqVXBwcGFh4U0TUlNTR48e3bJlSzMzs7Zt206aNOnixYv6\nTX/2oQ28BADuyER2AAD4S5YuXXry5Mnp06cbGxuvWbMmPDzcy8vr+eefv2nazJkzN27cOGHC\nhJkzZ2o0mm+//fb111/PzMx8//33Fy1a1Lx58y1btoSHh3ft2vWxxx67m889dOjQiBEjWrdu\nHR4e7ujomJycPGLECCOj//+/5ePHjz/11FPNmzd/7bXXnJycMjIyYmJivvvuuzNnzrRo0eK2\nH9rwSx7gLxoA1dIBgDJNnTpVCNG3b9/Kykr9yPHjx4UQI0aM0D/19/cXQuTk5Oh0umbNmvXu\n3bv+y+fMmTN27Njq6mqdTvf2228LIXbv3n33n/7ss88KIVJTU+tGQkJChBC+vr76p2vXrn3i\niSeSkpLqJqxZs0YIsWbNGv3TWz/0ji8BgIbxVSwAZZs3b56pqan+cdeuXY2NjbOzs2+dZmpq\nmpmZmZeXVzeyatWqHTt2GBsb38eH1tbW7tu3z9PTs0ePHnWD06ZNqz9n5syZx48f79+/vxCi\nqqqqvLxcv1uuga9W7+MlAFAfxQ6Asnl7e9c91mg01tbWZWVlt06LiorKzs729vYODAzcuHHj\nlStX/sqH5uTklJWVtW/fvv5ghw4dbpq2ZcuWp556ysHBwczMzNLScuDAgUKI6urqBt75Pl4C\nAHUodgCU7S4vU/fqq6/u3bt3wIABiYmJQUFBbm5uw4YNy8zMvL8PLS0tFUJYWFjUH7SwsNBo\nNHVPFyxYEBgYWFpa+u677+7bty8lJeV//ud/Gn7b+3gJANTHyRMAmooBAwYMGDCgoqLiwIED\nW7dujY2NHTRo0C+//GJmZnavb2VpaSmEKC8vrz9YXFys+78TcsvLy1evXu3m5paUlGRtba0f\nvH79egPveR8vAYCbsMcOQNNibm4+aNCgTZs2zZgxIy0t7eTJk/fxJk5OTmZmZlqttv7gTz/9\nVPc4Nze3rKyse/fudRVNCJGcnNzAe97HSwDgJhQ7AOp35MgRV1fX2NjY+oP6S5PoT7zQn0Jx\n24PzbsvExOTJJ59MS0urf5HhmJiYusetW7fWaDT1T3o4efKkPkDdfr6bPvRuXgIADeOrWADq\n17179+bNm0+bNu3gwYM+Pj4ajebYsWObNm3q27evj4+PEEJ/GsSKFSu0Wm2/fv3qn+v6Z15/\n/fXk5OThw4cHBQW1aNEiOTm5tLTUzs5Ov9XS0nLYsGFfffXVjBkz+vfvf+bMmffff//jjz8e\nMWLErl27tm3bNmLEiFs/9I4vsbKyepi/TgCUT/b1VgDgPumvY3fhwoX6g3Z2do8//rj+cf3r\n2OXn58+ePdvT07NZs2Z2dnZdunRZvnx5UVGRfmZlZeXYsWMtLS0dHBzi4+PvMsD27dv/9re/\nmZmZOTo6BgUFFRQUuLm5de3aVb81Ly9vwoQJjo6OdnZ2AwYMOHDggE6ni4yMtLa2dnJyysnJ\nufVD7/iSB/CrBkDVNLr/O9QXAAAAisYxdgAAACrBMXYA8AfV1dXFxcUNz7Gysqq73QUANB4U\nOwD4g2+++eYf//hHw3O2bdsWEBBgmDwAcPc4xg4A/qCgoOCXX35peE6HDh1atmxpmDwAcPco\ndgAAACrByRMAAAAqQbEDAABQCYodAACASlDsAAAAVIJiBwAAoBIUOwAAAJWg2AEAAKgExQ4A\nAEAlKHYAAAAqQbEDAABQCYodAACASlDsAAAAVIJiBwAAoBIUOwAAAJWg2AEAAKgExQ4AAEAl\nKHYAAAAqQbEDAABQCYodAACASlDsAAAAVIJiBwAAoBIUOwAAAJWg2AEAAKgExQ4AAEAlKHYA\nAAAq8f8AvPRhroBZ2psAAAAASUVORK5CYII="
          },
          "metadata": {
            "image/png": {
              "width": 420,
              "height": 420
            }
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Finding phenotype and genotype associations with GWAS\n",
        "## GWAS {auto-animate=true}\n",
        "Genome-wide association studies (GWAS) — це аналіз генотипу в різних особин для пошуку асоціацій з фенотиповими ознаками.  \n",
        "Наші дані містять три SNP для дослідження GWAS (це дуже мало для простоти).\n",
        "  \n",
        "## Завантажуємо пакети, встановлюємо сід {auto-animate=true}\n",
        "\n",
        "```{r, echo=TRUE}\n",
        "# BiocManager::install(\"VariantAnnotation\")\n",
        "library(VariantAnnotation)\n",
        "library(rrBLUP)\n",
        "set.seed(1234)\n",
        "```\n",
        "  \n",
        "Сід необхідний для відтворюваності аналізу, коли він містить елемент випадковості.  \n",
        "  \n",
        "## Завантажуємо дані {auto-animate=true}\n",
        "\n",
        "```{r, echo=TRUE}\n",
        "vcf_file <- file.path(getwd(), \"datasets\", \"ch2\", \"small_sample.vcf\")\n",
        "vcf <- readVcf(vcf_file, \"hg19\")\n",
        "```\n",
        "  \n",
        "## Завантажуємо дані {auto-animate=true}\n",
        "\n",
        "```{r, echo=TRUE}\n",
        "vcf_file <- file.path(getwd(), \"datasets\", \"ch2\", \"small_sample.vcf\")\n",
        "vcf <- readVcf(vcf_file, \"hg19\")\n",
        "```\n",
        "  \n",
        "```{r}\n",
        "vcf\n",
        "```\n",
        "\n",
        "## Екстрагуємо генотипи, зразки, маркерні позиції {auto-animate=true}\n",
        "\n",
        "```{r, echo=TRUE}\n",
        "gts <- geno(vcf)$GT\n",
        "samples <- samples(header(vcf))\n",
        "markers <- rownames(gts)\n",
        "chrom <- as.character(seqnames(rowRanges(vcf)))\n",
        "pos <- as.numeric(start(rowRanges(vcf)))\n",
        "```\n",
        "\n",
        "## Визначаємо функцію {auto-animate=true}\n",
        "\n",
        "```{r, echo=TRUE}\n",
        "convert <- function(v){\n",
        "  v <- gsub(\"0/0\", 1, v)\n",
        "  v <- gsub(\"0/1\", 0, v)\n",
        "  v <- gsub(\"1/0\", 0, v)\n",
        "  v <- gsub(\"1/1\",-1, v)\n",
        "  return(v)\n",
        "}\n",
        "```\n",
        "\n",
        "Функція конвертує VCF генотипи в неопхідний для GWAS функції.\n",
        "\n",
        "## Застосовуємо функцію {auto-animate=true}\n",
        "\n",
        "```{r, echo=TRUE}\n",
        "gt_char <- apply(gts, convert, MARGIN = 2)\n",
        "```\n",
        "\n",
        "## Конвертуємо результат в числову матрицю {auto-animate=true}\n",
        "\n",
        "```{r, echo=TRUE}\n",
        "genotype_matrix <- matrix(as.numeric(gt_char), nrow(gt_char))\n",
        "colnames(genotype_matrix) <- samples\n",
        "```\n",
        "\n",
        "## Створюємо датафрейм варіантів {auto-animate=true}\n",
        "\n",
        "```{r, echo=TRUE}\n",
        "variant_info <- data.frame(marker = markers,\n",
        "                           chrom = chrom,\n",
        "                           pos = pos)\n",
        "```\n",
        "\n",
        "## Створюємо комбінований варіант/генотип датафрейм {auto-animate=true}\n",
        "\n",
        "```{r, echo=TRUE}\n",
        "genotypes <-  cbind(variant_info, as.data.frame(genotype_matrix))\n",
        "genotypes\n",
        "```\n",
        "\n",
        "## Створюємо датафрейм генотипів {auto-animate=true}\n",
        "\n",
        "```{r, echo=TRUE}\n",
        "phenotypes <- data.frame(line = samples,\n",
        "                         score = rnorm(length(samples)))\n",
        "\n",
        "phenotypes\n",
        "```\n",
        "\n",
        "## Власне GWAS {auto-animate=true}\n",
        "\n",
        "```{r, echo=TRUE}\n",
        "GWAS(phenotypes, genotypes, plot=FALSE)\n",
        "```"
      ],
      "metadata": {
        "id": "hKAwVffoHEX6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"VariantAnnotation\")\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"rrBLUP\")\n",
        "\n",
        "library(VariantAnnotation)\n",
        "library(rrBLUP)\n",
        "set.seed(1234)\n",
        "\n",
        "vcf_file <- file.path(getwd(), \"datasets\", \"ch2\", \"small_sample.vcf\")\n",
        "vcf <- readVcf(vcf_file, \"hg19\")\n",
        "gts <- geno(vcf)$GT\n",
        "\n",
        "\n",
        "samples <- samples(header(vcf))\n",
        "markers <- rownames(gts)\n",
        "chrom <- as.character(seqnames(rowRanges(vcf)))\n",
        "pos <- as.numeric(start(rowRanges(vcf)))\n",
        "\n",
        "convert <- function(v){\n",
        "  v <- gsub(\"0/0\", 1, v)\n",
        "  v <- gsub(\"0/1\", 0, v)\n",
        "  v <- gsub(\"1/0\", 0, v)\n",
        "  v <- gsub(\"1/1\",-1, v)\n",
        "  return(v)\n",
        "}\n",
        "gt_char<- apply(gts, convert, MARGIN = 2)\n",
        "\n",
        "genotype_matrix <- matrix(as.numeric(gt_char), nrow(gt_char) )\n",
        "colnames(genotype_matrix)<- samples\n",
        "\n",
        "variant_info <- data.frame(marker = markers,\n",
        "                           chrom = chrom,\n",
        "                           pos = pos)\n",
        "\n",
        "genotypes <-  cbind(variant_info, as.data.frame(genotype_matrix))\n",
        "genotypes\n",
        "\n",
        "phenotypes <- data.frame(\n",
        "  line = samples,\n",
        "  score = rnorm(length(samples))\n",
        ")\n",
        "\n",
        "phenotypes\n",
        "\n",
        "GWAS(phenotypes, genotypes,plot=FALSE)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "M4AZvM8wHTKT",
        "outputId": "cbe4cbd2-03c8-4801-b181-a6d08431ac6b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'BiocVersion', 'VariantAnnotation'\n",
            "\n",
            "also installing the dependencies ‘formatR’, ‘lambda.r’, ‘futile.options’, ‘filelock’, ‘abind’, ‘SparseArray’, ‘futile.logger’, ‘snow’, ‘BH’, ‘plogr’, ‘png’, ‘rjson’, ‘BiocFileCache’, ‘matrixStats’, ‘RCurl’, ‘GenomeInfoDbData’, ‘S4Arrays’, ‘DelayedArray’, ‘bitops’, ‘BiocParallel’, ‘RSQLite’, ‘KEGGREST’, ‘XML’, ‘GenomicAlignments’, ‘BiocIO’, ‘restfulr’, ‘biomaRt’, ‘BiocGenerics’, ‘MatrixGenerics’, ‘GenomeInfoDb’, ‘GenomicRanges’, ‘SummarizedExperiment’, ‘Rsamtools’, ‘zlibbioc’, ‘Biobase’, ‘S4Vectors’, ‘IRanges’, ‘XVector’, ‘Biostrings’, ‘AnnotationDbi’, ‘rtracklayer’, ‘BSgenome’, ‘GenomicFeatures’, ‘Rhtslib’\n",
            "\n",
            "\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'rrBLUP'\n",
            "\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "Loading required package: BiocGenerics\n",
            "\n",
            "\n",
            "Attaching package: ‘BiocGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    IQR, mad, sd, var, xtabs\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
            "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
            "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
            "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
            "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
            "    table, tapply, union, unique, unsplit, which.max, which.min\n",
            "\n",
            "\n",
            "Loading required package: MatrixGenerics\n",
            "\n",
            "Loading required package: matrixStats\n",
            "\n",
            "\n",
            "Attaching package: ‘MatrixGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
            "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
            "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
            "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
            "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
            "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
            "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
            "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
            "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
            "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
            "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
            "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
            "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
            "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
            "    rowWeightedSds, rowWeightedVars\n",
            "\n",
            "\n",
            "Loading required package: GenomeInfoDb\n",
            "\n",
            "Loading required package: S4Vectors\n",
            "\n",
            "Loading required package: stats4\n",
            "\n",
            "\n",
            "Attaching package: ‘S4Vectors’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:utils’:\n",
            "\n",
            "    findMatches\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    expand.grid, I, unname\n",
            "\n",
            "\n",
            "Loading required package: IRanges\n",
            "\n",
            "Loading required package: GenomicRanges\n",
            "\n",
            "Loading required package: SummarizedExperiment\n",
            "\n",
            "Loading required package: Biobase\n",
            "\n",
            "Welcome to Bioconductor\n",
            "\n",
            "    Vignettes contain introductory material; view with\n",
            "    'browseVignettes()'. To cite Bioconductor, see\n",
            "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘Biobase’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:MatrixGenerics’:\n",
            "\n",
            "    rowMedians\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    anyMissing, rowMedians\n",
            "\n",
            "\n",
            "Loading required package: Rsamtools\n",
            "\n",
            "Loading required package: Biostrings\n",
            "\n",
            "Loading required package: XVector\n",
            "\n",
            "\n",
            "Attaching package: ‘Biostrings’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:base’:\n",
            "\n",
            "    strsplit\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘VariantAnnotation’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:base’:\n",
            "\n",
            "    tabulate\n",
            "\n",
            "\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A data.frame: 3 × 6</caption>\n",
              "<thead>\n",
              "\t<tr><th scope=col>marker</th><th scope=col>chrom</th><th scope=col>pos</th><th scope=col>NA00001</th><th scope=col>NA00002</th><th scope=col>NA00003</th></tr>\n",
              "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><td>rs6054257     </td><td>20</td><td>  14370</td><td>1</td><td>0</td><td>-1</td></tr>\n",
              "\t<tr><td>20:17330_T/A  </td><td>20</td><td>  17330</td><td>1</td><td>0</td><td> 1</td></tr>\n",
              "\t<tr><td>20:1230237_T/G</td><td>20</td><td>1230237</td><td>1</td><td>1</td><td> 0</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA data.frame: 3 × 6\n\n| marker &lt;chr&gt; | chrom &lt;chr&gt; | pos &lt;dbl&gt; | NA00001 &lt;dbl&gt; | NA00002 &lt;dbl&gt; | NA00003 &lt;dbl&gt; |\n|---|---|---|---|---|---|\n| rs6054257      | 20 |   14370 | 1 | 0 | -1 |\n| 20:17330_T/A   | 20 |   17330 | 1 | 0 |  1 |\n| 20:1230237_T/G | 20 | 1230237 | 1 | 1 |  0 |\n\n",
            "text/latex": "A data.frame: 3 × 6\n\\begin{tabular}{llllll}\n marker & chrom & pos & NA00001 & NA00002 & NA00003\\\\\n <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n\\hline\n\t rs6054257      & 20 &   14370 & 1 & 0 & -1\\\\\n\t 20:17330\\_T/A   & 20 &   17330 & 1 & 0 &  1\\\\\n\t 20:1230237\\_T/G & 20 & 1230237 & 1 & 1 &  0\\\\\n\\end{tabular}\n",
            "text/plain": [
              "  marker         chrom pos     NA00001 NA00002 NA00003\n",
              "1 rs6054257      20      14370 1       0       -1     \n",
              "2 20:17330_T/A   20      17330 1       0        1     \n",
              "3 20:1230237_T/G 20    1230237 1       1        0     "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A data.frame: 3 × 2</caption>\n",
              "<thead>\n",
              "\t<tr><th scope=col>line</th><th scope=col>score</th></tr>\n",
              "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><td>NA00001</td><td>-1.2070657</td></tr>\n",
              "\t<tr><td>NA00002</td><td> 0.2774292</td></tr>\n",
              "\t<tr><td>NA00003</td><td> 1.0844412</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA data.frame: 3 × 2\n\n| line &lt;chr&gt; | score &lt;dbl&gt; |\n|---|---|\n| NA00001 | -1.2070657 |\n| NA00002 |  0.2774292 |\n| NA00003 |  1.0844412 |\n\n",
            "text/latex": "A data.frame: 3 × 2\n\\begin{tabular}{ll}\n line & score\\\\\n <chr> & <dbl>\\\\\n\\hline\n\t NA00001 & -1.2070657\\\\\n\t NA00002 &  0.2774292\\\\\n\t NA00003 &  1.0844412\\\\\n\\end{tabular}\n",
            "text/plain": [
              "  line    score     \n",
              "1 NA00001 -1.2070657\n",
              "2 NA00002  0.2774292\n",
              "3 NA00003  1.0844412"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[1] \"GWAS for trait: score\"\n",
            "[1] \"Variance components estimated. Testing markers.\"\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A data.frame: 3 × 4</caption>\n",
              "<thead>\n",
              "\t<tr><th></th><th scope=col>marker</th><th scope=col>chrom</th><th scope=col>pos</th><th scope=col>score</th></tr>\n",
              "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><th scope=row>1</th><td>rs6054257     </td><td>20</td><td>  14370</td><td>0.3010543</td></tr>\n",
              "\t<tr><th scope=row>2</th><td>20:17330_T/A  </td><td>20</td><td>  17330</td><td>0.3010057</td></tr>\n",
              "\t<tr><th scope=row>3</th><td>20:1230237_T/G</td><td>20</td><td>1230237</td><td>0.1655498</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA data.frame: 3 × 4\n\n| <!--/--> | marker &lt;chr&gt; | chrom &lt;chr&gt; | pos &lt;dbl&gt; | score &lt;dbl&gt; |\n|---|---|---|---|---|\n| 1 | rs6054257      | 20 |   14370 | 0.3010543 |\n| 2 | 20:17330_T/A   | 20 |   17330 | 0.3010057 |\n| 3 | 20:1230237_T/G | 20 | 1230237 | 0.1655498 |\n\n",
            "text/latex": "A data.frame: 3 × 4\n\\begin{tabular}{r|llll}\n  & marker & chrom & pos & score\\\\\n  & <chr> & <chr> & <dbl> & <dbl>\\\\\n\\hline\n\t1 & rs6054257      & 20 &   14370 & 0.3010543\\\\\n\t2 & 20:17330\\_T/A   & 20 &   17330 & 0.3010057\\\\\n\t3 & 20:1230237\\_T/G & 20 & 1230237 & 0.1655498\\\\\n\\end{tabular}\n",
            "text/plain": [
              "  marker         chrom pos     score    \n",
              "1 rs6054257      20      14370 0.3010543\n",
              "2 20:17330_T/A   20      17330 0.3010057\n",
              "3 20:1230237_T/G 20    1230237 0.1655498"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Finding unannotated transcribed regions\n",
        "Поширеною проблемою є пошук і підрахунок зчитувань, які вирівнялися за межами анотованих ділянок. В RNAseq-експерименті ці зчитування можуть представляти неанотовані гени і нові транскрипти. По суті, у нас є деякі гени, про які ми знаємо, і ми бачимо, що вони транскрибуються, оскільки вони мають вирівняне покриття зчитування, але інші транскрибовані ділянки не підпадають під\n",
        "і ми хочемо знати розташування вирівняних зчитувань, які їх представляють.\n",
        "що їх представляють. У цьому рецепті ми розглянемо оманливо просту техніку для\n",
        "пошуку таких областей.\n",
        "Підготовка\n",
        "Наш набір даних буде синтетичним, який містить невелику ділянку геному розміром 6 000 п.н. і дві генні і третю неанотовану ділянку з вирівнюючими зчитуваннями.\n",
        "Нам знадобляться бібліотеки Bioconductor csaw, IRanges, SummarizedExperiment та\n",
        "rtracklayer, а також деякі функції з інших пакунків, що входять до складу базового Bioconductor. Дані знаходяться у сховищі даних цієї книги\n",
        "у файлах datasets/ch1/windows.bam та datasets/ch1/genes.gff\n",
        "Як це зробити...\n",
        "Аналіз потужності за допомогою powsimR можна виконати у наступні кроки:\n",
        "# 1 Налаштуйте функцію завантаження:\n",
        "```{r, echo=TRUE}\n",
        "get_annotated_regions_from_gff <- function(file_name) {\n",
        " gff <- rtracklayer::import.gff(назва_файлу)\n",
        " as(gff, \"GRanges\")}\n",
        " ```\n",
        " # 2 Отримати підрахунки у вікнах по всьому геному:\n",
        "```{r, echo=TRUE}\n",
        "whole_genome <- csaw::windowCounts(\n",
        " file.path(getwd(), \"datasets\", \"ch1\", \"windows.bam\"),\n",
        " bin = TRUE,\n",
        " filter = 0,\n",
        " width = 500,\n",
        " param = csaw::readParam(\n",
        " minq = 20,\n",
        " dedup = TRUE,\n",
        " pe = \"both\"\n",
        " )\n",
        ")\n",
        "colnames(whole_genome) <- c(\"small_data\")\n",
        "annotated_regions <-\n",
        "get_annotated_regions_from_gff(file.path(getwd(), \"datasets\",\n",
        "\"ch1\", \"genes.gff\"))\n",
        "```\n",
        "# 3 Знайдіть перекриття між анотаціями та нашими вікнами і створіть підмножину вікон:\n",
        "```{r, echo=TRUE}\n",
        "library(IRanges)\n",
        "library(SummarizedExperiment)\n",
        "windows_in_genes <-IRanges::overlapsAny(\n",
        "SummarizedExperiment::rowRanges(whole_genome), annotated_regions )\n",
        " ```\n",
        "# 4 Розбийте вікна на вікна в анотованих і неанотованих регіонах:\n",
        "```{r, echo=TRUE}\n",
        "annotated_window_counts <- whole_genome[windows_in_genes,]\n",
        "non_annotated_window_counts <- whole_genome[ ! windows_in_genes,]\n",
        "```\n",
        "# 5 Виводимо дані у матрицю підрахунків:\n",
        "```{r, echo=TRUE}\n",
        "assay(non_annotated_window_counts)\n",
        "```"
      ],
      "metadata": {
        "id": "vCGFECruRw_G"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"csaw\")\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"IRanges\")\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"SummarizedExperiment\")\n",
        "\n",
        "get_annotated_regions_from_gff <- function(file_name) {\n",
        "  gff <- rtracklayer::import.gff(file_name)\n",
        "  as(gff, \"GRanges\")\n",
        "}\n",
        "\n",
        "get_annotated_regions_from_bed <- function(file_name){\n",
        "  bed <- rtracklayer::import.bed(file_name)\n",
        "  as(bed, \"GRanges\")\n",
        "}\n",
        "\n",
        "#get counts in windows across whole genome\n",
        "whole_genome <- csaw::windowCounts(\n",
        "    file.path(getwd(), \"datasets\", \"ch1\", \"windows.bam\"),\n",
        "    bin = TRUE,\n",
        "    filter = 0,\n",
        "    width = 500,\n",
        "    param = csaw::readParam(\n",
        "        minq = 20,\n",
        "        dedup = TRUE,\n",
        "        pe = \"both\"\n",
        "    )\n",
        ")\n",
        "colnames(whole_genome) <- c(\"small_data\")\n",
        "\n",
        "#make a GRanges object of the annotated features\n",
        "annotated_regions <- get_annotated_regions_from_gff(file.path(getwd(), \"datasets\", \"ch1\", \"genes.gff\"))\n",
        "\n",
        "\n",
        "library(IRanges)\n",
        "library(SummarizedExperiment)\n",
        "#find the overlaps between the windows we made\n",
        "windows_in_genes <-IRanges::overlapsAny(\n",
        "  SummarizedExperiment::rowRanges(whole_genome),\n",
        "  annotated_regions\n",
        ")\n",
        "\n",
        "annotated_window_counts <- whole_genome[windows_in_genes,]\n",
        "non_annotated_window_counts <- whole_genome[ ! windows_in_genes,]\n",
        "\n",
        "\n",
        "assay(annotated_window_counts)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "4D8aeB8DSzVU",
        "outputId": "c472b809-5665-46e1-8475-e90afa98da0e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'csaw'\n",
            "\n",
            "also installing the dependencies ‘statmod’, ‘edgeR’, ‘limma’, ‘metapod’\n",
            "\n",
            "\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'IRanges'”\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'SummarizedExperiment'”\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A matrix: 6 × 1 of type int</caption>\n",
              "<thead>\n",
              "\t<tr><th scope=col>small_data</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><td> 0</td></tr>\n",
              "\t<tr><td>31</td></tr>\n",
              "\t<tr><td>25</td></tr>\n",
              "\t<tr><td> 0</td></tr>\n",
              "\t<tr><td>28</td></tr>\n",
              "\t<tr><td>31</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA matrix: 6 × 1 of type int\n\n| small_data |\n|---|\n|  0 |\n| 31 |\n| 25 |\n|  0 |\n| 28 |\n| 31 |\n\n",
            "text/latex": "A matrix: 6 × 1 of type int\n\\begin{tabular}{l}\n small\\_data\\\\\n\\hline\n\t  0\\\\\n\t 31\\\\\n\t 25\\\\\n\t  0\\\\\n\t 28\\\\\n\t 31\\\\\n\\end{tabular}\n",
            "text/plain": [
              "     small_data\n",
              "[1,]  0        \n",
              "[2,] 31        \n",
              "[3,] 25        \n",
              "[4,]  0        \n",
              "[5,] 28        \n",
              "[6,] 31        "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Finding regions showing high expression abinitio with bumphunter\n",
        "Пошук ділянок вирівнювання зчитування, які походять від однієї і тієї ж, потенційно неанотованої, геномної особливості, є поширеним завданням.\n",
        "Мета полягає в тому, щоб згрупувати вирівнювання зчитувань таким чином, щоб ми могли позначити регіони, які мають значне покриття, а потім порівняти зразки для виявлення відмінностей у рівнях експресії.\n",
        "Бібліотеки: Rsamtools та bumphunter\n",
        "Дані: datasets/ch1/windows.bam.\n",
        "\n",
        "# 1 Завантажте дані та отримайте покриття для кожної позиції:\n",
        "```\n",
        "library(Rsamtools)\n",
        "library(bumphunter)\n",
        "pileup_df <- Rsamtools::pileup(file.path(getwd(), \"datasets\",\n",
        "\"ch1\", \"windows.bam\"))\n",
        "```\n",
        "# 2 Знаходимо попередні кластери:\n",
        "```\n",
        "clusters <- bumphunter::clusterMaker(pileup_df$seqnames,\n",
        "pileup_df$pos, maxGap = 100)\n",
        "```\n",
        "# 3 Знаходимо бампери з мінімальним відсіканням:\n",
        "```\n",
        "bumphunter::regionFinder(pileup_df$count, pileup_df$seqnames,\n",
        "pileup_df$pos, clusters, cutoff=1)\n",
        "```\n"
      ],
      "metadata": {
        "id": "GQtU2Q65VEnC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"Rsamtools\")\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"bumphunter\")\n",
        "library(Rsamtools)\n",
        "library(bumphunter)\n",
        "\n",
        "pileup_df <- Rsamtools::pileup(file.path(getwd(), \"datasets\", \"ch1\", \"windows.bam\"))\n",
        "clusters <- bumphunter::clusterMaker(pileup_df$seqnames, pileup_df$pos, maxGap = 100)\n",
        "table(clusters)\n",
        "\n",
        "bumphunter::regionFinder(pileup_df$count, pileup_df$seqnames, pileup_df$pos, clusters, cutoff=1)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "GuRt-BRJU3vi",
        "outputId": "c9a69db6-bb68-4e6f-d331-7109e7ba3948"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'Rsamtools'”\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'bumphunter'\n",
            "\n",
            "also installing the dependencies ‘rngtools’, ‘doRNG’\n",
            "\n",
            "\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "Loading required package: foreach\n",
            "\n",
            "Loading required package: iterators\n",
            "\n",
            "Loading required package: parallel\n",
            "\n",
            "Loading required package: locfit\n",
            "\n",
            "locfit 1.5-9.8 \t 2023-06-11\n",
            "\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "clusters\n",
              "   1    2    3 \n",
              "1486 1552 1520 "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "getSegments: segmenting\n",
            "\n",
            "getSegments: splitting\n",
            "\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A data.frame: 3 × 10</caption>\n",
              "<thead>\n",
              "\t<tr><th></th><th scope=col>chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>value</th><th scope=col>area</th><th scope=col>cluster</th><th scope=col>indexStart</th><th scope=col>indexEnd</th><th scope=col>L</th><th scope=col>clusterL</th></tr>\n",
              "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><th scope=row>3</th><td>Chr1</td><td>4503</td><td>5500</td><td>10.401974</td><td>15811</td><td>3</td><td>3039</td><td>4558</td><td>1520</td><td>1520</td></tr>\n",
              "\t<tr><th scope=row>1</th><td>Chr1</td><td> 502</td><td>1500</td><td> 9.985868</td><td>14839</td><td>1</td><td>   1</td><td>1486</td><td>1486</td><td>1486</td></tr>\n",
              "\t<tr><th scope=row>2</th><td>Chr1</td><td>2501</td><td>3500</td><td> 8.657216</td><td>13436</td><td>2</td><td>1487</td><td>3038</td><td>1552</td><td>1552</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA data.frame: 3 × 10\n\n| <!--/--> | chr &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | value &lt;dbl&gt; | area &lt;int&gt; | cluster &lt;dbl&gt; | indexStart &lt;int&gt; | indexEnd &lt;int&gt; | L &lt;dbl&gt; | clusterL &lt;int&gt; |\n|---|---|---|---|---|---|---|---|---|---|---|\n| 3 | Chr1 | 4503 | 5500 | 10.401974 | 15811 | 3 | 3039 | 4558 | 1520 | 1520 |\n| 1 | Chr1 |  502 | 1500 |  9.985868 | 14839 | 1 |    1 | 1486 | 1486 | 1486 |\n| 2 | Chr1 | 2501 | 3500 |  8.657216 | 13436 | 2 | 1487 | 3038 | 1552 | 1552 |\n\n",
            "text/latex": "A data.frame: 3 × 10\n\\begin{tabular}{r|llllllllll}\n  & chr & start & end & value & area & cluster & indexStart & indexEnd & L & clusterL\\\\\n  & <fct> & <int> & <int> & <dbl> & <int> & <dbl> & <int> & <int> & <dbl> & <int>\\\\\n\\hline\n\t3 & Chr1 & 4503 & 5500 & 10.401974 & 15811 & 3 & 3039 & 4558 & 1520 & 1520\\\\\n\t1 & Chr1 &  502 & 1500 &  9.985868 & 14839 & 1 &    1 & 1486 & 1486 & 1486\\\\\n\t2 & Chr1 & 2501 & 3500 &  8.657216 & 13436 & 2 & 1487 & 3038 & 1552 & 1552\\\\\n\\end{tabular}\n",
            "text/plain": [
              "  chr  start end  value     area  cluster indexStart indexEnd L    clusterL\n",
              "3 Chr1 4503  5500 10.401974 15811 3       3039       4558     1520 1520    \n",
              "1 Chr1  502  1500  9.985868 14839 1          1       1486     1486 1486    \n",
              "2 Chr1 2501  3500  8.657216 13436 2       1487       3038     1552 1552    "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Finding protein domains with PFAM and bio3d\n"
      ],
      "metadata": {
        "id": "Aa9AXdG85WYW"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**1.Завантажння та встановлення необхідних пакетів.**"
      ],
      "metadata": {
        "id": "CODPhdYZZbAs"
      }
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "Dtgh7_6ZRIOB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"org.Hs.eg.db\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aSQ7PNNL51QS",
        "outputId": "e44ee99b-3516-4999-b8a3-4aa953401fd2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'BiocVersion', 'org.Hs.eg.db'\n",
            "\n",
            "also installing the dependencies ‘bitops’, ‘zlibbioc’, ‘RCurl’, ‘GenomeInfoDbData’, ‘XVector’, ‘GenomeInfoDb’, ‘plogr’, ‘png’, ‘Biostrings’, ‘BiocGenerics’, ‘Biobase’, ‘IRanges’, ‘RSQLite’, ‘S4Vectors’, ‘KEGGREST’, ‘AnnotationDbi’\n",
            "\n",
            "\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "BiocManager::install(\"org.EcK12.eg.db\")"
      ],
      "metadata": {
        "id": "TgEy_k6T9fIi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "BiocManager::install(\"org.At.tair.db\")"
      ],
      "metadata": {
        "id": "MZ2j5KI9-TLm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "install.packages(\"devtools\")\n",
        "library(devtools)\n",
        "install_bitbucket(\"Grantlab/bio3d\", subdir = \"ver_devel/bio3d/\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "eoi9xuyF-dfW",
        "outputId": "8c68744f-de6a-4134-d638-4db9d147b7b7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "also installing the dependencies ‘Rcpp’, ‘httpuv’, ‘xtable’, ‘sourcetools’, ‘later’, ‘promises’, ‘shiny’, ‘downlit’, ‘htmlwidgets’, ‘miniUI’, ‘pkgdown’, ‘profvis’, ‘urlchecker’\n",
            "\n",
            "\n",
            "Loading required package: usethis\n",
            "\n",
            "\n",
            "Attaching package: ‘devtools’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:BiocManager’:\n",
            "\n",
            "    install\n",
            "\n",
            "\n",
            "Downloading bitbucket repo Grantlab/bio3d@HEAD\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\u001b[36m──\u001b[39m \u001b[36mR CMD build\u001b[39m \u001b[36m─────────────────────────────────────────────────────────────────\u001b[39m\n",
            "* checking for file ‘/tmp/RtmpHRgGuv/remotes39428e9d9f4/Grantlab-bio3d-9686c49cf36d/ver_devel/bio3d/DESCRIPTION’ ... OK\n",
            "* preparing ‘bio3d’:\n",
            "* checking DESCRIPTION meta-information ... OK\n",
            "* cleaning src\n",
            "* checking for LF line-endings in source and make files and shell scripts\n",
            "* checking for empty or unneeded directories\n",
            "* building ‘bio3d_2.4-4.9000.tar.gz’\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "utils::download.file(url = \"https://bitbucket.org/Grantlab/bio3d/get/HEAD.tar.gz\",\n",
        "destfile = \"bio3d.tar.gz\", method = \"curl\")"
      ],
      "metadata": {
        "id": "8O1YyQlt-3JV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**2.Перевірка вмісту бази даних**\n",
        "\n",
        "• Функція keytypes() містить список доступних ключів пошуку. Використовується\n",
        "для того, щоб перевірити, чи містить база даних інформацію про домени\n",
        "PFAM. Вихідні дані відображають різні ключі пошуку, включаючи \"PFAM\".\n",
        "Отримуємо вектор ключів за допомогою функції keys()."
      ],
      "metadata": {
        "id": "35UuvowNZ6I8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "library(org.Hs.eg.db)\n",
        "keytypes(org.Hs.eg.db)\n",
        "k <- head(keys(org.Hs.eg.db, keytype = \"ENSEMBL\"), n = 3 )"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 69
        },
        "id": "ijSvV93VC1vp",
        "outputId": "26a0adcb-8ced-42e3-8e8e-974708cf7cf8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<style>\n",
              ".list-inline {list-style: none; margin:0; padding: 0}\n",
              ".list-inline>li {display: inline-block}\n",
              ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
              "</style>\n",
              "<ol class=list-inline><li>'ACCNUM'</li><li>'ALIAS'</li><li>'ENSEMBL'</li><li>'ENSEMBLPROT'</li><li>'ENSEMBLTRANS'</li><li>'ENTREZID'</li><li>'ENZYME'</li><li>'EVIDENCE'</li><li>'EVIDENCEALL'</li><li>'GENENAME'</li><li>'GENETYPE'</li><li>'GO'</li><li>'GOALL'</li><li>'IPI'</li><li>'MAP'</li><li>'OMIM'</li><li>'ONTOLOGY'</li><li>'ONTOLOGYALL'</li><li>'PATH'</li><li>'PFAM'</li><li>'PMID'</li><li>'PROSITE'</li><li>'REFSEQ'</li><li>'SYMBOL'</li><li>'UCSCKG'</li><li>'UNIPROT'</li></ol>\n"
            ],
            "text/markdown": "1. 'ACCNUM'\n2. 'ALIAS'\n3. 'ENSEMBL'\n4. 'ENSEMBLPROT'\n5. 'ENSEMBLTRANS'\n6. 'ENTREZID'\n7. 'ENZYME'\n8. 'EVIDENCE'\n9. 'EVIDENCEALL'\n10. 'GENENAME'\n11. 'GENETYPE'\n12. 'GO'\n13. 'GOALL'\n14. 'IPI'\n15. 'MAP'\n16. 'OMIM'\n17. 'ONTOLOGY'\n18. 'ONTOLOGYALL'\n19. 'PATH'\n20. 'PFAM'\n21. 'PMID'\n22. 'PROSITE'\n23. 'REFSEQ'\n24. 'SYMBOL'\n25. 'UCSCKG'\n26. 'UNIPROT'\n\n\n",
            "text/latex": "\\begin{enumerate*}\n\\item 'ACCNUM'\n\\item 'ALIAS'\n\\item 'ENSEMBL'\n\\item 'ENSEMBLPROT'\n\\item 'ENSEMBLTRANS'\n\\item 'ENTREZID'\n\\item 'ENZYME'\n\\item 'EVIDENCE'\n\\item 'EVIDENCEALL'\n\\item 'GENENAME'\n\\item 'GENETYPE'\n\\item 'GO'\n\\item 'GOALL'\n\\item 'IPI'\n\\item 'MAP'\n\\item 'OMIM'\n\\item 'ONTOLOGY'\n\\item 'ONTOLOGYALL'\n\\item 'PATH'\n\\item 'PFAM'\n\\item 'PMID'\n\\item 'PROSITE'\n\\item 'REFSEQ'\n\\item 'SYMBOL'\n\\item 'UCSCKG'\n\\item 'UNIPROT'\n\\end{enumerate*}\n",
            "text/plain": [
              " [1] \"ACCNUM\"       \"ALIAS\"        \"ENSEMBL\"      \"ENSEMBLPROT\"  \"ENSEMBLTRANS\"\n",
              " [6] \"ENTREZID\"     \"ENZYME\"       \"EVIDENCE\"     \"EVIDENCEALL\"  \"GENENAME\"    \n",
              "[11] \"GENETYPE\"     \"GO\"           \"GOALL\"        \"IPI\"          \"MAP\"         \n",
              "[16] \"OMIM\"         \"ONTOLOGY\"     \"ONTOLOGYALL\"  \"PATH\"         \"PFAM\"        \n",
              "[21] \"PMID\"         \"PROSITE\"      \"REFSEQ\"       \"SYMBOL\"       \"UCSCKG\"      \n",
              "[26] \"UNIPROT\"     "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**3.Запит інформації про домен**\n",
        "\n",
        "• Використовуючи функцію select(), надсилаємо запит до бази даних\n",
        "для отримання інформації про домен на основі наданих ключів.\n",
        "Зокрема, мета полягає в тому, щоб отримати ідентифікатори PFAM,\n",
        "пов’язаний з вибраними генами."
      ],
      "metadata": {
        "id": "xoPIDgu3aKGf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "result <- select(org.Hs.eg.db, keys = k, keytype=\"ENSEMBL\", columns\n",
        "= c(\"PFAM\"))\n",
        "result"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 415
        },
        "id": "2c88OdpkDqYZ",
        "outputId": "bdf9f914-0c11-4bb0-c01b-0de01dc34624"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'select()' returned 1:many mapping between keys and columns\n",
            "\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A data.frame: 9 × 2</caption>\n",
              "<thead>\n",
              "\t<tr><th scope=col>ENSEMBL</th><th scope=col>PFAM</th></tr>\n",
              "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><td>ENSG00000121410</td><td>PF13895</td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF07678</td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF17791</td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF07677</td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF00207</td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF17789</td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF07703</td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF01835</td></tr>\n",
              "\t<tr><td>ENSG00000291190</td><td>NA     </td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA data.frame: 9 × 2\n\n| ENSEMBL &lt;chr&gt; | PFAM &lt;chr&gt; |\n|---|---|\n| ENSG00000121410 | PF13895 |\n| ENSG00000175899 | PF07678 |\n| ENSG00000175899 | PF17791 |\n| ENSG00000175899 | PF07677 |\n| ENSG00000175899 | PF00207 |\n| ENSG00000175899 | PF17789 |\n| ENSG00000175899 | PF07703 |\n| ENSG00000175899 | PF01835 |\n| ENSG00000291190 | NA      |\n\n",
            "text/latex": "A data.frame: 9 × 2\n\\begin{tabular}{ll}\n ENSEMBL & PFAM\\\\\n <chr> & <chr>\\\\\n\\hline\n\t ENSG00000121410 & PF13895\\\\\n\t ENSG00000175899 & PF07678\\\\\n\t ENSG00000175899 & PF17791\\\\\n\t ENSG00000175899 & PF07677\\\\\n\t ENSG00000175899 & PF00207\\\\\n\t ENSG00000175899 & PF17789\\\\\n\t ENSG00000175899 & PF07703\\\\\n\t ENSG00000175899 & PF01835\\\\\n\t ENSG00000291190 & NA     \\\\\n\\end{tabular}\n",
            "text/plain": [
              "  ENSEMBL         PFAM   \n",
              "1 ENSG00000121410 PF13895\n",
              "2 ENSG00000175899 PF07678\n",
              "3 ENSG00000175899 PF17791\n",
              "4 ENSG00000175899 PF07677\n",
              "5 ENSG00000175899 PF00207\n",
              "6 ENSG00000175899 PF17789\n",
              "7 ENSG00000175899 PF07703\n",
              "8 ENSG00000175899 PF01835\n",
              "9 ENSG00000291190 NA     "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**4.Зіставлення ідентифікаторів PFAM з описами:**\n",
        "\n",
        "• Завантажуючи пакет PFAM.db, отримуємо доступ до об’єкта\n",
        "PFAMDE, який забезпечує зв’язок між ідентифікаторами PFAM та\n",
        "їхніми описами.\n",
        "\n",
        "• Вихідні дані демонструють список генів із відповідними\n",
        "ідентифікаторами PFAM та білками, для яких ці домени\n",
        "характерні."
      ],
      "metadata": {
        "id": "Ux_5w_8RaSIY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "\n",
        "BiocManager::install(\"PFAM.db\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "--fGzkkdQcbc",
        "outputId": "80e6b3f4-c971-429e-ccb5-4255bfa921a8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'PFAM.db'\n",
            "\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "library(PFAM.db)\n",
        "descriptions <- PFAMDE"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KtN0UHg3QmfD",
        "outputId": "b6fb584f-68d3-4bc4-cb8e-753526d36d65"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**5.Отримання описів:**\n",
        "\n",
        "• Функція mappedkeys() отримує фактичні описи ідентифікаторів PFAM."
      ],
      "metadata": {
        "id": "0s_aRHVdagmI"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "all_ids <- mappedkeys(descriptions)"
      ],
      "metadata": {
        "id": "gQ9s7aGiQvbL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**6.Перетворення на data frame:**\n",
        "\n",
        "• Описи з попереднього кроку перетворюються на фрейм даних для\n",
        "полегшеного маніпулювання та об’єднання на наступному кроці."
      ],
      "metadata": {
        "id": "HY2glBDTapZp"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "id_description_mapping <- as.data.frame(descriptions[all_ids])"
      ],
      "metadata": {
        "id": "cL3shR1SQw7M"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "BiocManager::install(\"Organism.dplyr\")"
      ],
      "metadata": {
        "id": "hGvq8-pDQ1JT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**7.Об’єднання описів за допомогою ідентифікаторів**:\n",
        "\n",
        "• Функція dplyr::left_join співставляє та прикріплює описи доменів\n",
        "PFAM до їхніх відповідних ідентифікаторів.\n",
        "\n",
        "• В результаті отримуємо таблицю з доменами PFAM,\n",
        "відповідними генами, що їх кодують та білками, для яких ці\n",
        "домени характерні."
      ],
      "metadata": {
        "id": "gcblii37a6Tp"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dplyr::left_join(result, id_description_mapping, by = c(\"PFAM\" = \"ac\"))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 380
        },
        "id": "A_AwWPz6Q87A",
        "outputId": "c3594236-a83f-41c1-8153-977431708322"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A data.frame: 9 × 3</caption>\n",
              "<thead>\n",
              "\t<tr><th scope=col>ENSEMBL</th><th scope=col>PFAM</th><th scope=col>de</th></tr>\n",
              "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><td>ENSG00000121410</td><td>PF13895</td><td>Immunoglobulin domain                   </td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF07678</td><td>A-macroglobulin TED domain              </td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF17791</td><td>Macroglobulin domain MG3                </td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF07677</td><td>A-macroglobulin receptor binding domain </td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF00207</td><td>Alpha-2-macroglobulin family            </td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF17789</td><td>Macroglobulin domain MG4                </td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF07703</td><td>Alpha-2-macroglobulin bait region domain</td></tr>\n",
              "\t<tr><td>ENSG00000175899</td><td>PF01835</td><td>MG2 domain                              </td></tr>\n",
              "\t<tr><td>ENSG00000291190</td><td>NA     </td><td>NA                                      </td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA data.frame: 9 × 3\n\n| ENSEMBL &lt;chr&gt; | PFAM &lt;chr&gt; | de &lt;chr&gt; |\n|---|---|---|\n| ENSG00000121410 | PF13895 | Immunoglobulin domain                    |\n| ENSG00000175899 | PF07678 | A-macroglobulin TED domain               |\n| ENSG00000175899 | PF17791 | Macroglobulin domain MG3                 |\n| ENSG00000175899 | PF07677 | A-macroglobulin receptor binding domain  |\n| ENSG00000175899 | PF00207 | Alpha-2-macroglobulin family             |\n| ENSG00000175899 | PF17789 | Macroglobulin domain MG4                 |\n| ENSG00000175899 | PF07703 | Alpha-2-macroglobulin bait region domain |\n| ENSG00000175899 | PF01835 | MG2 domain                               |\n| ENSG00000291190 | NA      | NA                                       |\n\n",
            "text/latex": "A data.frame: 9 × 3\n\\begin{tabular}{lll}\n ENSEMBL & PFAM & de\\\\\n <chr> & <chr> & <chr>\\\\\n\\hline\n\t ENSG00000121410 & PF13895 & Immunoglobulin domain                   \\\\\n\t ENSG00000175899 & PF07678 & A-macroglobulin TED domain              \\\\\n\t ENSG00000175899 & PF17791 & Macroglobulin domain MG3                \\\\\n\t ENSG00000175899 & PF07677 & A-macroglobulin receptor binding domain \\\\\n\t ENSG00000175899 & PF00207 & Alpha-2-macroglobulin family            \\\\\n\t ENSG00000175899 & PF17789 & Macroglobulin domain MG4                \\\\\n\t ENSG00000175899 & PF07703 & Alpha-2-macroglobulin bait region domain\\\\\n\t ENSG00000175899 & PF01835 & MG2 domain                              \\\\\n\t ENSG00000291190 & NA      & NA                                      \\\\\n\\end{tabular}\n",
            "text/plain": [
              "  ENSEMBL         PFAM    de                                      \n",
              "1 ENSG00000121410 PF13895 Immunoglobulin domain                   \n",
              "2 ENSG00000175899 PF07678 A-macroglobulin TED domain              \n",
              "3 ENSG00000175899 PF17791 Macroglobulin domain MG3                \n",
              "4 ENSG00000175899 PF07677 A-macroglobulin receptor binding domain \n",
              "5 ENSG00000175899 PF00207 Alpha-2-macroglobulin family            \n",
              "6 ENSG00000175899 PF17789 Macroglobulin domain MG4                \n",
              "7 ENSG00000175899 PF07703 Alpha-2-macroglobulin bait region domain\n",
              "8 ENSG00000175899 PF01835 MG2 domain                              \n",
              "9 ENSG00000291190 NA      NA                                      "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Aligning genomic length sequences with DECIPHER\n"
      ],
      "metadata": {
        "id": "UunvOMY1Rxb6"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Завантаження пакетів. Інсталюємо необхідні для роботи пакети (BiocManager,\n",
        "DECIPHER)"
      ],
      "metadata": {
        "id": "kXCkb8WxTADr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"DECIPHER\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XFEF0wbOR9hh",
        "outputId": "a80f31b6-5dc4-4eb3-9e5b-7854ec69d56c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'DECIPHER'\n",
            "\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Завантаження послідовності. Імпортуємо dataset з досліджуваними\n",
        "послідовностями в об’єкт під назвою long_seqs, тип - DNAStringSet."
      ],
      "metadata": {
        "id": "8fIj2rDlYDU0"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "library(DECIPHER)"
      ],
      "metadata": {
        "id": "CHiCeBXpY5lk"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "long_seqs <- readDNAStringSet(file.path(getwd(),\"Desktop\", \"datasets\", \"ch3\",\"plastid_genomes.fa\"))\n",
        "long_seqs"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 488
        },
        "id": "dx3RMu1Jbi4f",
        "outputId": "70fd1d8f-e86a-4ed7-d026-458681bf70fe"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ERROR",
          "evalue": "ignored",
          "traceback": [
            "Error in .Call2(\"new_input_filexp\", filepath, PACKAGE = \"XVector\"): cannot open file '/content/Desktop/datasets/ch3/plastid_genomes.fa'\nTraceback:\n",
            "1. readDNAStringSet(file.path(getwd(), \"Desktop\", \"datasets\", \"ch3\", \n .     \"plastid_genomes.fa\"))",
            "2. .read_XStringSet(filepath, format, nrec, skip, seek.first.rec, \n .     use.names, \"DNA\", with.qualities)",
            "3. fasta.index(filepath, nrec = nrec, skip = skip, seek.first.rec = seek.first.rec, \n .     seqtype = seqtype)",
            "4. open_input_files(filepath)",
            "5. lapply(filepath2, function(fp) {\n .     filexp <- new_input_filexp(fp)\n .     reg.finalizer(filexp, close_filexp, onexit = TRUE)\n .     filexp\n . })",
            "6. lapply(filepath2, function(fp) {\n .     filexp <- new_input_filexp(fp)\n .     reg.finalizer(filexp, close_filexp, onexit = TRUE)\n .     filexp\n . })",
            "7. FUN(X[[i]], ...)",
            "8. new_input_filexp(fp)",
            "9. .Call2(\"new_input_filexp\", filepath, PACKAGE = \"XVector\")"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Machine learning for novel feature detection in proteins**"
      ],
      "metadata": {
        "id": "FqUWtMC7evTn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"Biostrings\")\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"kebabs\")\n",
        "library(kebabs)\n",
        "library(Biostrings)\n",
        "\n",
        "seqs <- readAAStringSet(file.path(getwd(), \"datasets\", \"ch3\", \"ecoli_proteins.fa\"))\n",
        "classes <- readr::read_csv(file.path(getwd(), \"datasets\", \"ch3\", \"ecoli_protein_classes.txt\"), col_names = TRUE)\n",
        "classes <- classes$class\n",
        "\n",
        "\n",
        "num_seqs <- length(seqs)\n",
        "\n",
        "training_proportion <- 0.75\n",
        "training_set_indices <- sample(1:num_seqs, training_proportion * num_seqs)\n",
        "test_set_indices <- c(1:num_seqs)[-training_set_indices]\n",
        "\n",
        "kernel <- gappyPairKernel(k=1, m=3)\n",
        "\n",
        "model <- kbsvm(x=seqs[training_set_indices], y=classes[training_set_indices], kernel=kernel, pkg=\"e1071\", svm=\"C-svc\", cost=15)\n",
        "\n",
        "predictions <- predict(model, seqs[test_set_indices])\n",
        "evaluatePrediction(predictions, classes[test_set_indices], allLabels=c(1,-1) )\n",
        "\n",
        "\n",
        "##predict sites in sequence\n",
        "\n",
        "seq_to_test <- seqs[[1]][1:10]\n",
        "seq_to_test\n",
        "\n",
        "prediction_profile <-getPredictionProfile(seq_to_test, kernel, featureWeights(model), modelOffset(model) )\n",
        "plot(prediction_profile)"
      ],
      "metadata": {
        "id": "2GOX4cjHeskm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "  install.packages(\"BiocManager\")\n",
        "  BiocManager::install(\"kebabs\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5NLmQoMNkeq4",
        "outputId": "4d15ee25-7308-4633-9c8a-d03616747ab5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'kebabs'\n",
            "\n",
            "also installing the dependencies ‘proxy’, ‘kernlab’, ‘e1071’, ‘LiblineaR’, ‘apcluster’\n",
            "\n",
            "\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "  install.packages(\"BiocManager\")\n",
        "  BiocManager::install(\"Biostrings\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XtWfZQK3e7lc",
        "outputId": "4efb00ad-52e2-4bfe-d27d-324d191bd9e7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'Biostrings'”\n",
            "Old packages: 'bit', 'brio', 'bslib', 'cli', 'cpp11', 'curl', 'data.table',\n",
            "  'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven', 'highr', 'isoband',\n",
            "  'jsonlite', 'openssl', 'pkgbuild', 'processx', 'progress', 'ragg',\n",
            "  'roxygen2', 'sass', 'scales', 'stringi', 'svglite', 'testthat',\n",
            "  'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml', 'boot',\n",
            "  'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "library(kebabs)\n",
        "library(Biostrings)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "I5NT74RzlIDO",
        "outputId": "c5eb0574-9f60-430c-e2b3-112ad32e5502"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Loading required package: kernlab\n",
            "\n",
            "\n",
            "Attaching package: ‘kernlab’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:Biostrings’:\n",
            "\n",
            "    type\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:BiocGenerics’:\n",
            "\n",
            "    type\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "seqs <- readAAStringSet(file.path(setwd(\"C:/Users/my comp2/OneDrive/Рабочий стол\"), \"ecoli_proteins.fa\"))\n",
        "classes <- readr::read_csv(file.path(setwd(\"C:/Users/my comp2/OneDrive/Рабочий стол\"), \"ecoli_protein_classes.txt\"), col_names = TRUE)\n",
        "classes <- classes$class"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 262
        },
        "id": "ats3uJEBlpk1",
        "outputId": "4568a0e7-f065-442d-da01-c32a71a12e79"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ERROR",
          "evalue": "ignored",
          "traceback": [
            "Error in setwd(\"C:/Users/my comp2/OneDrive/Рабочий стол\"): cannot change working directory\nTraceback:\n",
            "1. readAAStringSet(file.path(setwd(\"C:/Users/my comp2/OneDrive/Рабочий стол\"), \n .     \"ecoli_proteins.fa\"))",
            "2. .read_XStringSet(filepath, format, nrec, skip, seek.first.rec, \n .     use.names, \"AA\", with.qualities)",
            "3. .is_filexp_list(filepath)",
            "4. file.path(setwd(\"C:/Users/my comp2/OneDrive/Рабочий стол\"), \n .     \"ecoli_proteins.fa\")",
            "5. setwd(\"C:/Users/my comp2/OneDrive/Рабочий стол\")"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "install.packages(\"usethis\")\n",
        "library(usethis)\n",
        "\n",
        "\n",
        "repo_url <- \"https://github.com/PacktPublishing/R-Bioinformatics-Cookbook.git\"\n",
        "\n",
        "\n",
        "system(sprintf(\"git clone %s\", repo_url), intern = TRUE)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 104
        },
        "id": "aR9ANbXD4qSj",
        "outputId": "5350fdd5-9369-47cf-be40-80db709e4f4e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "Warning message in system(sprintf(\"git clone %s\", repo_url), intern = TRUE):\n",
            "“running command 'git clone https://github.com/PacktPublishing/R-Bioinformatics-Cookbook.git' had status 128”\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [],
            "text/markdown": "",
            "text/latex": "",
            "text/plain": [
              "character(0)\n",
              "attr(,\"status\")\n",
              "[1] 128\n",
              "attr(,\"errmsg\")\n",
              "[1] \"Resource temporarily unavailable\""
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "setwd('/content/R-Bioinformatics-Cookbook')\n"
      ],
      "metadata": {
        "id": "VtDL6Y3VoOfB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
        "  install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"ape\")\n",
        "\n",
        "library(ape)\n",
        "\n",
        "newick <- read.tree(file.path(getwd(), \"datasets\", \"ch4\", \"mammal_tree.nwk\"))\n",
        "l <- subtrees(newick)\n",
        "plot(newick)\n",
        "plot(l[[4]], sub = \"Node 4\")\n",
        "\n",
        "small_tree <- extract.clade(newick, 9)\n",
        "\n",
        "new_tree <- bind.tree(newick, small_tree, 3)\n",
        "plot(new_tree)\n"
      ],
      "metadata": {
        "id": "4cEgprHxcQ_R"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Finding unannotated transcribed regions**"
      ],
      "metadata": {
        "id": "WMbKobq2azjo"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "BiocManager::install(\"SummarizedExperiment\")\n",
        "BiocManager::install(\"IRanges\")\n",
        "BiocManager::install(\"csaw\")\n",
        "BiocManager::install(\"rtracklayer\")\n",
        "get_annotated_regions_from_gff <- function(file_name) {\n",
        "  gff <- rtracklayer::import.gff(file_name)\n",
        "  as(gff, \"GRanges\")\n",
        "}\n",
        "\n",
        "get_annotated_regions_from_bed <- function(file_name){\n",
        "  bed <- rtracklayer::import.bed(file_name)\n",
        "  as(bed, \"GRanges\")\n",
        "}\n",
        "\n",
        "#get counts in windows across whole genome\n",
        "whole_genome <- csaw::windowCounts(\n",
        "    file.path(getwd(), \"datasets\", \"ch1\", \"windows.bam\"),\n",
        "    bin = TRUE,\n",
        "    filter = 0,\n",
        "    width = 500,\n",
        "    param = csaw::readParam(\n",
        "        minq = 20,\n",
        "        dedup = TRUE,\n",
        "        pe = \"both\"\n",
        "    )\n",
        ")\n",
        "colnames(whole_genome) <- c(\"small_data\")\n",
        "\n",
        "#make a GRanges object of the annotated features\n",
        "annotated_regions <- get_annotated_regions_from_gff(file.path(getwd(), \"datasets\", \"ch1\", \"genes.gff\"))\n",
        "\n",
        "\n",
        "library(IRanges)\n",
        "library(SummarizedExperiment)\n",
        "#find the overlaps between the windows we made\n",
        "windows_in_genes <-IRanges::overlapsAny(\n",
        "  SummarizedExperiment::rowRanges(whole_genome),\n",
        "  annotated_regions\n",
        ")\n",
        "\n",
        "annotated_window_counts <- whole_genome[windows_in_genes,]\n",
        "non_annotated_window_counts <- whole_genome[ ! windows_in_genes,]\n",
        "\n",
        "\n",
        "assay(annotated_window_counts)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "HuWrIqsoW2Ze",
        "outputId": "14356301-d019-4e9b-dc02-647369258376"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'SummarizedExperiment'”\n",
            "Old packages: 'bit', 'brew', 'brio', 'bslib', 'cli', 'cpp11', 'curl',\n",
            "  'data.table', 'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven',\n",
            "  'highr', 'isoband', 'jsonlite', 'openssl', 'pkgbuild', 'processx',\n",
            "  'progress', 'ragg', 'roxygen2', 'sass', 'scales', 'stringi', 'svglite',\n",
            "  'testthat', 'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml',\n",
            "  'boot', 'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'IRanges'”\n",
            "Old packages: 'bit', 'brew', 'brio', 'bslib', 'cli', 'cpp11', 'curl',\n",
            "  'data.table', 'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven',\n",
            "  'highr', 'isoband', 'jsonlite', 'openssl', 'pkgbuild', 'processx',\n",
            "  'progress', 'ragg', 'roxygen2', 'sass', 'scales', 'stringi', 'svglite',\n",
            "  'testthat', 'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml',\n",
            "  'boot', 'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Warning message:\n",
            "“package(s) not installed when version(s) same as or greater than current; use\n",
            "  `force = TRUE` to re-install: 'csaw'”\n",
            "Old packages: 'bit', 'brew', 'brio', 'bslib', 'cli', 'cpp11', 'curl',\n",
            "  'data.table', 'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven',\n",
            "  'highr', 'isoband', 'jsonlite', 'openssl', 'pkgbuild', 'processx',\n",
            "  'progress', 'ragg', 'roxygen2', 'sass', 'scales', 'stringi', 'svglite',\n",
            "  'testthat', 'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml',\n",
            "  'boot', 'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.18 (BiocManager 1.30.22), R 4.3.2 (2023-10-31)\n",
            "\n",
            "Installing package(s) 'rtracklayer'\n",
            "\n",
            "also installing the dependencies ‘rjson’, ‘XML’, ‘GenomicAlignments’, ‘BiocIO’, ‘restfulr’\n",
            "\n",
            "\n",
            "Old packages: 'bit', 'brew', 'brio', 'bslib', 'cli', 'cpp11', 'curl',\n",
            "  'data.table', 'desc', 'devtools', 'fansi', 'gargle', 'gert', 'haven',\n",
            "  'highr', 'isoband', 'jsonlite', 'openssl', 'pkgbuild', 'processx',\n",
            "  'progress', 'ragg', 'roxygen2', 'sass', 'scales', 'stringi', 'svglite',\n",
            "  'testthat', 'textshaping', 'vctrs', 'vroom', 'whisker', 'xml2', 'yaml',\n",
            "  'boot', 'cluster', 'Matrix', 'nlme', 'rpart'\n",
            "\n",
            "Loading required package: BiocGenerics\n",
            "\n",
            "\n",
            "Attaching package: ‘BiocGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:stats’:\n",
            "\n",
            "    IQR, mad, sd, var, xtabs\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
            "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
            "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
            "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
            "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
            "    table, tapply, union, unique, unsplit, which.max, which.min\n",
            "\n",
            "\n",
            "Loading required package: S4Vectors\n",
            "\n",
            "Loading required package: stats4\n",
            "\n",
            "\n",
            "Attaching package: ‘S4Vectors’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:utils’:\n",
            "\n",
            "    findMatches\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:base’:\n",
            "\n",
            "    expand.grid, I, unname\n",
            "\n",
            "\n",
            "Loading required package: MatrixGenerics\n",
            "\n",
            "Loading required package: matrixStats\n",
            "\n",
            "\n",
            "Attaching package: ‘MatrixGenerics’\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
            "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
            "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
            "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
            "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
            "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
            "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
            "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
            "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
            "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
            "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
            "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
            "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
            "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
            "    rowWeightedSds, rowWeightedVars\n",
            "\n",
            "\n",
            "Loading required package: GenomicRanges\n",
            "\n",
            "Loading required package: GenomeInfoDb\n",
            "\n",
            "Loading required package: Biobase\n",
            "\n",
            "Welcome to Bioconductor\n",
            "\n",
            "    Vignettes contain introductory material; view with\n",
            "    'browseVignettes()'. To cite Bioconductor, see\n",
            "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
            "\n",
            "\n",
            "\n",
            "Attaching package: ‘Biobase’\n",
            "\n",
            "\n",
            "The following object is masked from ‘package:MatrixGenerics’:\n",
            "\n",
            "    rowMedians\n",
            "\n",
            "\n",
            "The following objects are masked from ‘package:matrixStats’:\n",
            "\n",
            "    anyMissing, rowMedians\n",
            "\n",
            "\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<table class=\"dataframe\">\n",
              "<caption>A matrix: 6 × 1 of type int</caption>\n",
              "<thead>\n",
              "\t<tr><th scope=col>small_data</th></tr>\n",
              "</thead>\n",
              "<tbody>\n",
              "\t<tr><td> 0</td></tr>\n",
              "\t<tr><td>31</td></tr>\n",
              "\t<tr><td>25</td></tr>\n",
              "\t<tr><td> 0</td></tr>\n",
              "\t<tr><td>28</td></tr>\n",
              "\t<tr><td>31</td></tr>\n",
              "</tbody>\n",
              "</table>\n"
            ],
            "text/markdown": "\nA matrix: 6 × 1 of type int\n\n| small_data |\n|---|\n|  0 |\n| 31 |\n| 25 |\n|  0 |\n| 28 |\n| 31 |\n\n",
            "text/latex": "A matrix: 6 × 1 of type int\n\\begin{tabular}{l}\n small\\_data\\\\\n\\hline\n\t  0\\\\\n\t 31\\\\\n\t 25\\\\\n\t  0\\\\\n\t 28\\\\\n\t 31\\\\\n\\end{tabular}\n",
            "text/plain": [
              "     small_data\n",
              "[1,]  0        \n",
              "[2,] 31        \n",
              "[3,] 25        \n",
              "[4,]  0        \n",
              "[5,] 28        \n",
              "[6,] 31        "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "ccR979UuZGgs"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}